MSC4133: Extending User Profile API with Key:Value Pairs (#4133)
matrix-org/matrix-spec-proposals: proposals/4133-extended-profiles.md

Author: Thanks for this. I've adopted the specific edits mentioned, and also reorganised sections and corrected some unusual formatting/phrasing in the hope that it helps make the overall document a bit easier.

Reviewer: Moving to a thread for discussion to take place - @MTRNord said:
> So I am still not sure if this is a good idea.
> 
> I do like that we can represent timezones. I do not (and I am aware this was moved to $later) think this msc works well for many other fields. Pronouns for example are by nature freetext which means that MSC reintroduces one of the core issues which this MSC initially had. Imho dropping it here just delays this concerns but did not address it.
> 
> I am also not sure why there was no move towards instead “forking”/continuing the profiles as rooms instead. Why do we need to do it like this and compromise extremely on this? Why couldn't the time have been used on the other thing?
> 
> Missing is also how unstable/client specific namespaced fields can be moderated by room admins. This brings me back to my past concerns in my other comments. This again essentially allows unmoderated freetext fields which even more confusing may only show up in some clients. So it could a) target specific user groups and b) a room admin may not even see it and have people complain. This makes T&S quite a bit harder for room admins. For server admins it obviously is not an issue.
> 
> Last but not least I fear feature creep with this. Will we have now a UI related MSC for every field here? If I want to for example want to have my gender in this would this need to be another MSC because other want this? This feels like we are starting to micro manage the spec and I am not sure this is a good idea.

Reviewer: I'm failing to understand how this MSC creates problems for pronouns specifically. This MSC defines a *generic* key/value structure, where the value can be pretty much any type. This could be a user-supplied string, an object, or a boolean value depending on the key's requirements. If you have specific concerns with how the MSC is interfering with other MSCs, like pronouns, please leave inline comments for review.

This MSC also states it *complements* profiles as rooms, which is true: one of the concerns with profiles as rooms is it tries to do too much. By accepting this MSC, we're agreeing that arbitrary key/value pairs should be supported, which makes other MSCs easier to land. In essence, this is already a continuation of the work, just at an earlier stage than the current MSCs in the area. Forking the effort feels incorrect, given all the authors involved would like to work with each other towards a solution. If there's specific language which doesn't feel in line with this, please leave an inline comment for review.

As for room moderation: profile fields are deliberately kept at the user level in this MSC. A future MSC may introduce them to the room/member level as well. This makes the T&S model one where the server admin is responsible for the user's profile, not the rooms they are participating in. Room admins still have the option to eject users they don't feel fit their room's purpose, and can operate bots which scan profiles for potentially abusive content if they wish. Meanwhile, abusive users and their profiles should be reported to the homeserver admin, which this MSC calls out as the preferred mechanism for dealing with issues. A future, unrelated, MSC will make this easier. If there's specific areas which could be improved further, inline comments are appreciated for further review.

Creating an MSC for every feature may be what happens at the start of the transition towards this key/value system, but as we gain experience in reviewing these MSCs we'll be better able to calibrate the types of things which should be in the spec and which are best covered as unstable extensions. For example, we may prefer a general `m.bio` field over lots of fields for gender, websites, contact info, and interests. Or we may prefer to have a dedicated field for gender, and combine the rest under `m.bio`. These types of concerns are best expressed on the individual MSCs which get opened after this MSC lands - the influx is expected, and something we'd need to manage regardless.

Reviewer 2: I am currently on the move and cant draft a full reply in the meanwhile many of my concerns also have been raised by gnuxie at https://github.com/matrix-org/matrix-spec-proposals/pull/4133#discussion_r1926938930 which likely answer some of the questions. I will write a proper reply once i am back at a computer :)

Reviewer 2: I am currently on the move and cant draft a full reply in the meanwhile many of my concerns also have been raised by gnuxie at https://github.com/matrix-org/matrix-spec-proposals/pull/4133#discussion_r1926938930 which likely answer some of the questions. I will write a proper reply once i am back at a computer :)

Reviewer 2: I am currently on the move and cant draft a full reply in the meanwhile many of my concerns also have been raised by gnuxie at https://github.com/matrix-org/matrix-spec-proposals/pull/4133#discussion_r1926938930 which likely answer some of the questions. I will write a proper reply once i am back at a computer :)

Reviewer 3: > I am also not sure why there was no move towards instead forking/continuing profiles as rooms instead.
You mean #4201? Which exists just to give PaR another try.

Reviewer 4: @gnuxie wrote:

> Why is this MSC being considered in isolation of any specified custom profile field?
> 
>   * All fields contain user generated content.
> 
>   * No custom profile field has sufficiently demonstrated itself (via the spec process or otherwise).
> 
>   * There are no fields defined in this MSC
> 
>   * All fields used in the implementation are currently under review in other MSCs and have not demonstrated themselves.

Reviewer 4: Whilst we do make an effort to avoid speccing stuff that doesn't have a demonstrated purpose, there is a balance to make between that, and allowing things to land in manageable pieces. Historically, some features have struggled to make it through the spec process because their scope was just too broad. (MSC1849 was an example of that.) 

Reviewing large MSCs is really hard (as is reviewing a large PR of any kind), just because everybody has to keep lots of context in their head for weeks or even months, and breaking them up helps build stable foundations before moving onto the next storey.

Further: in the case of this particular MSC, I see it mostly as a generalisation of the existing API. Even if we never end up landing anything like MSC4175, it doesn't feel like this MSC is adding undue clutter to the API.

Reviewer 5: I think this has been covered by the other two threads https://github.com/matrix-org/matrix-spec-proposals/pull/4133#discussion_r1927321286 and https://github.com/matrix-org/matrix-spec-proposals/pull/4133#discussion_r1927322022. So it can be resolved.

Reviewer 4: @Gnuxie wrote:

> The MSC does not propose anything in itself except a place to store user generated content... without demonstrating consideration for user generated content, especially free-text content.
> 
> * Free-text content was originally part of the MSC, but got moved out to another proposal ([MSC4208](https://github.com/matrix-org/matrix-spec-proposals/pull/4208)) to "resolve" numerous blocking concerns about it.  But these concerns will appear in almost every single follow-up MSC related to specifying any field that builds on this proposal.
> 
> * The MSC still has a provision to display and enter custom fields:
>   > Clients MAY provide a UI for users to view and enter custom
>   > fields, respecting the appropriate namespaces.
> 
> * The demonstrated implementation uses [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175) which is a parsable field with a very narrow range of values. But this proposal is clearly intended to be followed up with fields for unparsable user generated content. And we expect there to be many more unparsable fields than restricted and narrowly defined ones.
>
> I believe the MSC should be modified to address the concerns with free-text content. And also demonstrate implementation within a client that does add free-text fields.
> 
> The discussion will have to happen with subsequent MSCs unless they define fields that are both very limited and restricted, which is obviously not the intent of this proposal. So it does not make sense to discuss this proposal in isolation of those concerns.

Reviewer 4: I don't think I really understand the concern here. It feels like you're arguing against this MSC because you're concerned that people may, in the future, use it as a basis for other MSCs which bring problems.  You've acknowledged yourself that such problems are not inherent by pointing to MSC4175, so it seems inappropriate to express your concerns on this MSC. If you have concerns with MSC4208, then they belong on MSC4208, not here.

> * Free-text content was originally part of the MSC, but got moved out to another proposal ([MSC4208](https://github.com/matrix-org/matrix-spec-proposals/pull/4208)) to "resolve" numerous blocking concerns about it.  But these concerns will appear in almost every single follow-up MSC related to specifying any field that builds on this proposal.

Yeeess, but then the time to discuss them is on those MSCs, not here.

> * The MSC still has a provision to display and enter custom fields:
>       > Clients MAY provide a UI for users to view and enter custom
>       > fields, respecting the appropriate namespaces.

To be clear, I'm reading this as: "clients MAY provide a UI for users to view and enter *their own* profile fields". On that basis, it's hard to see why this, on the face of it, is any different to, say, allowing users to send freetext events.

I must admit I'm not that familiar with the T&S space, so it's entirely possible I'm missing something here, but in discussions I've had with people who are more familiar with it than me, they've not been concerned about letting users *set* profile data: it's when clients start showing the profile data of other users that you need to be concerned. I'd be happy to be set straight here.

Reviewer 6: I've just reread this MSC again, and would agree that we should abuse concerns for specific field formats in the MSC which defines that specific field.  In terms of the custom fields in this MSC (e.g. someone publishing `m.nasty.stuff: true` on their profile) then yes, some clients may expose that to other users, and then it would surely be handled like any other abusive avatar or displayname today: you'd ignore or otherwise moderate the user. So, we can use the abuse potential of today's displayname/avatars as a precedent for how you'd handle this, and don't need to block on defining new mechanisms.

In other words, rather than waiting on future MSCs for new profile field types to land which flesh out antiabuse considerations which might then inform this MSC, let's keep *some* forwards momentum and actually land this as is - we can always have a future MSC for "antiabuse mechanisms common to different extensible profile fields" or something, if needed.

Author: If it helps, that's roughly what I'm looking to achieve with #4208 and Jim has offered to help shape it, so I'm hoping we're off to a good start.

The hope is that we can define the dependencies for freetext content in profiles to be commonly allowed in clients, then other profile field MSCs have an easy reference to quote (or use as a dependency) to make sure they're also safe.

As always, I welcome feedback and would love to gather different perspectives so I can start translating that into firm technical requirements, then figure out whether MSCs already exist (or need to) for the dependencies.

Reviewer 5: > You've acknowledged yourself that such problems are not inherent by pointing to [MSC4175](https://github.com/matrix-org/matrix-doc/pull/4175), so it seems inappropriate to express your concerns on this MSC

But MSC4175 is a very simplified example, and it would be fair to suggest that this MSC would not be considered in the same way if its only function was to provide the timezone field or similar narrowly defined ones, such as a flag for whether the user is a bot. Therefore it does seem appropriate for me to express these concerns on this MSC. 

>  it's hard to see why this, on the face of it, is any different to, say, allowing users to send freetext events.

Because there is already a significant amount of safety tooling for handling user generated content in events and all clients have been built with that in mind. With the profiles we are creating a new surface and new ways for users to interact. @turt2live has rightfully suggested in another comment https://github.com/matrix-org/matrix-spec-proposals/pull/4133#discussion_r1926148709 that bots can scan profiles for abusive content, but this isn't something anyone has implemented at the moment. And we don't yet understand from this proposal and its implementations how profiles will be used for spam or abuse in Matrix's public rooms.

> they've not been concerned about letting users set profile data: it's when clients start showing the profile data of other users that you need to be concerned. 

Yes this is true, but it is clear that clients will need to display profile data for this MSC to have utility.

Reviewer: It allows the behaviour because that's how most clients expose profiles to the user, but that doesn't make it part of the room. Many of the options already available on the opened card/pane are global actions, independent of the community the room is part of. 

This feature is more akin to presence, where it is displayed adjacent to room information, but ultimately the room's moderators are not responsible for banning people who set abusive status messages.

Reviewer 5: >  but ultimately the room's moderators are not responsible for banning people who set abusive status messages.

I'm not sure what you mean by responsibility here but room moderators will feel obligated to ban room members with abusive status messages. The only reason this doesn't already happen frequently is because it has been rare for clients to display status messages anywhere at all.

Reviewer 2: > but ultimately the room's moderators are not responsible for banning people who set abusive status messages.

This might be the current design but after all room moderators are the firdt line of defense. If someone has a $illegalThing status and someone sees it in the room member list there is a high chance that they will initially ask the room admin to do soemthing about that user.

Today this leaves 2 options: ban that user in the rooms or mediate a cobtact to the server admin (hoping the server admin itself isnt part of $illegalThing themself).

The big feature gap with that and this msc however is the reporting. Imho manual reporting in the room chat is not a satisfactory solution. Especially as some bot users will get triggered by that and start to issue spam. Which makes the already bad situation even more bad.

Author: If it helps, that MSC exists in #4202, which will most likely be a dependency of #4208

Reviewer: We can put reporting the user higher on the list independent of FCP here.

Reviewer 2: > If it helps, that MSC exists in #4202, which will most likely be a dependency of #4208

Imho ot doesnt help as there will be a multi month gap between that most likely simply due to the spec release cycle + implementation cycles. And the moderation issues are immediate ones. Not ones which we can fix later. That happenes in the past (example being reporting of messages not being federated for years) and caused countless problems for Moderators.

Reviewer: I've opened https://github.com/matrix-org/matrix-spec-proposals/pull/4260 with my T&S hat to get some short-term reporting options in place, while MSC4202 and similar are discussed. 

Implementation of MSC4260 is happening right now, and proposed-FCP is expected shortly after. 

I don't think it's feasible to block MSC4133 on a full and comprehensive reporting system, but we can increase baseline reporting mechanisms for the short term.

Reviewer: (I'm considering this thread resolved - if there's still outstanding thoughts, please re-raise them in a new thread or on the appropriate MSC)

Reviewer 5: @turt2live Even with a full and comprehensive reporting system you have still not addressed room moderators.  https://github.com/matrix-org/matrix-spec-proposals/pull/4133#discussion_r1935822840

Room moderators may not have a legal responsibility to cleanup profiles from users in their rooms. But it's very poor to use that as an excuse to ignore them. They still have an obligation to clean them up for their room members. And they should be allowed to stop the profile fields being pulled in preemptively rather than as a reaction on an individual basis (by banning users).

Reviewer: Room moderators aren't being ignored. In the T&S model for Matrix, the default responsibility for a user profile is the user's homeserver, not the rooms they participate in. Some rooms may still go above and beyond that model, but that is an elective choice. Those rooms may need to develop or acquire tools which help them accomplish that. Those tools don't block this MSC from progressing through FCP, though.

Reviewer 5: >  In the T&S model for Matrix

can you show me this model?

Reviewer: It's not written down in a useful format at the moment. Most aspects are covered implicitly in the spec itself, and other parts will be clarified as reporting v2 progresses further. 

The Trust & Safety Committee under the Foundation's Governing Board may be best positioned to write this down.

Reviewer 5: OK, so it does not currently exist and needs to be created by the Trust & Safety Committee?

Reviewer: It exists as an idea with the T&S team, but the Committee would be best positioned to review it and write it down.

Reviewer 5: The problem is @turt2live that the Matrix safety model could also say things like "It is discouraged for room admins to allow users to join rooms publicly from the open federation". To avoid making safety guarantees, and to avoid tackling tough problems like this one. And we wouldn't know that, and everyone on the open federation would still be doing that. And being actively encouraged by the foundation's press communications, but not their technical ones.

Reviewer: This is well beyond the scope of this MSC, but the T&S model is more about who handles a given report than specific recommended actions. There's some definitions which may change how reports are routed, but the model wouldn't go to the extreme of "don't allow people to join rooms". 

For example, some arbitrary definition of "community moderators" should receive and process reports from users in their rooms, at least where the server isn't involved. We need to ensure that tooling exists to allow them to make decisions related to those reports, and also ensure that reports can even go to the right people in the first place.

Reviewer 4: @Gnuxie wrote:

> Why is there urgency to merge this MSC?
>        
>  * Clients can currently use the proposal to implement custom fields under the unstable namespace without any changes or hindrance. Some of them already have done, so there should be no urgency to get this merged.
>  * Similar Concerns have been suppressed in Matrix rooms as bikeshedding or blocking the MSC. All developers are free to implement the proposal as is under the unstable flag. When the proposal becomes part of the specification it does not change the number of implementations, nor does it allow dependant MSCs to make any further progress.

Reviewer 4: Firstly, I'm sorry if your concerns haven't been heard. I don't think I've seen the discussions you refer to, and I wasn't aware of them before you raised them here. This is one reason we ask that concerns on MSCs be raised as comments on the MSC itself: conclusions in Matrix rooms get lost. Thank you for doing so now.

Secondly: I don't really see this as "urgency". The MSC has been 9 months in the making, and it feels like a relatively incremental change to the existing API. Honestly, it felt ready to land. I'll refer you to my earlier points that it can be very helpful to land MSCs incrementally rather than having a large amount to review all at once.

As the SCT, we've also had feedback in the past that it can be unreasonably difficult to get simple changes to the spec through the MSC process. There are numerous anecdotes from people who have just found the MSC process too arduous and have walked away in frustration. We have a balance to pick between rushing things into the protocol without due diligence, and letting the protocol ossify because changing it is just to damn hard.

All that said, you're right that this isn't actively blocking anything, and it's certainly not our goal to steamroller it through against significant opposition. On that basis: thank you again for raising your concerns. This is the FCP working as it should: we've put the FCP timer at least on hold for now and we'll see what can be done to address the concerns.

Whilst we can't promise that everyone is going to be delighted by every MSC that is accepted, it is important to make sure that members of the Matrix community have the opportunity to voice their concerns and that we take the time to consider those concerns properly.

Reviewer 5: >  I'll refer you to my earlier points that it can be very helpful to land MSCs incrementally rather than having a large amount to review all at once.

I think this is what I mean by urgency. It does make sense that you would want to review MSCs in smaller pieces. And it does seem like at the moment the only formal tool to review and freeze MSCs is by landing them into the specification.

> As the SCT, we've also had feedback in the past that it can be unreasonably difficult to get simple changes to the spec through the MSC process. There are numerous anecdotes from people who have just found the MSC process too arduous and have walked away in frustration. We have a balance to pick between rushing things into the protocol without due diligence, and letting the protocol ossify because changing it is just to damn hard.

Yes, this is a valid concern. Though I think there is another factor at play here, which is that clients may implement Matrix features selectively regardless of whether the features are specified in a spec release or an MSC. And at least some frustration comes from lack of implementation in a range of clients. Now I could misunderstand what people's frustrations have been, but to me that seemed to be close to the root. 

The thing is this MSC does not bring any feature directly to the end user in Matrix clients. Do we believe that landing MSC4133 in specification will make a difference to the uptake of the dependent MSCs (which are still iterating) that will? It just seems unlikely to me.

On the other hand, we do need to keep MSC authors satisfied that their work is making progress, is valued. And I will concede that while contribution is scarce an optimistic merge makes sense. But there could be other ways of assuring and rewarding authors.

Reviewer: Not all MSCs are required to bring features to clients. Architectural MSCs like this are still helpful for furthering the protocol. Landing them incrementally avoids a situation where projects use the unstable endpoints indefinitely (effectively forking the protocol), and another where the feature is killed under process and debate. As the SCT, we need to balance not only speed of the process, but also risk of de-facto spec. We typically favour to move faster where we can, because once something becomes anywhere close to de-facto spec, it's a truly painful experience to remedy. 

Features like extensible profiles (of any variety) have an extremely high potential for de-facto spec if left unmanaged or delayed by the SCT. Client developers in particular have been asking for such a feature for *years* so they can compete with other platforms, and so they can bring joy to their users. Now that the MSC is implemented and clients are starting to pick it up, there's serious risk that they'll just ignore the spec and continue to use those unstable implementations to bring it to their users. 

It's also not clear to me at this stage if we'd actually push this as far as *merging* into the spec. Similar to Extensible Events, we may stop short until 'enough' of the feature is used by accepted MSCs to allow for changes to be made more easily. I suspect we'll still go ahead and allow it to be merged to the spec anyway, particularly to encourage the ecosystem to explore use cases (see [my other comment](https://github.com/matrix-org/matrix-spec-proposals/pull/4133/files#r1932904934)). If we find that we need a bunch more work after that, we'll prioritize bug fix MSCs highly. 

In any case, nearly all downstream MSCs, including profiles as rooms, would rely on this MSC's system existing to provide maximum compatibility or just a foundational, first-party, mechanism to share information.

Reviewer 5: I understand that SCT has a lot to balance here. And I don't have answers for how to change the spec process, and it wouldn't be appropriate to either just for the sake of one MSC. So with that I do concede and you can mark the thread as resolved. 

> Now that the MSC is implemented and clients are starting to pick it up, there's serious risk that they'll just ignore the spec and continue to use those unstable implementations to bring it to their users.

As an aside i don't really understand why this is a risk? Especially for an MSC like this that has been built collaboratively with a lot of feedback from the SCT. For other MSCs where the authorship is less collaborative it could see why that would be more of an issue. 

>  there's serious risk that they'll just ignore the spec and continue to use those unstable implementations to bring it to their users.

This is true, but the same risk also applies for implementations based on versions of the spec. 

> It's also not clear to me at this stage if we'd actually push this as far as merging into the spec. Similar to Extensible Events, we may stop short until 'enough' of the feature is used by accepted MSCs to allow for changes to be made more easily

that makes sense

Reviewer: Off-spec largely means implementing things that don't reside on spec.matrix.org - collaboration with the SCT largely only helps get things moving through the process, but doesn't prevent forks of the protocol when MSCs get stuck in FCP.

Reviewer: (I'm considering this resolved - if there's further questions about the spec process, it's probably best to raise them in #matrix-spec rather than here)

  be published by the user about themselves without creating state events or other moderation issues.

## Potential Issues

Initial challenges are primarily ensuring uniform support for the `extended` field across different servers
and clients during the rollout phase: users may come to expect that other users will check their supported
languages before communicating with them, but the sender's server does not support extended profiles.

As such, this MSC is designed to be as simple as possible to get initial functionality and data structures
implemented widely in both clients and servers, so further extensions can be debated, implemented, and tested
over time in the future.

As this data is stored only at the global level, it won't allow users to modify fields per-room, or track
historical changes in profile fields. However, this is for performance and moderation reasons, as many users
will struggle to maintain many fields of personal data between different rooms, and publishing state events
for every field change could become an additional burden on servers and moderators.

## Alternatives

> An alternative could be to allow for more specific endpoint modifications or to introduce a completely new

Reviewer 7: For those of us with extremely long memories of this feature request, a semi-alternative is discussed in https://github.com/matrix-org/matrix-spec-proposals/pull/1769.

I'm not actually sure the two things here are incompatible, since 1769 is more about the storage mechanism (rooms) rather than the API surface.

Anyway, I think it's worth mentioning where the thinking went before.

Author: Indeed, it was talking about profile rooms that led me to propose this.

Profile rooms are considered dependent on a number of other MSCs, including fast joins and peeking, and that's taken years so far.

The objective of this MSC is to offer a far more basic level of service (that would be very simple to implement in servers and clients) to allow people to start designing UI, bots, and other features, that could fairly easily be migrated to use profile rooms later should that proposal finally make it into stable.

I also suspect a number of servers will ultimately disable/restrict profile rooms, but may allow this feature to pass (even just with a whitelist of keys/values) so offering both options could be a way to offer a "Extendable Profiles Lite" even once profile rooms are live. 

I'm certainly happy to explain some of this background - I wasn't sure if adding the extra detail would be distracting!

Reviewer 7: Yep, I agree. This MSC should happen first at any rate. I'd probably put in a mention to say that these approaches are compatible, should anyone be wondering why we're going for one rather than the other.

Author: I've added [this](https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/c0bc005d898eb099ee9e4fb85644300902704a1b) to clarify how this MSC relates to [MSC1769](https://github.com/matrix-org/matrix-spec-proposals/pull/1769) - does that look ok?

Reviewer 8: My 2c: It would be great to have both of these features.  We have played around with a (limited) version of profile rooms in Circles, and one of the things that's leading to us cutting that feature is that we have no nice way for a user to share the room id of their profile room.  With this MSC, it would be easy - just put the profile room id in the extended profile.

Author: Indeed, I very much am trying to avoid making this the "kitchen sink" profile, but more a very basic way to be able to publish arbitrary bits of data in a simple way without history. In a communication platform, it's important to have plenty of options for publishing bits of data to/from the people we're communicating with, we just need to make sure they're voluntary and respect users' privacy!

API specifically for extended information. However, these approaches could lead to more significant changes
to the existing API and higher complexity for client developers.

> ## Security Considerations

Reviewer 7: We may want to put a "sensible" limit on the size and number of keys on a profile, to avoid potential issues.

Author: Very good point. Perhaps a maximum total number of keys of any depth, but also a maximum depth for sanity?

Reviewer 7: Thinking about this, maybe we just do what we do for events and cap this to 64KiB overall https://spec.matrix.org/v1.10/client-server-api/#size-limits. Whether or not we still want a depth limit, I'm not sure.

Author: I've clarified the entire profile JSON should not be larger than 64KiB, and that servers are allowed to reject JSON received for remote users if it is larger than 64KiB. Is that suitable?

Reviewer 7: Seems reasonable to me!

arbitrary key:value pairs at the global level.

This proposal does not seek to enforce the content or usage of these fields but rather to add a
framework for users to have extra data that can be further clarified and extended in the future as
community usage of these fields grows.

Homeservers could disable the ability for users to update these fields, or require a specific list
of fields, but the intention of this proposal is that users will be presented with a form to enter
their own free-text fields and values. After using these very flexible fields, the community may
opt to request a further extension to promote one or more fields to be specified per-room via
member events.

### Client-Server API Changes

1. **GET `/_matrix/client/v3/profile/{userId}/{key_name}`**: This endpoint will replace the existing
   profile endpoints. It will return the value of the specified `key_name`:

    ```json
    {
      "key_name": "field_value"
    }
    ```

    For example, requesting `/_matrix/client/v3/profile/@alice:matrix.org/displayname` would return:

    ```json
    {
      "displayname": "Alice"
    }
    ```

> 2. **PUT `/_matrix/client/v3/profile/{userId}/{key_name}`**: This endpoint will set the value of the

Reviewer 9: To avoid the issue we have with account data, should there be a corresponding `DELETE` endpoint? Could be a follow up MSC if you don't want to expand though.

Author: That's a great suggestion - I completely missed this, as we were previously overwriting the entire profile each time. I've added that now, hopefully this looks ok?

   specified `key_name`:

    ```json
    {
      "key_name": "new_value"
    }
    ```

    For example, setting `/_matrix/client/v3/profile/@alice:matrix.org/displayname` with:

    ```json
    {
      "displayname": "Alice Wonderland"
    }
    ```

3. **GET `/_matrix/client/v3/profile/{userId}/`**: This endpoint will retrieve all profile fields:

    ```json
    {
      "avatar_url": "mxc://matrix.org/MyC00lAvatar",
      "displayname": "John Doe",
      "custom_field1": "value1",
      "custom_field2": ["value2", "value3"]
    }
    ```

4. **POST `/_matrix/client/v3/profile/{userId}`**: This endpoint will accept a complete JSON object
   to replace the entire profile:

    ```json
    {
      "avatar_url": "mxc://matrix.org/MyNewAvatar",
      "displayname": "John Doe",
      "custom_field1": "new_value1",

>       "custom_field2": ["new_value2", "new_value3"]

Reviewer 9: I expected these to be reverse DNS unstable identifiers as we normally use for event fields. I see no mention of that, however.

Additionally, clarity on the data values should probably be added. It seems this expands it to more than strings. What happens if you set an invalid data type?

Author: I wasn't sure how best to group the unstable stuff, so it's all together in a "Unstable prefix" section at the bottom of the document, which lists the unstable fields looking like this:

```json
{
    "avatar_url": "mxc://matrix.org/MyC00lAvatar",
    "displayname": "John Doe",
    "uk.tcpip.msc4133.custom_field1": "field_value",
    "uk.tcpip.msc4133.custom_field2": ["one value", "another value"]
}
```

Is this ok, or would you prefer the document be refactored to have each unstable alongside the stable version?

Reviewer 9: So the unstable section should probably only contain the bit about the endpoint and capabilities prefixes.

Author: Got it! Thanks for clarifying, I have added a section around namespaces (borrowing heavily from [Room Tagging](https://spec.matrix.org/v1.10/client-server-api/#events-12) in the spec) and also used the opportunity to specify a maximum key length.

I clarified that the `m.*` namespace fields should be ignored if the client does not recognise them, as these will have been entered via MSC and may have special entry/display requirements the client is not aware of.

User-defined custom fields can be any valid JSON type, and there are no special requirements for their UI/display - my assumption being that clients are more innovative than we are and will find useful ways to display/arrange these fields that we can't foresee years in advance.

Reviewer 9: > Additionally, clarity on the data values should probably be added. It seems this expands it to more than strings. What happens if you set an invalid data type?

I don't think this part was handled? I'm inclined to only accept strings.

Author: I've done this now, specifying that all custom fields must be string values, but that future MSC'd fields may use other types (including a payload example showing `m.allowed_list` containing an array) so hopefully this reasonably covers this concern? https://github.com/tcpipuk/matrix-spec-proposals/commit/5fdd69cae3884421b190d2fba804e5cc8ac5c41f

    }
    ```

> ### Distinction Between Existing and Custom Fields

Reviewer 9: Other clients (both your own and other users) will have no way to know that a profile was changed. I don't think it blocks this MSC, but we might need more changes to keep clients in sync.

Author: Agreed. I like the idea of (at some point) an MSC to make an EDU like Presence communicate that a user's global profile has changed (rather than just specifically the name/avatar) but I don't want to make the scope of this MSC too broad.

Currently it just about stays compatible with the current spec, which should help client/server developers implement the changes more gracefully before the community starts creating more MSCs to extend the functionality further!

Reviewer 9: I agree not to change scope of this MSC, but we should probably document this limitation as a potential issue!

Author: Another very good point! I've added this now: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/0b88d85375f277e500f5910ad40ed3f7ff6a5501

2. **PUT `/_matrix/client/v3/profile/{userId}/{key_name}`**: This endpoint will set the value of the
   specified `key_name`:

    ```json
    {
      "key_name": "new_value"
    }
    ```

    For example, setting `/_matrix/client/v3/profile/@alice:matrix.org/displayname` with:

    ```json
    {
      "displayname": "Alice Wonderland"
    }
    ```

3. **DELETE `/_matrix/client/v3/profile/{userId}/{key_name}`**: This endpoint will remove the key
   (and associated value) from the profile, if permitted by the homeserver.

4. **GET `/_matrix/client/v3/profile/{userId}/`**: This endpoint will retrieve all profile fields:

    ```json
    {
      "avatar_url": "mxc://matrix.org/MyC00lAvatar",
      "displayname": "John Doe",
      "custom_field1": "value1",
      "custom_field2": ["value2", "value3"]
    }
    ```

5. **POST `/_matrix/client/v3/profile/{userId}`**: This endpoint will accept a complete JSON object
   to replace the entire profile:

    ```json
    {
      "avatar_url": "mxc://matrix.org/MyNewAvatar",
      "displayname": "John Doe",
      "custom_field1": "new_value1",
      "custom_field2": ["new_value2", "new_value3"]
    }
    ```

### Server-Server API Changes

> 1. **GET `/_matrix/federation/v1/query/profile/{userId}/{key_name}`** will mirror the client-server

Reviewer 10: Will this endpoint only allow homeservers to query for users that belong to the requested homeserver? Generally should be a safe assumption but might be good to explicitly state that.

Author: Servers should definitely only be requesting profiles from the homeserver in the MXID, so I will clarify that - thank you!

I'll also clarify that servers should only ever serve profiles for their own users, rather than expect people to assume that.

Author: I've [clarified that](https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/e836057e61d8d97e01a36877b46b8deb15a59037) servers should only accept requests for a user local to that server, and that servers should only request profiles from the server specified in that user's MXID - thank you!

>    API changes to ensure profile information is consistently available across the federated network.

Reviewer 11: I think a new endpoint is redundant, the existing `/query/profile` API has an optional `field` parameter

Author: This was to replace the existing [avatar_url](https://spec.matrix.org/v1.10/client-server-api/#get_matrixclientv3profileuseridavatar_url) and [displayname](https://spec.matrix.org/v1.10/client-server-api/#get_matrixclientv3profileuseriddisplayname) endpoints, which I wouldn't want to immediately remove compatibility for.

Do you think that's reasonable, or should I specify they be removed anyway? In the meantime, I've updated the full profile endpoint to clarify it still has the `field` parameter.

Reviewer 11: I was commenting on the federation API, not the C-S API

Author: Ah, I misunderstood, sorry - I've amended this now: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/b37af654a5265dfcc18c441a0eefb0270c63f151

of fields, but the intention of this proposal is that users will be presented with a form to enter
their own free-text fields and values. After using these very flexible fields, the community may
opt to request a further extension to promote one or more fields to be specified per-room via
member events.

### Client-Server API Changes

1. **GET `/_matrix/client/v3/profile/{userId}/{key_name}`**: This endpoint will replace the existing
   profile endpoints. It will return the value of the specified `key_name`:

    ```json
    {
      "key_name": "field_value"
    }
    ```

    For example, requesting `/_matrix/client/v3/profile/@alice:matrix.org/displayname` would return:

    ```json
    {
      "displayname": "Alice"
    }
    ```

2. **PUT `/_matrix/client/v3/profile/{userId}/{key_name}`**: This endpoint will set the value of the
   specified `key_name`:

    ```json
    {
      "key_name": "new_value"
    }
    ```

    For example, setting `/_matrix/client/v3/profile/@alice:matrix.org/displayname` with:

    ```json
    {
      "displayname": "Alice Wonderland"
    }
    ```

3. **DELETE `/_matrix/client/v3/profile/{userId}/{key_name}`**: This endpoint will remove the key

>    (and associated value) from the profile, if permitted by the homeserver.

Reviewer 9: See also #3754.

Author: That seems mainly for the existing avatar and name fields - are you suggesting I should be integrating it into this to make the existing MSC redundant?

Reviewer 9: I had just forgotten it existed, so figured it should be referenced somewhere. There's a few options:

* Consider it a partial alternative.
* Add a dependency.
* Remove the deletes and let the other MSC handle it.

I'd suggest not the second option as decoupling is preferred where possible.

Reviewer 9: That's great. Thanks.

4. **GET `/_matrix/client/v3/profile/{userId}/`**: This endpoint will retrieve all profile fields:

    ```json
    {
      "avatar_url": "mxc://matrix.org/MyC00lAvatar",
      "displayname": "John Doe",
      "custom_field1": "value1",
      "custom_field2": ["value2", "value3"]
    }
    ```

5. **POST `/_matrix/client/v3/profile/{userId}`**: This endpoint will accept a complete JSON object
   to replace the entire profile:

    ```json
    {
      "avatar_url": "mxc://matrix.org/MyNewAvatar",
      "displayname": "John Doe",
      "custom_field1": "new_value1",
      "custom_field2": ["new_value2", "new_value3"]
    }
    ```

### Server-Server API Changes

> **GET `/_matrix/federation/v1/query/profile/{userId}`** will mirror the client-server API changes

Reviewer 11: ```suggestion
**GET `/_matrix/federation/v1/query/profile`** will mirror the client-server API changes
```

Author: I completely overlooked that, thank you!

Homeservers could disable the ability for users to update these fields, or require a specific list
of fields, but the intention of this proposal is that users will be presented with a form to enter
their own free-text fields and values. After using these very flexible fields, the community may
opt to request a further extension to promote one or more fields to be specified per-room via
member events.

### Client-Server API Changes

1. **GET `/_matrix/client/v3/profile/{userId}/{key_name}`**: This endpoint will replace the existing
   profile endpoints. It will return the value of the specified `key_name`:

    ```json
    {
      "key_name": "field_value"
    }
    ```

    For example, requesting `/_matrix/client/v3/profile/@alice:matrix.org/displayname` would return:

    ```json
    {
      "displayname": "Alice"
    }
    ```

2. **PUT `/_matrix/client/v3/profile/{userId}/{key_name}`**: This endpoint will set the value of the
   specified `key_name`:

    ```json
    {
      "key_name": "new_value"
    }
    ```

    For example, setting `/_matrix/client/v3/profile/@alice:matrix.org/displayname` with:

    ```json
    {
      "displayname": "Alice Wonderland"
    }

>     ```

Reviewer 3: Wouldn't it make sense for the key to be `value` instead of the field's name?
This would avoid potential future expansion clashing with existing keys (eg. a `privacy` field, `rooms`, etc. etc.)

Author: I'd far rather we maximise compatibility with current clients than a breaking change that requires all clients to change to update basic fields like name/picture.

The only way we could do that at the moment would be to add workarounds/exceptions for `displayname` and `avatar_url` which seems like a backward step when this MSC aims to offer a clean way to expand the current set of profile fields.

There's no reason this couldn't change in a future/extra MSC, which could investigate how extra metadata would be treated, but it could also just as easily add metadata keys (like "privacy" or "rooms" keys) while communicating the actual value under the name of the key.

I can imagine that being quite a contentious debate that could hold the MSC up for a while, when this one's main purpose is to get some basic profile field functionality out quickly while we wait for more advanced profile functionality (like "profile rooms") to be ironed out.

Reviewer 9: > The only way we could do that at the moment would be to add workarounds/exceptions for `displayname` and `avatar_url` which seems like a backward step when this MSC aims to offer a clean way to expand the current set of profile fields.

Another option would be to rev the endpoint version. So `v4/profile` uses `{"value": "blah"}` for arbitrary keys. `v3/profile` only supports `displayname` and `avatar_url`.

Author: That being the case, can this MSC stick with the current behaviour and a future MSC add another versioned endpoint that allows metadata for fields?

Reviewer 12: While we don't seem to have validated expansion use cases yet, I'd personally be in favor of the MSCs current behavior, too, because it limits overhead both when writing and implementing the specification.

Also, _if_ we later extend this endpoint, I suspect that would require updating client implementations anyway meaning we might not save a lot by trying to prepare for that now? 🤔

Author: I totally agree preparing for future problems is valuable, but this MSC aims to start with a simple implementation to move quickly, so that users get the extra functionality quickly while waiting for more advanced features like profile rooms to be completed.

By enabling extra custom fields in a simple way, servers and clients can get rolling, then future expansions using a new versioned endpoint can still offer a fallback on the present one, e.g. clients could still display a plain string when they don't understand timezones/pronouns/links, etc.

to ensure profile information is consistently available across the federated network.

As per the current stable endpoint, it accepts an optional `field` query string parameter to
request a single field.

As there is no method to verify the history of these fields over federation, this endpoint must
only accept requests for local users on the current homeserver, and homeservers must only request
a profile from the homeserver specified in that user's MXID.

### Capabilities

A new capability `m.set_profile_fields` will be introduced to control the ability to set custom
profile fields. The client should assume setting fields is allowed when this capability is missing.

Example capability object:

```json
{
  "capabilities": {
    "m.set_profile_fields": {
      "enabled": false
    }
  }
}
```

### Distinction Between Existing and Custom Fields

The existing fields, `avatar_url` and `displayname`, will continue to trigger state events in each
room. These fields are replicated per-room via member events. Custom fields, however, will **not**
trigger state events in rooms. They will exist solely at the global level and are intended for

> storing metadata about the user that does not need to be replicated in each room.

Reviewer 3: I would love to see the same semantics here as in MSC4069, where a client can inhibit updating the profile info in rooms.

Author: I wasn't aware of #4069 and this looks like great functionality!

I don't think either MSC really clashes with the other, only that #4069 may want a tweak to mention allowing more "special" keys than just `avatar_url` and `displayname` in the future if they're added.

Author: (it's worth noting this MSC doesn't add any new fields that are published per-room in member events, this MSC exists solely to add fields to the user's public profile at an account-wide level)

Reviewer 9: I think leaving that up to MSC4069 is best to avoid mixing concerns. 👍

- **avatar_url** and **displayname**: Changes to these fields will generate state events in all
  rooms the user is a member of.
- **Custom fields**: These are stored in the user's global profile and do not generate state events
  in rooms.

### Key/Namespace Requirements for Custom Fields

The namespace for field names is defined as follows:

- The namespace `m.*` is reserved for fields defined in the Matrix specification. Clients should
  ignore any fields in this namespace that they don't understand, as this field may have special
  entry/display requirements that are defined in the Matrix specification.
- The namespace `u.*` is reserved for user-defined fields. The portion of the string after the `u.`
  is defined the display name of this field.

> 

Reviewer 13: Should probably define how names that are not part of these namespaces should be handled. e.g. the hypothetical `org.matrix.msc9876.pronouns` mentioned below.  Presumably, they should be ignored by clients just like `m.*` names.

Do servers need to handle names in each namespace differently?

Author: My expectation was that anything with "special abilities" would be MSC'd to be defined in the spec:

For example, let's say someone MSCs `m.timezone` that requires IANA format like `Europe/Paris` that clients can have a special picker for on the editing side, then other users' clients can display  your local time in the profile (or even next to messages). The MSC could define that servers can do things with that information, but that's up to the MSC.

Anything currently going through the process would have the unstable MSC prefix, but otherwise before someone creates an MSC, if someone entered a "timezone" key manually it'd create a `u.timezone` key that has completely free entry requirements, so they might enter `CEST` or `+2` or `None` and servers/clients wouldn't be expected to do anything special with that except display it.

Does that seem reasonable, or do you think that behaviour should be different? I can try to make this clearer in the text either way! 🙂

Author: I've replaced the pronoun example with this timezone one to (hopefully) better illustrate this:
https://github.com/matrix-org/matrix-spec-proposals/commit/a11f8724957ceafcc23058fb93a5e6b42d646d51

Reviewer 13: So, if I'm understanding correctly, anything that isn't in the `u.*` namespace needs to be treated as if it's in the `m.*` namespace, since it may need special processing.  It sounds like clients need to ignore unknown names that aren't in `u.*`, and servers should return some sort of error if clients try to send an unknown name that isn't in `u.*`?

Reviewer 9: I think of it as event content -- you can put whatever keys you want in there but you might have special behavior for some keys.

I do not think servers would need to reject unknown keys but they must know how to treat two keys special (`displayname` and `avatar_url`).

Author: What Patrick said, but my assumption was also that homeserver implementations could easily choose to automatically prepend `u.` to a key if it doesn't have a prefix already (and isn't `displayname` or `avatar_url`).

Do you think I should explicitly decline some keys (e.g. if it doesn't have a prefix) or leave that up to implementations to figure out?

Reviewer 9: > my assumption was also that homeserver implementations could easily choose to automatically prepend u. to a key if it doesn't have a prefix already

I don't think that's a good idea (and the MSC doesn't say that).

Author: Sorry, that was badly phrased. Long work day!

I mean: I hadn't explicitly said in the MSC that homeservers should _deny_ keys outside the spec, so I presume people will deal with that problem in different ways... should I therefore explicitly define what should happen, or leave it as we already specify the prefixes people _should_ be using?

Reviewer 9: I think it is fairly clear -- it says that it doesn't enforce anything about the keys.

Will let @uhoreg answer though.

Reviewer 13: > I think it is fairly clear -- it says that it doesn't enforce anything about the keys.

Yes, I'm kind of asking whether we *should* enforce something about the keys.  For example, if we allow people to save anything in undefined names, people start writing random things in `m.timezone`, and later we have an MSC that defines that `m.timezone` must use one of the IANA names, what do we do with existing data?  Also, do clients need a way to detect the difference between servers that will allow anything in `m.timezone` or those that perform validation?

Personally, I think the way that I'd do it is that I'd say that servers will accept any `u.*` names, but return an error on any other name that it doesn't recognize.  Future MSCs can define additional names (such as `m.timezone`) that servers should accept.

There are other ways to do it (such as specifying how to deal with old data when defining new names, or saying that it's up to future MSCs to define such things).  I think that limiting it to `u.*` names for now is the easiest and least risky way to do it, but you could make the case for something else if you think that it's better.

Reviewer 9: But we don't do that with events or account data, etc. There will likely be profile fields that the server doesn't care about so validating them will limit the ecosystem.

Author: My main concern is that servers may not recognise all of the `m.*` namespace and limiting clients by requiring the server to recognise them could be a problem.

There's nothing in the MSC that says servers _can't_ limit which fields a user is allowed to update (so a homeserver could technically whitelist which fields a user can update if the admin so chooses) but I think hardcoding a requirement to only accept recognised `m.*` fields could prove to be something we regret later.

Reviewer 13: > My main concern is that servers may not recognise all of the `m.*` namespace and limiting clients by requiring the server to recognise them could be a problem.

Yes, there's a question of which is worse: having servers that allow everything, and then having to worry about what happens if we later add names that require special behaviour (validating values, generating events, etc.); or having servers lock it down by default and then users having to wait for their servers to support whatever new feature even if it's just a plain string with no special behaviour needed.

> There's nothing in the MSC that says servers _can't_ limit which fields a user is allowed to update (so a homeserver could technically whitelist which fields a user can update if the admin so chooses) but I think hardcoding a requirement to only accept recognised `m.*` fields could prove to be something we regret later.

IMHO, I'd prefer if homeservers behaved fairly consistently, or else it may be confusing for clients.  If a homeserver decides to gatekeep some names in its own namespace, that's fine, but if some homeservers allow everything, and some homeservers only allow `u.*` plus some hand-pick names, that would be too much deviation for my tastes.  Even if they may both be *technically* compliant.

Anyways, the main thing that I'm looking for at the moment is something indicating that the issue has been considered, and either an explanation for how the issue is addressed, or an explanation that the issue isn't a showstopper.  It sounds to me like you are of the opinion that it isn't a showstopper, and any validation issues can be handled client-side (which is fair, since clients need to validate values anyways).  And it may be sufficient to say that future MSCs that define profile fields should define how to deal with any additional issues.  Can you add something to this MSC that summarizes this, maybe in the "Potential Issues" section?

Author: I've tried to clarify in my latest updates that homeservers will not enforce the namespaces, but are allowed to forbid users from creating/modifying/deleting certain keys if configured/implemented to do so.

I've also clarified that clients may attempt to display `m.*` keys they don't "understand" but should never attempt to modify them in case they have special formatting requirements.

I have also specified that all custom fields must be string values, but that future spec'd fields may use other formats (the payload example shows an array list).

A future MSC may add a versioned endpoint that allows specifying more detailed metadata about each field, but hopefully this one helps keep things consistent and reliable while also working with the profile field framework we already have!

allowing for a more versatile user profile.

This proposal is designed to complement rather than replace
[MSC1769](https://github.com/matrix-org/matrix-spec-proposals/pull/1769) (Extensible profiles as
rooms). While [MSC1769](https://github.com/matrix-org/matrix-spec-proposals/pull/1769) offers a more
complex solution for extensible profiles, this proposal focuses on enabling the storage of small,
arbitrary key:value pairs at the global level.

This proposal does not seek to enforce the content or usage of these fields but rather to add a
framework for users to have extra data that can be further clarified and extended in the future as
community usage of these fields grows.

Homeservers could disable the ability for users to update these fields, or require a specific list
of fields, but the intention of this proposal is that users will be presented with a form to enter
their own free-text fields and values. After using these very flexible fields, the community may
opt to request a further extension to promote one or more fields to be specified per-room via
member events.

### Client-Server API Changes

> 1. **GET `/_matrix/client/v3/profile/{userId}/{key_name}`**: This endpoint will replace the existing

Reviewer 12: It might be worth noting somewhere that this is not a breaking change or deprecation but rather a generalization of the existing endpoints which will continue to exist.

Author: Does this clarification at the top help? https://github.com/tcpipuk/matrix-spec-proposals/blob/555388f2f72b815ab626595b1b3c0aef6775d898/proposals/4133-extended-profiles.md?plain=1#L14-L16

Reviewer 12: Yes, sounds good, thank you. 👍

      "displayname": "Alice Wonderland"
    }
    ```

3. **DELETE `/_matrix/client/v3/profile/{userId}/{key_name}`**: This endpoint will remove the key
   (and associated value) from the profile, if permitted by the homeserver. Could be considered a
   partial alternative to [MSC3754](https://github.com/matrix-org/matrix-spec-proposals/pull/3754)
   which specifies `DELETE` endpoints for specifically `/avatar_url` and `/displayname`.

4. **GET `/_matrix/client/v3/profile/{userId}/`**: This endpoint will retrieve all profile fields:

    ```json
    {
      "avatar_url": "mxc://matrix.org/MyC00lAvatar",
      "displayname": "John Doe",
      "custom_field1": "value1",
      "custom_field2": ["value2", "value3"]
    }
    ```

5. **POST `/_matrix/client/v3/profile/{userId}`**: This endpoint will accept a complete JSON object
   to replace the entire profile:

    ```json
    {
      "avatar_url": "mxc://matrix.org/MyNewAvatar",
      "displayname": "John Doe",
      "custom_field1": "new_value1",
      "custom_field2": ["new_value2", "new_value3"]
    }
    ```

### Server-Server API Changes

**GET `/_matrix/federation/v1/query/profile`** will mirror the client-server API changes

> to ensure profile information is consistently available across the federated network.

Reviewer 12: I might be overthinking this but I think it could help to spell out the (probably few) concrete changes to the SS API here. For instance, the `field` parameter is currently [documented](https://spec.matrix.org/v1.11/server-server-api/#get_matrixfederationv1queryprofile) as

> One of: `[displayname, avatar_url]`.

which I suspect will be lifted?

Author: Does the existing line you've highlighted not clarify enough that the field requirements will mirror the CS ones, or does it need to explicitly announce that the `field` parameter can accept more than two values?

Reviewer 12: I think it's implicitly clear, yes. I'm just trying to uncover any ambiguity before the spec writing stage. But maybe there is none here.

The current spec currently has some wording around fields that "can be made public" in the response

> If no `field` was specified, the response should include the fields of the user’s profile that can be made public, such as the display name and avatar.

I'm unsure whether or not that will change with this MSC but that's technically covered in the [comment below](https://github.com/matrix-org/matrix-spec-proposals/pull/4133/files#r1682494213).

Author: Oh, I see what you mean. Hmm.

My reading of that part of the spec makes me think it's distinguishing profile fields from account data, so that endpoint shouldn't return any non-public data?

I'll have a think about tweaking the wording to make this 100% clear just in case!

Author: How does this read to you? https://github.com/tcpipuk/matrix-spec-proposals/commit/72fcd90934b591c8c5ab4746231936b6111c89dd

Reviewer 12: Sounds good, thank you. 👍

As per the current stable endpoint, it accepts an optional `field` query string parameter to
request a single field.

As there is no method to verify the history of these fields over federation, this endpoint must
only accept requests for local users on the current homeserver, and homeservers must only request
a profile from the homeserver specified in that user's MXID.

### Capabilities

A new capability `m.set_profile_fields` will be introduced to control the ability to set custom

> profile fields. The client should assume setting fields is allowed when this capability is missing.

Reviewer 12: Maybe something like `m.set_custom_profile_fields` could help emphasize that this is orthogonal to the `m.set_displayname` and `m.set_avatar_url` capabilities?

Author: That makes sense, though I'd anticipated this would advertise everything added by this MSC being available, with the assumption that at some point the old capabilities might be retired.

I do think in the future we'll need a good way to mark which fields are editable or not, but held that back for a future MSC to define as I'm far from an expert on handling this type of data!

Do you think I should add both capabilities, i.e. a required one to advertise that we support extra fields at all, then an optional one to specify whether the user is allowed to set custom fields?

Author: Oops, I read the other comment first! Let's resolve this thread as the two have ended up being about the same thing 🙂

in advance - those would be defined in that field's MSC.

In contrast, if users wish to display their timezone immediately after this specification change,
they could enter a "My Timezone" field manually that would create a `u.My Timezone` key in their
profile - this would have open entry requirements, so they might enter `CEST` or `+2` or `None`
and clients/servers would not be expected to do anything with the value except store/display it.

### Size Limit

The name of a field must not exceed 255 bytes.

To ensure efficient handling and storage of profile data, the entire user profile JSON object
cannot exceed 64KiB. This follows the same size limit as events. Homeservers are allowed to limit
the fields (or content) that their local users can set. However, the only limit they should impose
on remote users is that the entire profile JSON block should not be larger than 64KiB.

The content of a field can be any valid JSON type, as long as the total size of the user profile
does not exceed 64KiB.

### Implementation Details

- This feature will be implemented as optional but recommended, enabling a smooth transition and

>   minimal disruption to existing deployments.

Reviewer 12: Does this refer to clients only? I think for servers, implementing it would be a MUST?

Author: I see what you're saying, and am willing to make the change, it's just the aim is to avoid incompatibility when a client/server is updated without the other.

There's already nothing to stop a server adding extra fields to the profile, there's just no API endpoints available to add/edit/delete them yet, so the short-term goal is for requests to read/update the `avatar_url` and `displaynane` fields to look/act exactly the same on both sides, but to transparently allow more fields later when the server advertises the capability.

As this MSC only proposes a framework to allow other people to add spec'd fields later, I thought it might be safer to stipulate that this shouldn't yet be _required_ for a server to be a valid homeserver, but that a future MSC could update this to say "Now we're pushing global adoption of timezones so need every client/homeserver to support this endpoint and recognise this field" for example.

Reviewer 12: Ah, so a server could declare it's not supporting the new endpoints by setting the capability to `false`? 

It might make sense to rename the capability then. It currently refers to "set" operations but if I'm understanding correctly the new `GET` endpoints could also be missing when the value is `false`.

Overall this sounds sensible though. 👍

Author: Good point.

I'd have thought GETing fields would be simpler because both clients and servers can just GET a user's entire profile and (in the case of [MSC4170](https://github.com/matrix-org/matrix-spec-proposals/pull/4170)) if the server doesn't return a 403 then the fields you get are the ones you know you're allowed to see.

I agree a future MSC for a specific field (e.g. let's say the presence status message gets moved to a profile field) might want that field to be checked regularly so would need a way to confirm whether that particular field can be read, but I'm hoping to leave the problem for that MSC to solve because I'm not even sure if it'll be required yet?

To my eyes, the main part this MSC needs to answer today is that clients/servers don't know if each other supports reading/writing extra fields individually without a capability, so perhaps I should simply make the capability `m.profile_fields` so servers can confirm they support this behaviour? If the capability is missing then clients know not to try using any extra field endpoints, but if it's set to _false_ then clients know the server recognises custom fields but not writing any of this user's fields?

Reviewer 12: > To my eyes, the main part this MSC needs to answer today is that clients/servers don't know if each other supports reading/writing extra fields individually without a capability, so perhaps I should simply make the capability m.profile_fields so servers can confirm they support this behaviour?

That sounds reasonable to me. 👍

> If the capability is missing then clients know not to try using any extra field endpoints, but if it's set to false then clients know the server recognises custom fields but not writing any of this user's fields?

That could work but I don't personally feel strongly about differentiating those two cases. As a client, knowing whether or not I can use the new endpoints seems most important to me.

Author: I've rewritten the capability section - does this look ok? https://github.com/tcpipuk/matrix-spec-proposals/commit/55e9e83a498684edadb13675ff9ecb819b1600da

Reviewer 12: Seems ok to me, thanks. 👍

- The profile data will be public by default, and compliance with GDPR and other privacy

>   regulations will be enforced, particularly in terms of data deletion and retention policies.

Reviewer 12: The wording "by default" makes it sound as if there is a mechanism to conceal things in certain situations?

Author: I should probably remove this, but was worried about forbidding making a field local-only, in case there's a possible useful implementation for corporate servers where other users on the same server can see a field that federated servers cannot, for example.

What do you think?

Reviewer 12: I think hiding certain fields on the SS API but not on the CS API is definitely a use case, yes.

I was just curious if you had foreseen a mechanism to achieve that as a user or if it would be a server admin's decision.

The term "public" is also slightly ambiguous here because [`GET /_matrix/client/v3/profile/{userId}`](https://spec.matrix.org/v1.11/client-server-api/#get_matrixclientv3profileuserid) doesn't currently require authorization and I suppose this will be true for the new `GET` endpoints, too.

FWIW, I think allowing the server admin to configure which fields to return on either of the APIs would fine within the scope of this proposal. Just wasn't sure if that was your intention.

Author: Ah, I see [MSC4170](https://github.com/matrix-org/matrix-spec-proposals/pull/4170) now exists and that I may have foreseen part of what you were looking for!

My goal in this MSC is to introduce a standard for these profile fields but not dictate how people might want to extend this in the future.

In this way, I'm adding/redefining endpoints to unlock the editing of fields that homeservers have always (technically) been allowed to publish, and introducing namespaces to make it easier to define future fields, but also trying to avoid intentionally limiting future implementations.

I can update this MSC to clarify it's compatible with [MSC4170](https://github.com/matrix-org/matrix-spec-proposals/pull/4170), but would you prefer I add anything in particular to help here? 🙂

Reviewer 12: Ok, those intentions sound good to me then. 👍

An optional cross-link to MSC4170 might help but I don't feel strongly.

Author: I've written up error codes, and referenced your MSC in the process - hopefully this looks ok? https://github.com/tcpipuk/matrix-spec-proposals/commit/f20b1a600d00923b98e746af09be8db3590b4b80

Reviewer 12: Perfect, thanks. 👍

  is defined the display name of this field.

For example, if a future MSC were to add a field for the user's timezone (not included in this MSC)
it might become `m.timezone` after entering the spec, but during the unstable process it might be
`org.matrix.msc9876.timezone`.

The MSC could define format requirements (e.g. "must be in IANA format like `Europe/Paris`") that
clients must validate, or that the field should be displayed in profiles showing the local time for
that user, or even that client/server behaviour is affected by this field (e.g. overriding display
from the device timezone to this configured one) but this MSC does not seek to define those options
in advance - those would be defined in that field's MSC.

In contrast, if users wish to display their timezone immediately after this specification change,
they could enter a "My Timezone" field manually that would create a `u.My Timezone` key in their
profile - this would have open entry requirements, so they might enter `CEST` or `+2` or `None`
and clients/servers would not be expected to do anything with the value except store/display it.

### Size Limit

> The name of a field must not exceed 255 bytes.

Reviewer 9: What's the error status & errcode that gets returned for this?

Reviewer 9: > I'm not sure what is common elsewhere in the spec - I'll have a look around now, but what do you think?

I think we generally use a 400 and then one of the `M_INVALID_PARAM` or whatever errcodes, but I'm not against using more HTTP status code.

Author: I've had a go at this - how does this look? https://github.com/tcpipuk/matrix-spec-proposals/commit/55e9e83a498684edadb13675ff9ecb819b1600da

To ensure efficient handling and storage of profile data, the entire user profile JSON object
cannot exceed 64KiB. This follows the same size limit as events. Homeservers are allowed to limit

> the fields (or content) that their local users can set. However, the only limit they should impose

Reviewer 9: Allowing servers to choose this means that clients can't offer local validation which is unfortunate.

Author: Allowing servers to allowlist/denylist which keys a user can update would stop local validation?

Or is it just that I haven't been clear enough on what homeservers are allowed to limit and should clarify that? 😄

Reviewer 9: > Allowing servers to allowlist/denylist which keys a user can update would stop local validation?

No -- having a generic length that's defined by homeservers and not the spec.

Author: Oh I see what you mean... perhaps I should _suggest_ a default (or minimum) limit of 64KiB?

My main concern is about people slapping 10MiB of base64 URIs in their profile and then other homeservers having to interpret/cache that... unless we allow servers to reject a remote user's profile if it's above a size they'll accept? 🤔

Author: I've done that now - how does this look? https://github.com/tcpipuk/matrix-spec-proposals/commit/5fdd69cae3884421b190d2fba804e5cc8ac5c41f

    }
    ```

    For example, setting `/_matrix/client/v3/profile/@alice:matrix.org/displayname` with:

    ```json
    {
      "displayname": "Alice Wonderland"
    }
    ```

3. **DELETE `/_matrix/client/v3/profile/{userId}/{key_name}`**: This endpoint will remove the key
   (and associated value) from the profile, if permitted by the homeserver. Could be considered a
   partial alternative to [MSC3754](https://github.com/matrix-org/matrix-spec-proposals/pull/3754)
   which specifies `DELETE` endpoints for specifically `/avatar_url` and `/displayname`.

4. **GET `/_matrix/client/v3/profile/{userId}`**: This endpoint will retrieve all profile fields:

    ```json
    {
      "avatar_url": "mxc://matrix.org/MyC00lAvatar",
      "displayname": "John Doe",
      "u.Custom Field": "value1",
      "m.allowed_list": ["value2", "value3"]
    }
    ```

5. **PUT `/_matrix/client/v3/profile/{userId}`**: This endpoint will accept a complete JSON object
   to replace the entire profile:

    ```json
    {
      "avatar_url": "mxc://matrix.org/MyNewAvatar",
      "displayname": "John Doe",
      "u.Custom Field": "new_value1",
      "m.allowed_list": ["new_value2", "new_value3"]
    }

>     ```

Reviewer 14: What is the use case for replacing the entire profile object? I fear this may lead to data loss if while one client tries to replace the whole profile, another client adds/updates a field which the first client isn't aware of.

It may make more sense to make this endpoint append/overwrite existing fields. That would allow clients to update multiple fields (avatar, displayname, bio) all in one request, without needing to specify all existing fields or worry about erasing a field that it doesn't know about.

Author: I agree this endpoint would be subject to race conditions, but my presumed use-case was for single-client accounts (e.g. bots, bridges, etc) to update the entire profile in a single request.

Bridges in particular might want to update hundreds (or thousands) of profiles, so it seemed important to offer an option to allow them to add *and remove* fields en masse in a single request.

As you've realised, I'd hope any user-interactive client would choose the "single-field at a time" endpoints to avoid race conditions.

Reviewer 14: It would help then to add what you've said as recommendations to the MSC, as those are all good points!

Author: It occurred to me that we could just have `PUT` and `POST` methods to choose between these behaviours, so I've written that, explained the difference, and a note about how/why you'd use the `POST` one: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/36aa7d479a4967ede351a5d7f1c933bb1d151256

request a single field. At time of writing, the Matrix specification says:

> If no `field` was specified, the response should include the fields of the user’s profile that
> can be made public, such as the display name and avatar.

Given this wording, homeservers currently already have the flexibility to decide whether some
fields are published over federation, and this proposal continues to allow this behaviour.

### Capabilities

A new capability `m.profile_fields` will be introduced to control the ability to set custom
profile fields.

For backwards compatibility purposes, clients should assume these extended endpoints are not
supported when this capability is missing.

When the capability exists but is set to `false`, clients should expect to read/display extended
fields, but should expect the server to deny creating/updating any custom fields. When this is set
to `true` it should be possible to create/update custom fields, but individual updates may receive
a 400/403 from the homeserver following the error codes listed below.

Example capability object:

```json
{
  "capabilities": {
    "m.profile_fields": {
      "enabled": false
    }
  }
}
```

### Error Handling

> To ensure clear communication of issues, the following error codes and messages will be used:

Reviewer 14: Do these error codes apply to the federation endpoint as well? Currently whether it does or not is not clear.

Author: Yeah, it was intentional to not specifically say they were only for the CS API.

Some actions (e.g. creating/editing a key) won't be occurring over federation, but if the server uses [MSC4170](https://github.com/matrix-org/matrix-spec-proposals/pull/4170) then 403 responses might be required, so I've tried not to be over-prescriptive on what a homeserver is _required_ to implement because there are multiple MSCs that touch on complementary behaviour to this one.

trigger state events in rooms. They will exist solely at the global level and are intended for
storing metadata about the user that does not need to be replicated in each room.

- **avatar_url** and **displayname**: Changes to these fields will generate state events in all
  rooms the user is a member of.
- **Custom fields**: These are stored in the user's global profile and do not generate state events
  in rooms.

### Key/Namespace Requirements for Custom Fields

Homeservers are not expected to enforce these namespaces, as future expansions may be unknown to
the server, but clients are expected to use the correct namespace for field creation/updates.

The namespace for field names is defined as follows:

- The namespace `m.*` is reserved for fields defined in the Matrix specification. This field may
  have special entry/display requirements that are defined in the Matrix specification. If a client
  does not recognise a field in this namespace, it may attempt to display it, but should not
  attempt to update the content in case it has special requirements.
- The namespace `u.*` is reserved for user-defined fields. The portion of the string after the `u.`

>   is defined the display name of this field. These user-defined fields will always be string format.

Reviewer 14: > The portion of the string after the `u.` is defined the display name of this field.

I'm not sure what this means. Does it mean that if there's a field `u.favourite_restaurant` then a client should display that field as `favourite_restaurant`?

If I were writing a mainstream client, I don't think I'd want to show unknown fields unless the user had enabled developer mode or similar; which reduces their usefulness dramatically.

Given that, I wonder whether `u.*` fields are even necessary. Is their purpose to eventually be promoted to specified fields? If so, why have them when you already have both stable fields (`m.timezone`) and unstable fields (`com.example.favourite_restaurant`) at your disposal?

I ask, as we've never needed a prefix other than `m.*` in the spec before.

Author: The `m.timezone` field doesn't exist yet, it is an example of a possible use for the `m.*` namespace, as explained in the document:

```md
For example, if a future MSC were to add a field for the user's timezone (not included in this MSC)
it might become `m.timezone` after entering the spec, but during the unstable process it might be
`org.matrix.msc9876.timezone`.

The MSC could define format requirements (e.g. "must be in IANA format like `Europe/Paris`") that
clients must validate, or that the field should be displayed in profiles showing the local time for
that user, or even that client/server behaviour is affected by this field (e.g. overriding display
from the device timezone to this configured one) but this MSC does not seek to define those options
in advance - those would be defined in that field's MSC.
```

This MSC defines endpoints to *allow* clients to create/edit/delete custom user-defined fields, and establishes a framework for future MSCs to define spec-defined fields that have more strictly defined content/validation.

As part of this MSC, homeservers are *allowed* to refuse the creation/editing/deletion of custom fields (whether allowlist/denylisting keys, or performing some other validation) but the MSC does not seek to demand strict implementations of a feature that will clearly grow to be used in different ways than I could possibly foresee.

The immediate functionality this MSC provides is the *ability* of endpoints to allow the creating/editing/deleting of profile fields that there is currently no way to edit:

```md
In contrast, if users wish to display their timezone immediately after this specification change,
they could enter a "My Timezone" field manually that would create a `u.My Timezone` key in their
profile - this would have open entry requirements, so they might enter `CEST` or `+2` or `None`
and clients/servers would not be expected to do anything with the value except store/display it.
```

Reviewer 14: Sorry, I didn't mean to imply that `m.timezone` already existed - I was using it as an example of the type of fields that could be used: those with stable (`m.*`) and unstable (i.e. `org.example.*`) prefixes.

However, I had missed the `u.My Timezone` example, which shows that the intention is for these fields to be shown in the UI as written (stripping the `u.`). I had concerns about `u.my_timezone`, but I can see `u.My Timezone` - resolving to `My Timezone` - in the UI working out!

I now see the value in having all three "types" of fields:

1. Those that have been spec'd and are given special meaning, i.e. `m.bio`.
2. Those that have unstable prefixes as part of the MSC process, i.e. `org.matrix.msc1234.bio`.
3. Those that are user-defined and not (yet) spec'd, i.e. `u.Favourite Spider`.

> The `u.*` namespace was lifted almost exactly from the definitions for room tagging: https://spec.matrix.org/v1.11/client-server-api/#room-tagging

Whoops, shows how often I look at the room tagging spec. Thanks!

Author: @anoadragon453 would you mind marking this review as resolved please? 🙂

Reviewer 14: Sure? I've dismissed it.

Author: Thanks! It's cleared the flag 🎉

in advance - those would be defined in that field's MSC.

In contrast, if users wish to display their timezone immediately after this specification change,
they could enter a "My Timezone" field manually that would create a `u.My Timezone` key in their
profile - this would have open entry requirements, so they might enter `CEST` or `+2` or `None`
and clients/servers would not be expected to do anything with the value except store/display it.

### Size Limit

The name of a field must not exceed 255 bytes.

To ensure efficient handling and storage of profile data, this specification proposal *recommends*
that the entire user profile JSON object not exceed 64KiB. However, homeservers are encouraged to
limit the fields (or content) that their local users can set, setting an upper size limit per field
and/or for the entire profile. Homeservers may also limit the maximum number of keys a user may set.

Likewise, homeservers may set an upper limit on the size of remote profiles they will accept from
remote homeservers over federation. It is recommended this minimum be set to *at least* 64KiB for

> the entire profile JSON block.

Reviewer 14: What happens if you set a large profile which some homeservers accepts, and others do not? Your profile will never update again on the latter until the size is reduced.

Recently we had a failure mode on matrix.org where a homeserver was trying to send a batch of presence EDUs that were larger than what matrix.org would accept. This caused the remote homeserver to continue retrying to send the batch after new presence events were generated, hammering matrix.org with requests that kept failing.

To alleviate this, we bounded the size of presence EDUs on the sending side in https://github.com/element-hq/synapse/pull/17371#issuecomment-2197107561. However, you could still run into the same issue if a remote homeserver defined their EDU limit lower than what Synapse now sends out. This implementation mismatch came from a lack of a consistently-defined request size limit.

Mandating a max size limit for requests in the spec that all homeservers must adhere to would avoid this.

Author: Until the most recent commit, all profiles had a max limit of 64KiB, but it was requested this become a recommendation rather than a requirement.

The main reason for this is that, in order to check the size of someone's profile, the server would need to pull all keys in the user's profile, generate a JSON blob, and measure the size of it.

Some implementations may then prefer to limit the max quantity/size of keys instead of an overall profile size, or other implementations (e.g. for local-only keys) may want to exclude some keys from public limits, so it would be safer to not be aggressively prescriptive here.

I've recommended a minimum amount of JSON to accept over federation, but wouldn't want to force huge homeservers like Matrix.org to accept 1MiB of JSON per user for a theoretically unlimited number of remote users, especially if profile data is going to be cached for a period of time.

I agree future MSCs may want to add mechanisms for servers to negotiate profile limits, but this is a simplified addition to get the framework in place so the community can get wider implementations in place and discover (and debate) what works best for these situations.

Reviewer 14: > The main reason for this is that, in order to check the size of someone's profile, the server would need to pull all keys in the user's profile, generate a JSON blob, and measure the size of it.

I'm not sure I buy this argument. You could make calculating the size more efficient in a number of ways; a naive method would be storing the profile size itself in the DB for each user, and updating that value when the user's profile is modified.

Unless I've missed something, it looks like the conversation that lead to the minimum recommendation was [this one](https://github.com/matrix-org/matrix-spec-proposals/pull/4133#discussion_r1691668918), which doesn't mention resource usage of calculating the total. Rather, it mentions that not spec'ing a max size prevents clients from being able to communicate that limit to the user before attempting to set the data. The current proposal still does not solve this.

The problem with setting a minimum is that some homeservers that you talk to will choose that minimum, effectively making it a maximum. Thus I don't think the current text actually changes anything, other than adding ambiguity as to what size limit a given homeserver may support.

I'd recommend defining a size limit, of which 64KiB seems reasonable for profile information.

> I agree future MSCs may want to add mechanisms for servers to negotiate profile limits, but this is a simplified addition to get the framework in place so the community can get wider implementations in place and discover (and debate) what works best for these situations.

I don't think adding negotiation would make sense I'm afraid. When talking to other homeservers, your homeserver would have to strip some fields but not others, which sounds complicated. On the client side, negotiation could work for local-only profile fields. But you'd hit up against federation size limits for any non local-only fields.

Author: I appreciate that view, yeah.

I've just reworked this to say that there should be a 64KiB limit *over federation* which can then potentially be bypassed locally by a homeserver according to its own policies: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/c3b2beb71e3fc53188c8b558e36b6faee1fb5469

Does that seem reasonable?

Reviewer 14: Hm, that still doesn't provide an ability for clients to know the size limit locally.

An ability to negotiate the size limit for remote-available fields could solve this. However, we don't actually have a way for the client to tell the homeserver which fields it wants to make available over federation, and which it does not yet. This makes it difficult to design on top of that idea.

Do you have a use case in mind for very large profile fields for use by non-federating homeservers? I can't think of any given what I've been asked for by Element customers over the years. Typically companies just want small bits of data to be stored, such as:

* Job title
* Holiday status
* Which "department" they're in
* Bio (2000 chars. max)

All of which combined won't come close to a ~65536 character limit.

Author: The expectation currently is not that users can choose sharing settings (room profiles will cover this later) but that we want homeserver implementations to have the option to limit what is federated outbound.

For example, your timezone or job title may be set automatically from your organisation's LDAP/OIDC server, and it would be desirable for your coworkers to see them, but your organisation might not want these publicly available to every server on the planet. This MSC *allows* a homeserver to choose what it shares over federation, so it *could* choose to denylist some fields from being federated, or allowlist a set list of fields to be federated.

Regarding sizes, I agree that 64KiB should be loads, I'm just wary of someone finding (for example) an amazing use case for a local-only field that requires a 100KiB data URI inside the profile fields - I don't personally see that specific application, just also can't see a good reason not to accommodate this?

I agree the ability for clients to see what fields they can/can't write (or the size limits) is difficult in the current iteration, but there are a number of improvements (discussed above) that I think would make for a great second MSC using a new versioned endpoint later.

What we want to do here is gently extend the existing endpoints to add the ability to add/edit profile fields with minimal client/server changes, so people can start developing wild ideas (and log little MSCs for specific fields like `m.timezone`) while we come back in a second MSC and debate what the perfect kitchen-sink endpoints will look like - I'd love the ability to set metadata on the fields, I just think it'd be biting off too much when profile fields in the current state could be live on real clients and servers this year.

Reviewer 14: Let's forget about having clients know the size locally for now then. Given that one of the examples I gave above mentioned a per-field size limit (a 2000 char bio), it may make sense to negotiate a per-field size for clients in the future.

I do think that a set size over federation is still necessary, and the proposal currently specifies that.

So I'm happy with where we are now. Thanks for going back and forth with me on it!

Author: Thanks a lot - your perspective on this has been really useful!

I totally agree the *ideal* implementation would need plenty of metadata on each field, but also trying to be pragmatic about phasing these changes - after this MSC, we may find room profiles suddenly spring to life and the extra metadata isn't quite as important, but equally someone may decide a little metadata would negate the need for room profiles.

I don't think we can truly know how big the appetite will be until people can get their hands on it 🙂

compatibility with existing servers. A future proposal may wish to use an EDU (such as Presence) to
notify users and homeservers that these custom fields have been updated.

## Alternatives

An alternative could be to allow for more specific endpoint modifications or to introduce a
completely new API specifically for extended information. However, these approaches could lead
to more significant changes to the existing API and higher complexity for client developers.

## Security Considerations

Since the profile data is public, there are minimal security risks associated with the transmission
of sensitive information; however, it is critical to ensure that all data handled through these
endpoints complies with GDPR and other privacy regulations, particularly in the event of user data
deletion.

It is crucial to ensure that clients inform users this data will be public, does not include
sensitive personal information, and complies with legal frameworks such as GDPR. Homeservers will
be encouraged to implement data caching strategies that do not exceed 24 hours to minimise the risk
of unintended data persistence.

> ## Unstable prefix

Author: Egads. Do I just abandon the merge option for now, and make a smaller MSC dependent of these two?

> ## Unstable prefix

Reviewer 9: > Egads. Do I just abandon the merge option for now, and make a smaller MSC dependent of these two?

It certainly benefits speed to reduce dependencies between MSCs. It *might* make sense to move the changes to set data via the `/profile` endpoint (except for fetching the custom fields) as it could become a sticking point w/ race conditions etc. That could be hashed out in a separate MSC which has the `POST`/`PATCH` calls in it? 🤷 

(The same could maybe be said for `DELETE`, but that's much easier to implement.)

> ## Unstable prefix

Reviewer 11: PUT is already in the spec

4138 is one tick away from FCP though, so it might be better to just keep all methods here :thinking:

> ## Unstable prefix

Reviewer 11: MSC4138 has entered FCP 🎉

> ## Unstable prefix

Author: Thank you! Reverted my last commit 😄

The [current Matrix specification](https://spec.matrix.org/v1.10/#profiles) technically already
allows extra custom fields to be published in a user's profile, however as this proposal introduces
additional requirements and allows custom user-defined fields, an unstable prefix should be used on
these fields until this proposal has entered the API as stable:

```json
{
    "avatar_url": "mxc://matrix.org/MyC00lAvatar",
    "displayname": "John Doe",
    "uk.tcpip.msc4133.u.Custom Field": "field_value",
    "uk.tcpip.msc4133.m.allowed_list": ["one value", "another value"]
}
```

> `/_matrix/client/uk.tcpip.msc4133/profile/{userId}` would be necessary for the `PATCH` and `PUT`

Reviewer 11: ```suggestion
`/_matrix/client/unstable/uk.tcpip.msc4133/profile/{userId}` would be necessary for the `PATCH` and `PUT`
```

Author: Thanks!

   specified `key_name`:

    ```json
    {
      "key_name": "new_value"
    }
    ```

    For example, setting `/_matrix/client/v3/profile/@alice:matrix.org/displayname` with:

    ```json
    {
      "displayname": "Alice Wonderland"
    }
    ```

3. **DELETE `/_matrix/client/v3/profile/{userId}/{key_name}`**: This endpoint will remove the key
   (and associated value) from the profile, if permitted by the homeserver. Could be considered a
   partial alternative to [MSC3754](https://github.com/matrix-org/matrix-spec-proposals/pull/3754)
   which specifies `DELETE` endpoints for specifically `/avatar_url` and `/displayname`.

4. **GET `/_matrix/client/v3/profile/{userId}`**: This endpoint will retrieve all profile fields:

    ```json
    {
      "avatar_url": "mxc://matrix.org/MyC00lAvatar",
      "displayname": "John Doe",
      "u.Custom Field": "value1",
      "m.allowed_list": ["value2", "value3"]
    }
    ```

5. **PUT `/_matrix/client/v3/profile/{userId}`**: This endpoint will accept a complete JSON object
   to *merge* into the current profile, updating any changed keys without removing/changing any

>    absent ones:

Reviewer 9: Isn't this usually what `PATCH` is used for?

Author: Fair point! In $dayjob we typically use `PUT` for this, but I'm very happy to confirm to this standard if the spec uses `PATCH` which I've now put here: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/267d663c8ff73b587c10c0790e05f06c0a62c225 🙂

    ```json
    {
      "avatar_url": "mxc://matrix.org/MyNewAvatar",
      "displayname": "John Doe",
      "u.Custom Field": "new_value1",

>       "m.allowed_list": ["new_value2", "new_value3"]

Author: Fair point! Is this clear enough? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/6038e95864e8e288f1239f0b025fd235e068a317#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R246-R253

When the capability exists but is set to `false`, clients should expect to read/display extended
fields, but should expect the server to deny creating/updating any custom fields. When this is set
to `true` it should be possible to create/update custom fields, but individual updates may receive
a 400/403 from the homeserver following the error codes listed below.

Example capability object:

```json
{
  "capabilities": {
    "m.profile_fields": {
      "enabled": false
    }
  }
}
```

### Error Handling

To ensure clear communication of issues, the following error codes and messages will be used:

- **400 Bad Request**: When the request is malformed, exceeds specified limits, or the profile
  JSON object is larger than 64KiB.
  - **Error Code for Malformed Request**: `M_BAD_JSON`
  - **Example Response**:

    ```json
    {
        "errcode": "M_BAD_JSON",
        "error": "The provided JSON is malformed."
    }
    ```

  - **Error Code for Exceeding Size Limit**: `M_TOO_LARGE`
  - **Example Response**:

    ```json
    {
        "errcode": "M_TOO_LARGE",
        "error": "The profile data exceeds the maximum allowed size of 64KiB."
    }
    ```

>   - **Error Code for Key Name Exceeding Length**: `M_KEY_TOO_LONG`

Author: You make a good point.

I wanted to give clients a clever way to tell whether the key or the value was the problem, but in hindsight I think it's safer to just offer `M_INVALID_DATA` for now, then the server can always provide different `"error"` messages to clarify the exact problem if needed: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/267d663c8ff73b587c10c0790e05f06c0a62c225#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R186-R194

>   - **Error Code for Key Name Exceeding Length**: `M_KEY_TOO_LONG`

Reviewer 9: I think `M_INVALID_PARAM` is the one to use. I don't think `M_INVALID_DATA` exists.

>   - **Error Code for Key Name Exceeding Length**: `M_KEY_TOO_LONG`

Author: Does that mean the change in https://github.com/matrix-org/matrix-spec-proposals/commit/267d663c8ff73b587c10c0790e05f06c0a62c225 should be ok?

>   - **Error Code for Key Name Exceeding Length**: `M_KEY_TOO_LONG`

Reviewer 9: No, that uses `M_INVALID_DATA`

>   - **Error Code for Key Name Exceeding Length**: `M_KEY_TOO_LONG`

Author: Oh! I typoed originally and just keep reusing the typo 🤦‍♂️

I'll aim to fix this tomorrow...

>   - **Error Code for Key Name Exceeding Length**: `M_KEY_TOO_LONG`

Author: Hopefully this looks ok? https://github.com/matrix-org/matrix-spec-proposals/commit/08099c0f3142b0b55a220bfb435b093a4a2eb624#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439L186-L194

  - **Example Response**:

    ```json
    {
        "errcode": "M_KEY_TOO_LONG",
        "error": "The key name exceeds the maximum allowed length of 255 bytes."
    }
    ```

- **403 Forbidden**: When the user does not have permission to take this action on a specific key,
  such as when the server policy (e.g.
  [MSC4170](https://github.com/matrix-org/matrix-spec-proposals/pull/4170)) restricts such actions:
  - **Error Code**: `M_FORBIDDEN`
  - **Example Response**:

    ```json
    {
        "errcode": "M_FORBIDDEN",
        "error": "You do not have permission to modify this field."
    }
    ```

- **404 Not Found**: When attempting to `GET` or `DELETE` a profile key that does not exist:
  - **Error Code**: `M_NOT_FOUND`
  - **Example Response**:

    ```json
    {
        "errcode": "M_NOT_FOUND",
        "error": "The requested profile key does not exist."
    }
    ```

> ### Distinction Between Existing and Custom Fields

Reviewer 9: ```suggestion
### Propagation of profile fields to membership events
```

The existing fields, `avatar_url` and `displayname`, will continue to trigger state events in each

> room. These fields are replicated per-room via member events. Custom fields, however, will **not**

Reviewer 9: I think this whole section should say "other" not "custom"? It is marking that `avatar_url` and `displayname` have additional special rules.

```suggestion
room. These fields are replicated per-room via member events. Other fields, however, will **not**
```

- **Custom fields**: These are stored in the user's global profile and do not generate state events
  in rooms.

### Key/Namespace Requirements for Custom Fields

Homeservers are not expected to enforce these namespaces, as future expansions may be unknown to
the server, but clients are expected to use the correct namespace for field creation/updates.

The namespace for field names is defined as follows:

- The namespace `m.*` is reserved for fields defined in the Matrix specification. This field may
  have special entry/display requirements that are defined in the Matrix specification. If a client
  does not recognise a field in this namespace, it may attempt to display it, but should not
  attempt to update the content in case it has special requirements.
- The namespace `u.*` is reserved for user-defined fields. The portion of the string after the `u.`
  is defined the display name of this field. These user-defined fields will always be string format.

For example, if a future MSC were to add a field for the user's timezone (not included in this MSC)
it might become `m.timezone` after entering the spec, but during the unstable process it might be

> `org.matrix.msc9876.timezone`.

Reviewer 9: I wrote this up as [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175) to show a concrete usage.

Author: Thanks! I've adjusted my example to show this: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/6038e95864e8e288f1239f0b025fd235e068a317

> 
> For example, if a future MSC were to add a field for the user's timezone (not included in this MSC)
> it might become `m.timezone` after entering the spec, but during the unstable process it might be
> `org.matrix.msc9876.timezone`.
> 
> The MSC could define format requirements (e.g. "must be in IANA format like `Europe/Paris`") that
> clients must validate, or that the field should be displayed in profiles showing the local time for
> that user, or even that client/server behaviour is affected by this field (e.g. overriding display
> from the device timezone to this configured one) but this MSC does not seek to define those options
> in advance - those would be defined in that field's MSC.
> 
> In contrast, if users wish to display their timezone immediately after this specification change,
> they could enter a "My Timezone" field manually that would create a `u.My Timezone` key in their
> profile - this would have open entry requirements, so they might enter `CEST` or `+2` or `None`
> and clients/servers would not be expected to do anything with the value except store/display it.

Reviewer 9: ```suggestion
- Client-specific or unstable fields MUST use the Java package naming convention: `tld.name.*`.
```

Author: I kept this chunk in because there have been several questions about naming conventions (especially the namespaces) and whether this MSC includes any pre-baked `m.*` fields.

If I remove this, would you suggest I add any clarification anywhere else for this stuff?

Author: I've chopped this down to focus on the custom field string requirement and used the timezone MSC as an example: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/6038e95864e8e288f1239f0b025fd235e068a317

Then added the clarification regarding client-specific or unstable fields: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/306ded2a6933d6eb65cb596ad5759aab262a336e

### Size Limit

The key *must* be a string of *at least* one character, and *must* not exceed 255 bytes.

To ensure efficient handling and storage of profile data, this proposal requires the entire user
profile JSON object not exceed 64KiB *over federation*.

As it is possible for a homeserver to allow a user to edit a field that it then doesn't provide
over federation, homeservers are encouraged to limit the fields (or content) that their local users
can set, setting an upper size limit per field and/or for the entire profile. Homeservers may also

> limit the maximum number of keys a user may set.

Reviewer 9: I think we should remove the "over federation" requirement and just straight require this. I suspect at some point we're going to want to propagate/push profile changes to local users.

Author: My theory was that there's no real reason a profile couldn't be larger within a homeserver if that implementation allowed it.

If a server doesn't publish some fields over federation (or doesn't federate at all) then should we limit it because of federation requirements?

Reviewer 9: I think it should be limited *regardless* so event local clients know what to expect when fetching it.

Author: Fair. This is what I wanted originally, so I'm happy to put it back! 😄

How's this? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/fe813d3e6995b4c85036a1a816159a572999d9f1

completely new API specifically for extended information. However, these approaches could lead
to more significant changes to the existing API and higher complexity for client developers.

## Security Considerations

Since the profile data is public, there are minimal security risks associated with the transmission
of sensitive information; however, it is critical to ensure that all data handled through these
endpoints complies with GDPR and other privacy regulations, particularly in the event of user data
deletion.

It is crucial to ensure that clients inform users this data will be public, does not include
sensitive personal information, and complies with legal frameworks such as GDPR. Homeservers will
be encouraged to implement data caching strategies that do not exceed 24 hours to minimise the risk
of unintended data persistence.

## Unstable prefix

The [current Matrix specification](https://spec.matrix.org/v1.10/#profiles) technically already
allows extra custom fields to be published in a user's profile, however as this proposal introduces
additional requirements and allows custom user-defined fields, an unstable prefix should be used on

> these fields until this proposal has entered the API as stable:

Reviewer 9: It didn't jump out to me that this requires an unstable `_matrix/client/uk.tcpip.msc4133/profile/{userId}` as well due to the `on_PUT` and `on_POST` changes.

Author: Good shout, I've done that: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/6038e95864e8e288f1239f0b025fd235e068a317

      "enabled": true,
      "allowed": ["m.example_field", "org.example.job_title"],
      "disallowed": ["org.example.secret_field"]
    }
  }
}
```

### Behaviour

- **When capability is missing**: Clients SHOULD assume extended profiles are supported and that
  they can be created or modified. If a server intends to deny some (or all) changes, it SHOULD use
  the capability to advertise this, improving the client experience.

- **When `enabled` is `false`**: Clients SHOULD expect to display profiles but NOT create or update
  fields. Any attempt to do so SHOULD result in a `403 Forbidden` error. This does not affect
  `avatar_url` and `displayname` fields, which are allowed for compatibility purposes.

- **When `enabled` is `true`**: Clients SHOULD allow users to create or update fields, except those
  keys listed in the `disallowed` array. Servers MAY specify an `allowed` array to allowlist fields
  that users can set. If both `allowed` and `disallowed` keys are provided, the `disallowed` one
  should be ignored. Clients SHOULD receive `400 Bad Request` or `403 Forbidden` responses if
  server-side policies prevent them.

## Key and Namespace Requirements

> Profiles MUST be at most 64 KiB (65,536 bytes) in size, as measured in Canonical JSON, including

Reviewer: link to Canonical JSON spec please

Author: Done: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/21ad83f5dd94f9cff0cecc67f721f247073b472c

the `avatar_url` and `displayname` fields.

Homeservers SHOULD NOT enforce namespaces, as future expansions may be unknown to the server, but
clients SHOULD use the correct namespace when creating or updating fields.

> Keys MUST follow the [Common Namespaced Identifier Grammar](https://spec.matrix.org/unstable/appendices/#common-namespaced-identifier-grammar),

Reviewer: ```suggestion
Keys MUST follow the [Common Namespaced Identifier Grammar](https://spec.matrix.org/v1.13/appendices/#common-namespaced-identifier-grammar),
```

linking to unstable is volatile

Author: Committed with thanks 🙂

- **Namespace `tld.name.*`**: For client-specific or unstable fields, using Java package naming
  convention (e.g., `com.example.custom_field`).

Following this change, clients could use `m.example_field` if that field is defined by the Matrix
specification, or `org.example.job_title` for organisation, client-specific fields, or MSC-backed
unstable features. Proposal [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175)
demonstrates the process of defining new fields in the `m.*` namespace.

## Error Handling

### 400 Bad Request: Request Exceeds Limits or Is Malformed

- **`M_BAD_JSON`**: Malformed request.

```json
{
  "errcode": "M_BAD_JSON",
  "error": "The provided JSON is malformed."
}
```

- **`M_PROFILE_TOO_LARGE`**: Exceeds total profile size limits.

```json
{
  "errcode": "M_PROFILE_TOO_LARGE",
  "error": "The profile data exceeds the maximum allowed size of 64 KiB."
}
```

- **`M_KEY_TOO_LARGE`**: Exceeds individual key length limits.

```json
{
  "errcode": "M_KEY_TOO_LARGE",
  "error": "The key name exceeds the maximum allowed length of 255 bytes."
}
```

### 403 Forbidden: User Lacks Permission

> Unchanged from the [current spec](https://spec.matrix.org/v1.13/client-server-api/#server-behaviour).

Reviewer: The server will need an ability to deny setting/creating a value even if the capability allows it - if this error code can be used for that, please clarify in the MSC.

Author: Ah, gotcha. Is this the sort of thing you had in mind? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/a948f5d03a7e25fc6d6ddea6f855a5733b165282

Reviewer: (yes)

  have special entry/display requirements that are defined in the Matrix specification. If a client
  does not recognise a field in this namespace, it may attempt to display it, but should not
  attempt to update the content in case it has special requirements.
- The namespace `u.*` is reserved for user-defined fields. The portion of the string after the `u.`
  is defined the display name of this field. These user-defined fields will always be string format.
- Client-specific or unstable fields MUST use the Java package naming convention: `tld.name.*`.

Following support for this specification change, a user could enter a "My Timezone" field manually
in their client and the client would be expected to add a `u.My Timezone` key in their profile.
However, this would be expected to be a string value, even if the user types "null" or an integer
into the box.

In contrast, [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175) would use the
unstable key `us.cloke.msc4175.tz` and following approval would then support clients using the
`m.tz` key with the values/validation that MSC requires.

### Size Limit

> The key *must* be a string of *at least* one character, and *must* not exceed 255 bytes.

Reviewer 9: Both of this are `M_INVALID_PARAM`? (Personally I'd find it helpful if the errors were mixed into this section instead of a separate one, but don't feel the need to refactor.)

The mixture of characters and bytes here is kind of weird?

Author: I've been trying to find the best options to use from [existing error codes](https://spec.matrix.org/v1.11/client-server-api/#common-error-codes) but I guess could add new ones if needed?

The spec seems to reserve `M_LIMIT_EXCEEDED` for rate-limits exceeded, so I've avoided using that.

Currently `M_TOO_LARGE` is reserved for operations that cause the profile to be too big (I guess this would also be used if a homeserver limited individual value lengths too?)

I assumed `M_INVALID_PARAM` would be used for keys that don't match the required spec, but I'm open to suggestions on better ways to handle this!

I could also specify `M_RESOURCE_LIMIT_EXCEEDED` is to be used when a user is not allowed to create more than the current number of keys? 🤔

Author: I've had a go at improving this a bit in the meantime: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/3157982521a857dc76722ebb82600a0ef4103329

Reviewer 9: `M_RESOURCE_LIMIT_EXCEEDED` seems wrong -- that is implied to be from an external factor. I think we're making this more complicated than it needs to be somehow. :(

Author: Ah, I figured it could be used when the user's exceeded a configured limit (a size rather than rate limit in this case) but I'm happy to remove, just thought the previous suggestion was that we shouldn't be reusing the same error codes for different issues.

I assume we want to avoid defining a bunch of new error codes, so would love some guidance here! 😇

To ensure efficient handling and storage of profile data, this proposal requires the entire user
profile JSON object not exceed 64KiB.

A future MSC may increase this limit or add exceptions, but this current limit has been chosen to
allow both servers and clients to have predictable upper limits, especially for caching.

Homeservers are allowed to limit the fields (or content) that their local users can set, setting
an upper size limit per field and/or for the entire profile, and may also limit the maximum number

> of keys a user may set.

Reviewer 9: This is *really* annoying and potentially error prone to implement. I would recommend instead setting a maximum value size & a maximum number of keys which will guarantee staying below the limit. Maybe a max size of 1024 bytes and a maximum count of 50 values?

Note that I don't think there's a current limit of display names and avatars so this defacto applies to it now. We should probably point that out explicitly?

Author: I've added a clarification that homeserver implementations may choose how to ensure their profiles remain within the limit, giving the 50x1KiB as an example: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/c5a30157e624fc91148423f7b0b169ab64035e94

I've also added a notice to warn that if a homeserver enforces limits per-field, it also would need to apply them to `avatar_url` and `displayname`: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/c8a5a1a7d2da613addebf4b3b31b03a4a26b8c5a

Do those seem reasonable?

Reviewer 7: I would also suggest a maximum for cross compatibility. We've had issues before where being lax here means one server implementing a very high limit impacts users on a lower limit. 50 keys sounds reasonable. 

I would say that field values should still follow the max size of events, which has so far seemed to go down reasonably well in the ecosystem.

If we find this isn't enough, maybe a future MSC can tackle it. Extended Extended Profiles :)

Reviewer 7: As far as avatar_url and displayname goes...tricky. We can either specialcase them for compatibility (and implementations may choose to do this?) or be strict now and homeservers have to figure out how to trim or remove lenghy keys. To be honest, I'd find a hard rationale to exceed 1KB for either key, let alone 64KB.

Reviewer 9: Another thought I had -- limiting only the maximum size vs. a per-value size might be weird for fields that take a user-inputted value? It becomes hard to do proper client size validation and could be frustrating to make a good UI for.

Author: I'd hope most clients would implement a sane upper value for most fields, but also wary that if someone implements "bio" type fields then 1024 bytes might be tight (including multi-byte characters) and some implementations might include files or GPG keys or something.

I really like the idea of enforcing an overall max as it makes it very easy for servers to reject profiles early if the content length is way too long, e.g. if a server sends a `Content-Length` header of 200MiB then we can drop it before it causes trouble.

I can certainly lock it down if we want to say data should never exceed X per field, I'm just wary that as soon as I put some limit in then someone is likely to complain it breaks their Super Awesome Idea™️.

Reviewer 9: > I really like the idea of enforcing an overall max as it makes it very easy for servers to reject profiles early if the content length is way too long, e.g. if a server sends a `Content-Length` header of 200MiB then we can drop it before it causes trouble.

I think we're cross-talking here. I'm talking about the Client-Server API and you're talking about the Server-Server API, I think? I'm not suggesting getting rid of the overall size limit.

Reviewer 9: We talked a bit about this on Matrix: https://matrix.to/#/!ewdjhNcPcEmYNKzlWp:t2l.io/$uOOBzRmu4iPpFuAMCl4jeJQXfcKxRfvFSCIEGXe3zxA?via=matrix.org&via=element.io&via=envs.net

The tl;dr was:

* @tulir mentioned hungryserv stores the extended profile as a JSONB blob, while my current Synapse implementation stores them as individual rows (I originally did the JSONB blob and it was annoying).
* The `PATCH` endpoint might make it a bit more fiddly.
* The size of the display name / avatar URL needs to be accounted for manually.
* Some discussion of code points vs. characters vs. bytes.

Reviewer 7: Noting https://github.com/element-hq/synapse/pull/17488/files#r1698867572 that we currently have a 64KB limit for the *whole* profile, which will need adjusting if we change the rules here.

Author: I think we're always going to have a 64KiB limit on the whole profile, the question is just about how Synapse's implementation enforces that. Hungryserv's implementation puts all of the fields in a single JSON object, but the current Synapse implementation stores each field separately, which would make limiting "max X fields of max Y length" more efficient in their case.

Reviewer 9: @erikjohnston and I had quite a bit of back-and-forth about this in a DM that I'll try to distill:

1. The main point Erik made was that *if* we hope to *one day* send this data over federation than we need to ensure the 64 KiB limit is well defined. That probably implies that this needs to be serialized using canonical JSON?
2. You probably don't want huge fields (keys or values) anyway, just for sanity's sake. So limiting the individual key names and values makes sense.
3. You probably don't want huge fields (keys or values) for performance reasons (both at the database and application level).

We also discussed a bit about how Synapse should store the data, but that's somewhat unrelated here.

Author: That all makes sense.

I could set something like an upper limit of 255 bytes on key names?

My main concern with values is that there are so many possible uses for them. For example, I've recently become aware of someone who wants to use profile fields to store VR avatar information for an implementation of VR in Matrix, and it'd also be valuable to have decent size fields for things like "bio" fields.

We could limit custom keys, then MSC specific longer ones, but that would pose similar issues for storage, so it's definitely something worth thinking about now.

Reviewer 7: To be honest I'd argue that if you are in the business of essentially storing big json blobs or otherwise in the profile, the media repository might be a better place to go. I.e. You'd reference your VR profile as a MXC link.

That means that clients in the know can load and cache the data, and the ones that don't (and the server(s)) don't need to concern themselves with storing your JSON blob. I'd probably even state as such on the spec that the profile is for short snippets of information like displaynames, location, short bios, etc. Not big blobs of data.

Author: That's a good point!

I suppose it'd also make it easier to quickly switch between these VR profiles, as you could toggle the profile field entry between two previously stored MXCs.

Ok, that being the case, do I set the value limit to 1024 bytes or shorter? I had a look through the spec and couldn't see an upper limit on MXC length, but figure we'd want to accommodate a reasonable size here.

Reviewer 7: This is where I'd really love some clients to weight in but I'd be keen to say 1024 is a little too short, but maybe something in the region of 8KB is okay. My current thinking is maybe you'd have a 500 character bio in maybe plaintext&html, and accounting for bits of unicode it feels like a reasonable baseline.

50 keys at 8KB is a 400KB profile which feels like basically the maximum of acceptableness to me.

Author: GitHub's "Bio" field accepts 160 bytes (at least I believe it's bytes, I added an emoji in mine and it limited to 159 characters) I assume because it accepts social links in other fields, and you can post gists to represent longer profiles if needed.

LinkedIn's "About" field accepts up to 2,600 bytes (also tested this with emoji) but that seems quite extreme, and they don't have that gist-type option. I've got quite a decent summary in mine and it's using 540 bytes.

If we're excluding the use-case of large chunks of data being stored in the profile, and want them to be small and performant, we probably shouldn't go above a 64KiB total limit, especially as 8KiB per field is a huge chunk of text. If clients need to download that every time someone checks what timezone you're in, I imagine that's going to hurt performance.

I'm becoming more convinced that 1KiB is plenty, especially as it doesn't stop a later MSC for a special field that has special requirements later.

Author: (very much welcome more opinions, I just think that even if we MSC'd something like a "profile stylesheet" field, you'd still probably use media and link a CSS file rather than embed a huge chunk of markup in the profile itself?)

Reviewer 7: I don't have strong enough opinions on 1kb v.s. 8 v.s. 64, whatever gets this unblocked. If we discover usecases, we can always increase the limit in another MSC.

Author: After [an extended conversation in #matrix-spec](https://matrix.to/#/!NasysSDfxKxZBzJJoE:matrix.org/$TKeMSZbWFraxQMSlzwi6afrZCYe9W7lKHs2Y0ai7hBY?via=matrix.org&via=envs.net&via=element.io), I've written this update: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/a81f21f8eff4023b7acffd7442ee913caed4b6a9

The important change is this section:

```md
### Size Limits

Until another MSC specifies otherwise:

- Each profile *must* be *at most* 64KiB in size
- Each profile may contain *up to* 50 keys (including `avatar_url` and `displayname`)
- Each key *must* be a string of *at least* one byte, and *must* not exceed 128 bytes
- Each value *must* be a string and *must* not exceed 512 bytes

These sizes are measured *after* being escaped for JSON, e.g. `🛸` would become `\uD83D\uDEF8`.

This 64KiB size limit should be calculated on the full canonicalised JSON payload a homeserver
would receive when retrieving the full profile, so a `Content-Length` header on the HTTP response
must not exceed 65536 bytes.

Future MSCs may allow fields that are not string values, or add exceptions to the size limits,
but these current limits have been chosen to allow both servers and clients to have predictable
upper limits for performance and caching purposes.
```

Author: I misread and the max size of an event is 65536 bytes *as encoded in Canonical JSON* so I've updated the wording to use the same measurement here: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/e688eb137a7368c0ec71cac1566a22baa9204fdc

  convention (e.g., `com.example.custom_field`).

Following this change, clients could use `m.example_field` if that field is defined by the Matrix
specification, or `org.example.job_title` for organisation, client-specific fields, or MSC-backed
unstable features. Proposal [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175)
demonstrates the process of defining new fields in the `m.*` namespace.

## Error Handling

### 400 Bad Request: Request Exceeds Limits or Is Malformed

- **`M_BAD_JSON`**: Malformed request.

```json
{
  "errcode": "M_BAD_JSON",
  "error": "The provided JSON is malformed."
}
```

- **`M_PROFILE_TOO_LARGE`**: Exceeds total profile size limits.

```json
{
  "errcode": "M_PROFILE_TOO_LARGE",
  "error": "The profile data exceeds the maximum allowed size of 64 KiB."
}
```

- **`M_KEY_TOO_LARGE`**: Exceeds individual key length limits.

```json
{
  "errcode": "M_KEY_TOO_LARGE",
  "error": "The key name exceeds the maximum allowed length of 255 bytes."
}
```

- **`M_TOO_MANY_KEYS`**: Exceeds key limits.

```json
{
  "errcode": "M_TOO_MANY_KEYS",
  "error": "The user has exceeded the maximum number of allowed keys in their profile."
}

> ```

Reviewer 11: There doesn't seem to be a key limit anymore, so `M_TOO_MANY_KEYS` should at least be removed, and possibly `M_KEY_TOO_LARGE` too?

Author: Agreed regarding "too many" although my assumption was that homeservers would probably enforce limits and it'd be worth providing errors that clients can comprehend when that happens. I'm happy to remove if you disagree? 

Regarding keys too long, the limit is set by the common naming schema, so although I don't specify a limit anymore, there is a limit of 256 imposed by the spec.

Reviewer 11: Oh yes the length limit is indeed there. Servers probably shouldn't be enforcing extra limits on number of keys, but not completely sure about that 🤔

Author: That's fair, I'm happy to remove it if the consensus is that it doesn't have any value!

Author: Done!

profile management.

## Unstable Prefixes

### Unstable Profile Fields

Until this proposal is stable, fields SHOULD use an unstable prefix:

```json
{
  "avatar_url": "mxc://matrix.org/MyC00lAvatar",
  "displayname": "John Doe",
  "uk.tcpip.msc4133.m.example_field": "field_value"
}
```

### Unstable Endpoints

Use unstable endpoints when the capability is not yet stable:

- **Get/Set/Delete Profile Fields**:
  - `/_matrix/client/unstable/uk.tcpip.msc4133/profile/{userId}/{key_name}`

### Unstable Capability

Advertise the capability with an unstable prefix:

```json
{
  "capabilities": {
    "uk.tcpip.msc4133.profile_fields": {
      "enabled": true,
      "disallowed": ["org.example.secret_field"]
    }
  }
}
```

### Unstable Client Features

The client feature `uk.tcpip.msc4133` SHOULD be advertised on the `/_matrix/client/versions`
endpoint when the `PUT` and `PATCH` methods are accepted on the
`/_matrix/client/unstable/uk.tcpip.msc4133/profile/{userId}` endpoint.

Once this MSC is merged, the client feature `uk.tcpip.msc4133.stable` SHOULD be advertised when the
`PUT` and `PATCH` methods are accepted on the `/_matrix/client/v3/profile/{userId}` endpoint until

> the next spec version where this endpoint is officially written into the spec, e.g.:

Reviewer 11: PATCH endpoints don't exist anymore (the replace everything PUT would've been nice, but maybe it's better to leave for another MSC)

Author: Agreed, I'll remove this. 

The `PUT` was intentionally left out because it was just causing confusion - adding both `PATCH` and `PUT` in a simple MSC afterwards makes sense, I just haven't raised it yet 🙂

Author: Done!

Reviewer 11: I think that section is still necessary to have the unstable and stable version flags, just the endpoint was wrong because only GET/PUT/DELETE `/profile/{userId}/{key_name}` exists now

Author: Oh, I thought that particular flag was only required for the `PUT` and `PATCH` endpoints and everything else was covered by the capability?

Reviewer 11: They're different things. Capabilities are for allowed/disallowed, unstable features (and `/versions` in general) are for supported/unsupported.

Author: Oh, I thought `avatar_url` and `displayname` already supported `GET` and `DELETE` so theoretically the only change is that more keys are supported, which is signified by the capability?

This hasn't ever been mentioned before, hence I only had that section in for the entire profile `PUT` and `PATCH` methods.

Reviewer 11: DELETE wasn't supported and even GET/PUT are really new endpoints (as the field was hardcoded rather than being a parameter in the old endpoints), that's why it requires an unstable prefix too. The capabilities section also says omitting the capability should be assumed to mean it's supported, which is not how unstable features work.

The [capabilities spec](https://spec.matrix.org/v1.13/client-server-api/#capabilities-negotiation) also says:

> Some examples of what should **not** be a capability are:
> * Whether the server supports a feature in the unstable specification.

Author: Got it, I'm on my way to bed, but will write up a new unstable endpoint section in the morning that makes it clear it's for keys rather than just the profile.

Author: @tulir how does this read to you? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/9fa009607fcec1419454dc120464a4dc0e5abef9

Reviewer 11: Looks good

There is no method to verify the history of global profile fields over federation, so this endpoint
must only accept requests for local users on the current homeserver, and homeservers must only
request a profile from the homeserver specified in that user's MXID.

As per the current stable endpoint, it accepts an optional `field` query string parameter to
request a single field. At time of writing, the Matrix specification says:

> If no `field` was specified, the response should include the fields of the user's profile that
> can be made public, such as the display name and avatar.

Given this wording, homeservers currently already have the flexibility to decide whether some
fields are published over federation, and this proposal continues to allow this behaviour.

### Capabilities

A new capability `m.profile_fields` controls the ability to *set* custom profile fields and is
advertised on the `GET /_matrix/client/v3/capabilities` endpoint. Clients should check for this
capability before attempting to create or modify a profile field.

- **Capability Structure**:

```json
{
  "capabilities": {
    "m.profile_fields": {
      "enabled": true,

>       "disallowed": ["org.example.job_title"]

Reviewer 15: Given the flexibility of user-defined fields, this looks unwieldy. Have you considered an allow-list approach, instead of a deny-list?

Reviewer 15: Or in addition to the deny-list approach!

Author: I like that idea! I'll add both as optional keys so that we're not including empty lists unnecessarily, but gives homeservers plenty of scope to tell clients what they can do!

Author: We have this now, hopefully that works for you! https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/d97189eccaa475ca3767478c0e977ff2f9ed1458#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R146

Author: I've updated the capability and added this now, which hopefully covers this: https://github.com/tcpipuk/matrix-spec-proposals/commit/d97189eccaa475ca3767478c0e977ff2f9ed1458#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R162-R167

Author: Totally agree, thanks for calling that out! I've clarified that here: https://github.com/tcpipuk/matrix-spec-proposals/commit/4afb8b84e2879dbd85affaeb65e468eec9a3710d#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R163-R169

    }
  }
}
  ```

- **Behaviour**
  - **When capability missing**: Clients should assume extended profiles are supported, and that
    they can be created/written to. If a server intends to deny some (or all) changes, it SHOULD
    use the capability to advertise this to improve client experience.

>   - **When `enabled` is `false`**: Clients should expect to display profiles but NOT create or

Reviewer 15: There is a missing behaviour here - servers may wish to initially support profiles, but decide to restrict them in response to misuse. Changing `enabled` to `false` will prevent changes, but not prevent display of the custom fields that already exist.

Author: In the model defined so far, servers can choose which fields they present to clients, so a server could provide a limited profile that only contains `avatar_url` and `displayname` while not allowing any other fields to be added.

Do you think that's sufficient? I can at least clarify that in the text so implementations are aware.

I've got a future versioned endpoint in mind that would include metadata per-profile and per-field, so a server could easily instruct clients which fields should be hidden, but the aim here was to reuse the existing endpoints for an MVP then start a wider discussion on how much extra data would add value in the future.

Author: I've added this, which hopefully covers the issue? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/ae197255f91d936527285f65ca891757e29859a6

due care over time.

As this data is stored only at the global level, it won't allow users to modify fields per-room,
or track historical changes in profile fields. However, many users would struggle to track their
own data across many rooms, and publishing state events for every field change could quickly become
a heavy burden on servers and moderators.

This proposal recommends future MSCs add certain fields to per-room member events when there is
value in doing so - the current free-text fields added by this proposal are not considered to have
this value.

This proposal also does not offer a method to "broadcast" to other users or homeservers that
changes have occurred. This is intentional to keep the scope of this change narrow and maximise
compatibility with existing servers. A future proposal may wish to use an EDU (such as Presence) to
notify users and homeservers that these custom fields have been updated.

## Alternatives

An alternative approach could involve introducing a completely new API for extended profile information. However, this may lead to more significant changes and increased complexity for client developers.

## Security Considerations

> Since profile fields are public, there are minimal security risks associated with the transmission

Reviewer 15: This seems unfortunately worded — public fields are inherently more risky, because the contents are publicly accessible. This is an acceptable risk when folks fully understand that what they include will be made public, and they do so intentionally (which your next line gets to), but public != minimal security risks, so I'd find a way to reword this.

Author: I see your point. I was thinking from the standpoint of this being the security section (so the threat of leaking data shared with the server is minimal) but agree it would make sense to categorise the risk for clarity!

Author: I've tried to rephrase this in a balanced way: https://github.com/tcpipuk/matrix-spec-proposals/commit/d97189eccaa475ca3767478c0e977ff2f9ed1458#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439L325-R346

Does that seem appropriate, or could you recommend better wording? 🙂

of sensitive information; however, users should be made aware that any information they add will be
visible to others on the federated network. Clients *should* inform them of this.

Homeservers and clients *must* comply relevant privacy regulations, particularly regarding data
deletion and retention. Profile data *should* be cleared when a user is deactivated, and while
homeservers *should* cache remote profiles, they *should* avoid caching beyond 24 hours to minimise
the risk of unintended data persistence.

> To minimise the impact of abuse, clients should offer suitable defaults for the users they will

Reviewer 15: For server operators, what is the mechanism for removing abusive profile fields from users? 

As an example: a user creates a custom field, and as the key value, they include an online fundraising link for a proscribed terrorist organisation, or perhaps a few lines from their manifesto. 

Server operators are obliged to remove that content. Deactivation may be the method they choose, but it should not be the only method. I would like to see more addressing the mechanisms for server operators to manage the risk involved with these custom fields.

Author: Who may modify a users fields is defined by the server, so an implementation should be free to remove a user's fields.

I'd initially conceived of the reverse (e.g. an organisation with OIDC login would automatically populate some basic fields from the user DB like timezone or job title) but it could easily go both ways. 

I could clarify that in the text if that helps?

Reviewer 15: It would - more broadly I'd like to see safety considerations moved up to a first-class element of MSCs. Documenting some ways that the proposal could be used in harmful ways, and the steps implementations should consider to mitigate or prevent those harms would be time well spent.

That way more of us can play with these existing new features safely 😄

Author: I agree! I was following the proposal template from the repo when I wrote the initial draft of this, so perhaps it'd be worth raising an issue/PR to add another section that can broadly encourage more coverage of potential safety issues?

In the meantime, I'll try to get adjustments for this MSC written up in the next couple of hours 🙂

Author: I've added this so far: https://github.com/tcpipuk/matrix-spec-proposals/commit/d97189eccaa475ca3767478c0e977ff2f9ed1458#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R294-R297

Reviewer 16: This is where report user feature would be helpful, IMHO.

Author: I've raised follow-up [MSC4202](https://github.com/matrix-org/matrix-spec-proposals/pull/4202) to request this additional moderation feature, depending on [MSC3843](https://github.com/matrix-org/matrix-spec-proposals/pull/3843).

for a more versatile user profile. Specifically, it redefines the existing `avatar_url` and
`displayname` endpoints to be more flexible, while attempting to maximise compatibility with
existing clients and servers to help speed adoption.

Likewise, this proposal is designed to complement rather than replace
[MSC1769](https://github.com/matrix-org/matrix-spec-proposals/pull/1769) (Extensible Profiles as
Rooms). While [MSC1769](https://github.com/matrix-org/matrix-spec-proposals/pull/1769) offers a
more complex solution for extensible profiles, this proposal focuses on enabling the storage of
small, arbitrary key:value pairs at the global level.

This proposal does not seek to enforce the exact content or usage of these fields but rather to add
a framework for users to have extra data that can be further clarified and extended in the future
as community usage of these fields grows.

Homeservers could disable the ability for users to update these fields or require a specific list
of fields, but the intention of this proposal is that users will be presented with a form to enter
their own free-text fields and values. After using these very flexible fields, the community may

> opt to request a further extension to promote one or more fields to be specified per-room via

Reviewer 7: nit pick, but I don't know whether we should mention the mechanism of promotion at this juncture. I'm not sure it's *actually* codified that profile fields should always end up in member events if they are sound.

  "displayname": "John Doe",
  "m.example_field": "value1",
  "org.example.job_title": "Software Engineer"
}
```

## Server-Server API Changes

The federation endpoint `GET /_matrix/federation/v1/query/profile` will mirror the client-server
API changes to ensure profile information is consistent between local and federated users.

As per the current stable endpoint, it accepts an optional `field` query string parameter to
request a single field. At the time of writing, the Matrix specification says:

> If no `field` was specified, the response should include the fields of the user's profile that
> can be made public, such as the display name and avatar.

Given this wording, homeservers currently have the flexibility to decide whether some fields are
published over federation, and this proposal continues to allow this behaviour.

## Capabilities

A new capability `m.profile_fields` controls the ability to *set* profile fields and is advertised
on the `GET /_matrix/client/v3/capabilities` endpoint.

This capability deprecates the use of `m.set_displayname` and `m.set_avatar_url`, which are not
required when this capability is present.

> Clients SHOULD check for this capability before attempting to create or modify a profile field.

Reviewer 4: Presumably clients could just try setting it and see if they get a 403? I feel like clients only need to check this if they want to find out if it *can* be set before exposing that option to the user.

```suggestion
Clients MAY check for this capability before attempting to create or modify a profile field.
```

  they can be created or modified. If a server intends to deny some (or all) changes, it SHOULD use
  the capability to advertise this, improving the client experience.

- **When `enabled` is `false`**: Clients SHOULD expect to display profiles but NOT create or update
  fields. Any attempt to do so SHOULD result in a `403 Forbidden` error. This does not affect
  `avatar_url` and `displayname` fields, which are allowed for compatibility purposes.

- **When `enabled` is `true`**: Clients SHOULD allow users to create or update fields, except those
  keys listed in the `disallowed` array. Servers MAY specify an `allowed` array to allowlist fields
  that users can set. If both `allowed` and `disallowed` keys are provided, the `disallowed` one
  should be ignored. Clients SHOULD receive `400 Bad Request` or `403 Forbidden` responses if
  server-side policies prevent them.

## Key and Namespace Requirements

Profiles MUST be at most 64 KiB (65,536 bytes) in size, as measured in
[Canonical JSON](https://spec.matrix.org/v1.13/appendices/#canonical-json), including the
`avatar_url` and `displayname` fields.

> Homeservers SHOULD NOT enforce namespaces, as future expansions may be unknown to the server, but

Reviewer 4: > Homeservers SHOULD NOT enforce namespaces

I feel like homeservers SHOULD enforce at least *some* of the requirements of the CNIG (key length, character set, starts with `[a-z]`)?

I notice that an error is defined for `M_KEY_TOO_LARGE` which implies that there is some expectation here.

Author: Ah yes, the wording of this should be clearer... homeservers should _not_ be checking whether the key uses a _known_ namespace, but they must still enforce the common grammar. I'll try to make this a bit more explicit.

Author: I've had a go at this - what do you think? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/9b2918e3735f5aec02f6309fcbc81feaca985804

Author: Oh, I've screwed up some line wrapping, so I'll fix that in the meantime!

>   capabilities.

Reviewer 7: I really don't like this. Capabilities are *not* there for determining if a feature is supported by the server but whether it is permitted. Typically it's assumed by most spec features that a missing capability implies it's enabled for you, not disabled.

The versions endpoint is used to determine support, and it's virtually assured that most clients will be requesting that information anyway. Can we please have it so that `/versions` is used to determine support, and this field is used to determine if setting profiles is permitted (a distinction I find valuable).

Reviewer 12: The [capabilities system](https://spec.matrix.org/v1.11/client-server-api/#capabilities-negotiation) appears to be specifically geared towards optional features.

> The capabilities advertised through this system are intended to advertise functionality which is optional in the API, or which depend in some way on the state of the user or server.

And as far as I understood the new APIs will be optional?

> This feature will be implemented as optional but recommended, enabling a smooth transition and minimal disruption to existing deployments.

Author: Exactly, my assumption was that a future version might make the feature mandatory and remove the capability, but this allows a good chunk of time for people to move over without frustrating errors.

We know clients *will* want to implement this feature, we don't need to twist their arms over it.

Reviewer 7: I'm quite worried about making this optional. I understand the need to not want to twist arms, but that's why we have  Matrix spec *version*. Taking this apart, clients don't really need to do much to support this particular feature and those that want to support vNext Matrix version will have ample time to implement in the period where this is merged but not versioned.

I think we need to be ultra careful here about optional features in Matrix, the spec intentionally does not carve up features like this so the experience across clients and servers is predictable. Extensible profiles to me is analogous to custom event types in Matrix, and core to building good experiences on the platform (unlike say VoIP or url previews perhaps).

So, it's my belief here that we should be treating this proposal as a core enhancement to Matrix, and shouldn't make this optional to support in clients and servers who want to support the next version of Matrix.

Reviewer 11: Implementing the endpoint obviously shouldn't be optional. I'm not sure if there are any optional endpoints in the spec? (the module system in the C-S API is for features that are optional to clients, but all features are required for servers)

Allowing custom fields in profiles can be optional though, just like the ability to set a displayname/avatar is already optional.

Author: Fair enough, at the point I proposed this I was clearly too humble! 😂

We still want the capability to say clients are allowed to modify fields, and @Half-Shot [proposed](https://github.com/matrix-org/matrix-spec-proposals/pull/4133#discussion_r1760675217) adding the "disallowed" field to the capability to let clients know what fields they *can't* modify, so what do I now do with the capability we do have?

Is the suggestion then that when the capability is *missing* we assume *full* support, but if the server disables writes they can include the capability as "false" (because you can't) or "true" with a "disallowed" key to say there are exceptions to what you can write?

Reviewer 12: > Implementing the endpoint obviously shouldn't be optional. I'm not sure if there are any optional endpoints in the spec? (the module system in the C-S API is for features that are optional to clients, but all features are required for servers)

I think only very few are optional. For instance, `get_token` (which also has a capability): https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv1loginget_token

I'd agree that the new APIs from this proposal probably can be mandatory though.

Reviewer 7: > Is the suggestion then that when the capability is missing we assume full support, but if the server disables writes they can include the capability as false (because you can't) or true with a disallowed key to say there are exceptions to what you can write?

Correct. If missing then it's assumed true. If true with disabled then some fields are not writable, if false then the feature is disabled entirely.

>Allowing custom fields in profiles can be optional though, just like the ability to set a displayname/avatar is already optional.

Yep, and the job title example is the one I wanted to point at as the obvious use case for this.

Author: Thanks for confirming! How about this? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/591999fe535f79d4bb131d88bc88ff086aac7305

Reviewer 7: Wording LGTM!

- **When `enabled` is `true`**: Clients MAY allow users to create or update custom fields. However,
  individual updates may still receive a `400 Bad Request` or `403 Forbidden` response from the

>   homeserver if specific server-side policies prevent the action.

Reviewer 7: This makes it sound like some fields might be readonly. Would it not make more sense for the server to declare these fields explicity in capabilities, otherwise the UI impact will be intensely frustrating:

If this is explicit in the capabilities response, you can disable inputs with a message to explain the lack of support. If it's given only in response to a PUT/POST, then you can only show an error.

For example:

```json
    "capabilities": {
        "m.profile_fields": {
            "enabled": true,
            "disallowed": ["org.example.job_title"]
        }
    }
```

would be immeasurably helpful to client devs

Author: Great idea, how about this? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/fbb4e4497b70708862c9b81f62fc11b3204dff6b

Reviewer 7: I like, thank you :green_heart:

- **Namespace `m.*`**: Reserved for fields explicitly defined in the Matrix specification. Clients
  that do not recognise a field in this namespace MAY attempt to display it but SHOULD NOT attempt
  to update the content unless they understand its formatting and validation requirements.

- **Namespace `tld.name.*`**: For client-specific or unstable fields, using Java package naming
  convention (e.g., `com.example.custom_field`).

Following this change, clients could use `m.example_field` if that field is defined by the Matrix
specification, or `org.example.job_title` for organisation, client-specific fields, or MSC-backed
unstable features. Proposal [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175)
demonstrates the process of defining new fields in the `m.*` namespace.

> ## Error Handling

Reviewer 9: I copied that from the avatar URL profile endpoint, so probably is a spec bug that it was never documented?

> ## Error Handling

Author: I'm happy to take this opportunity to document it now? 🙂

> ## Error Handling

Reviewer 9: Sounds good! Was mostly implying it housing be controversial!

> ## Error Handling

Author: How's this? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/a11286a5091c0e8ac279265f8295ad819c336379

(I've since fixed the line-wrapping, as I forgot it the first time, oops...)

> ## Error Handling

Reviewer 14: That change documents it for the endpoints relevant to this MSC, but not for the original avatar URL profile endpoint.

I think that should probably be tackled separately anyways. Thanks for including the errcode here!

### 400 Bad Request: Request Exceeds Limits or Is Malformed

- **`M_BAD_JSON`**: Malformed request.

```json
{
  "errcode": "M_BAD_JSON",
  "error": "The provided JSON is malformed."
}
```

- **`M_PROFILE_TOO_LARGE`**: Exceeds total profile size limits.

```json
{
  "errcode": "M_PROFILE_TOO_LARGE",
  "error": "The profile data exceeds the maximum allowed size of 64 KiB."
}
```

- **`M_KEY_TOO_LARGE`**: Exceeds individual key length limits.

```json
{
  "errcode": "M_KEY_TOO_LARGE",

>   "error": "The key name exceeds the maximum allowed length of 255 bytes."

Reviewer 4: Technically speaking:

```suggestion
  "error": "The key name exceeds the maximum allowed length of 255 characters."
```

... not that it makes much difference if we only allow ASCII characters and use UTF-8 encoding.

Author: Yeah, I may link [the grammar](https://spec.matrix.org/v1.13/appendices/#common-namespaced-identifier-grammar) in a couple of places to ensure it's unmissable to the average reader 😅

  format, and as measured in UTF-8 byte counts, the key and value must not be longer than 128 bytes
  and 512 bytes respectively.
- Client-specific or unstable fields MUST use the Java package naming convention: `tld.name.*`.

Following support for this specification change, a user could enter a "My Timezone" field manually
in their client and the client would be expected to add a `u.My Timezone` key in their profile.
However, this would be expected to be a string value, even if the user types "null" or an integer
into the box.

In contrast, [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175) would use the
unstable key `us.cloke.msc4175.tz` and following approval would then support clients using the
`m.tz` key with the values/validation that MSC requires.

### Size Limits

Whenever "bytes" are referred to as a limit, this is calculated as UTF-8 bytes, so a two-byte code
point consumes two bytes of this limit. This size would be measured before any JSON encoding.

Until another MSC specifies otherwise:

- Each profile *must* be *at most* 64KiB (65536 bytes) in size, as measured in Canonical JSON.

> - Each key *must* be a string of *at least* one byte and *must* not exceed 128 bytes.

Reviewer 5: Is the size limit on keys intentionally inclusive of all fields? In the prior section the limit is only mentioned for user-defined fields.

Is there an explanation for how the limit of 128 bytes was decided?

Author: Mainly because it's easier to increase later than decrease.

128 characters is very long for a key name, and no one has ever told me a use case where they foresee needing more than ~50.

Do you think longer keys are likely to be required? The *values* are not limited outside of the `u.*` namespace, which was the main request I received before.

Reviewer 5: > The values are not limited outside of the u.* namespace,

I'm not too confident that the MSC is explicit about that, which is why i highlighted the section here. I'm fine with the restriction on the user-defined fields, since clients need to display them and those would cause issues if they were too long.

Author: That's fair. I'm AFK for a few hours, but is there anything you'd particularly like to see stated to make sure I don't disappoint? 🙂

Reviewer 5: ```suggestion
- Each key *must* be a string of *at least* one byte.
- Each key in the `u.*` namespace *must* not exceed 128 bytes.
```

Reviewer 5: I think the wording was implicitly catching all keys, do you see what i mean? ^

Author: Ah, I see the misunderstanding. That wording is specifically for the *keys*, not the *values*.

There's no minimum length for values, they can be empty/null values (otherwise we'd require everyone to have an `avatar_url` and `displayname`.

The *values* in the `u.*` namespace are limited to a maximum 512 bytes, but have the same limit on key length as any other field, so I'll try to make that clearer.

Author: Is this clear enough? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/4a9557f27ba755931dcf497a3c3ee5eeb14005d3

regarding data deletion and retention. Profile data *should* be cleared when a user is deactivated.
While homeservers *should* cache remote profiles, they *should* implement strategies that do not
exceed 24 hours to minimize the risk of unintended data persistence.

## Unstable Prefixes

### Unstable Profile Fields

The current Matrix specification technically already allows extra custom fields to be published in
a user's profile. However, as this proposal introduces additional requirements and allows custom
user-defined fields, an unstable prefix should be used on these fields until this proposal has
entered the API as stable:

```json
{
    "avatar_url": "mxc://matrix.org/MyC00lAvatar",
    "displayname": "John Doe",
    "uk.tcpip.msc4133.u.Custom Field": "field_value"
}
```

### Unstable Endpoints

`/_matrix/client/unstable/uk.tcpip.msc4133/profile/{userId}` would be necessary for the `PATCH` and
`PUT` methods allowed when the unstable capability (detailed below) is advertised by the server.

The existing `GET` method would act as normal and remain on `/_matrix/client/v3/profile/{userId}`

> without need for an unstable endpoint.

Reviewer 7: I'm *slightly* worried about this. I know that the spec allows custom fields, but I've never once seen an implementation take advantage of it. I guess is there a concern here that extra fields appearing on an existing endpoint might cause crashes in the cases of sensitive parsers (looking at you serde), or data previously thought to be small and cachable suddenly filling peoples local storages.

maybe the SCT can share their thoughts

Author: This is partly why the upper limit on profile size is quite strict. We don't want people to suddenly be able to have 1MiB of JSON on an endpoint that was likely under 1KiB before.

Currently I believe implementations expect the profile to be valid JSON, but read these specific 2 keys out, and ignore everything else. Servers are *allowed* to include extra fields in the profile, but there's no agreed standard on what they might contain, so clients and servers typically ignore any extra data and just individually handle the two fields they know about.

My assumption is that most implementations will separately store custom fields to `avatar_url` and `displayname` (because it's a simpler database change) but this MSC just provides APIs for this data. I'd personally prefer implementations store a block of JSON and just ensure the writes are atomic (as hungryserv has done) as it'd allow a lot more flexibility on types and sizes, as well as making it easier to calculate the full profile size.

Author: @tulir does Beeper publish anything over federation at all? Can you show an example?

Reviewer 11: The custom profiles are only for bridged users which are unfederated

`avatar_url` and `displayname` fields.

Keys MUST follow the [Common Namespaced Identifier Grammar](https://spec.matrix.org/v1.13/appendices/#common-namespaced-identifier-grammar),
with the following considerations:

- **Namespace `m.*`**: Reserved for fields explicitly defined in the Matrix specification:
  - Servers SHOULD NOT check whether a key is known to be in the Matrix specification, as future
    expansions may be unknown to it.
  - Clients that do not recognise a field in this namespace MAY attempt to display it but
    SHOULD NOT attempt to update the content unless they understand its formatting and validation
    requirements.

- **Namespace `tld.name.*`**: For client-specific or unstable fields, using Java package naming
  convention (e.g. `com.example.custom_field`).

Following this change, clients could use `m.example_field` if that field is defined by the Matrix
specification, or `org.example.job_title` for organisation, client-specific fields, or MSC-backed
unstable features. Proposal [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175)
demonstrates the process of defining new fields in the `m.*` namespace.

> ## Error Handling

Reviewer 14: When it comes time to write the spec PR, the author will be trying to match up error codes to each individual endpoint.

Is it expected that all status code/error code combinations here can be used for all Client-Server and Server-Server endpoints introduced in this MSC? If so, could you explicitly say so at the beginning of this section?

Author: My assumption was that the Server-Server endpoints would have identical errors to the current ones, as federation access is still read-only in this MSC.

I'm happy to try to make the Client-Server endpoints a bit clearer on which errors apply to each of them though!

I might not have time to work on this over the weekend as my schedule is pretty full, but I'll try to get time ASAP 🙂

Reviewer 14: > My assumption was that the Server-Server endpoints would have identical errors to the current ones, as federation access is still read-only in this MSC.

That's fine! It's just not clear with the current copy. I'd explicitly state what you just said in the MSC to make it clear.

> I'm happy to try to make the Client-Server endpoints a bit clearer on which errors apply to each of them though!

That would be appreciated!

Author: I've written this: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/2a862350e6847297c420174576d6f87273dc1459

Would you prefer I literally write which errors apply to each endpoint, or is a section like this ok? I'm not sure what works best for spec-writers, so let me know!

Reviewer 14: That indeed makes it crystal clear - thank you!

This proposal does not seek to enforce the exact content or usage of these fields but rather to add
a framework for users to have extra data that can be further clarified and extended in the future as
community usage of these fields grows.

Homeservers could disable the ability for users to update these fields, or require a specific list
of fields, but the intention of this proposal is that users will be presented with a form to enter
their own free-text fields and values. After using these very flexible fields, the community may
opt to request a further extension to promote one or more fields to be specified per-room via
member events.

### Client-Server API Changes

1. **GET `/_matrix/client/v3/profile/{userId}/{key_name}`**: This endpoint will replace the existing
   profile endpoints. It will return the value of the specified `key_name`:

```json
{
    "key_name": "field_value"
}
```

For example, requesting `/_matrix/client/v3/profile/@alice:matrix.org/displayname` would return:

```json
{
    "displayname": "Alice"
}
```

2. **PUT `/_matrix/client/v3/profile/{userId}/{key_name}`**: This endpoint will set the value of the
   specified `key_name`:

```json
{

>     "key_name": "new_value"

Reviewer 12: Clarifying question: Would `null` be allowed here, effectively making the operation equivalent to `DELETE`?

Reviewer 9: I think that's how displayname and avatar URL work, but it is pretty meh as a solution...

Author: Are the `avatar_url` and `displayname` actually deleted when that happens? I assumed we'd want to allow keys to exist with null values.

Reviewer 12: Hm, the current spec for the display name and avatar endpoints doesn't seem to cover null values explicitly.

It's slightly confusing actually. The GET endpoints say the field is missing in the 200 response if it hasn't been "set":

> The user’s avatar URL if they have set one, otherwise not present.

But they also define a 404 response when "there is no" value:

> There is no avatar URL for this user or this user does not exist.

I'm not sure if that was somehow meant to differentiate a `null` value from a missing value? 😵‍💫

Reviewer 9: They're marked as null in the database and...Synapse does weird things with them. Some APIs it pretends they're unset and others it nulls them, I think?

Reviewer 12: Synapse also maps empty strings to `None` [apparently](https://github.com/element-hq/synapse/blob/4ac783549c5bac7a490a715d359f330bb0b1a161/synapse/handlers/profile.py#L305). :eyes: 

FWIW, my main point was that it might be good to be explicit. Allowing `null` values (making `PUT null` different from `DELETE`) seems ok to me.

Author: Great! Does that mean you're happy as is?

Reviewer 12: Yes, thank you. :+1:

**400 Bad Request**: When the request is malformed, exceeds specified limits, or the profile JSON
object is larger than 64KiB:

- **Error code for malformed request**: `M_BAD_JSON`

```json
{
    "errcode": "M_BAD_JSON",
    "error": "The provided JSON is malformed."
}
```

- **Error code for profile/field exceeding size limit**: `M_TOO_LARGE`

```json
{
    "errcode": "M_TOO_LARGE",
    "error": "The profile data exceeds the maximum allowed size of 64KiB."
}
```

or

```json
{
    "errcode": "M_TOO_LARGE",
    "error": "Profile field u.Bio exceeds maximum allowed size of 512 bytes."
}
```

or

```json
{
    "errcode": "M_TOO_LARGE",
    "error": "The key name exceeds the maximum allowed length of 128 bytes."
}
```

**403 Forbidden**: When the user does not have permission to take this action on a specific key,
such as when the server policy (e.g.
[MSC4170](https://github.com/matrix-org/matrix-spec-proposals/pull/4170)) restricts such actions:

- **Error code refusing to create additional keys**: `M_FORBIDDEN`

```json
{
    "errcode": "M_FORBIDDEN",
    "error": "The user has exceeded the maximum number of allowed keys in their profile."
}

> ```

Reviewer 9: This feels like it should be `M_TOO_LARGE` too, but would be curious what others think.

Author: Ah, I wasn't sure if it was an "you're not allowed to create more" rather than "your input is too large" but probably quite a subjective interpretation!

I'm happy to use "too large" if that's easier, it's no bother, but also happy to wait if you'd rather poll the SCT?

Reviewer 12: I'd personally lean towards `M_TOO_LARGE`, too. The spec currently describes `M_FORBIDDEN` as "Forbidden access". However, the issue here is less that you don't have access and more that you sent illegal data.

Reviewer 9: Yes, pretty much what @Johennes said. 👍

Author: Hopefully this suits now! 🙂 https://github.com/tcpipuk/matrix-spec-proposals/commit/39d5fa271589824bf0a88778e233710c16f928f7?diff=unified&w=0#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439L207-L210

in their client and the client would be expected to add a `u.My Timezone` key in their profile.
However, this would be expected to be a string value, even if the user types "null" or an integer
into the box.

In contrast, [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175) would use the
unstable key `us.cloke.msc4175.tz` and following approval would then support clients using the
`m.tz` key with the values/validation that MSC requires.

### Size Limits

Until another MSC specifies otherwise:

- Each profile *must* be *at most* 64KiB (65536 bytes) in size
- Each profile may contain *up to* 50 keys (including `avatar_url` and `displayname`)
- Each key *must* be a string of *at least* one byte, and *must* not exceed 128 bytes
- Each value *must* be a string and *must* not exceed 512 bytes

These sizes are measured as encoded in
[Canonical JSON](https://spec.matrix.org/v1.11/appendices/#canonical-json), so `🛸` would consume
two bytes of the limit.

Future MSCs may allow fields that are not string values, or add exceptions to the limits, but these
current limits have been chosen to allow servers and clients to have predictable upper limits for

> performance and caching purposes.

Reviewer: Up until the last paragraph of this section, the size requirements could be argued as supporting non-string values through Canonical JSON. Similar to how events cannot exceed 65KiB, and are objects, profiles can be serialized and size-checked against that too. 

I'd suggest removing the last paragraph of this section, and clarifying more overtly that profiles are measured against their encoded forms. 

Pushing expansion of the grammar to a future MSC also doesn't really feel like it'd work: clients and servers may heavily validate the endpoints, and thus error out when a client tries to use MSC$future's object storage mechanism. This would also force that future MSC to create new endpoints, leaving the user with 2 or more "versions" of their profile. 

If there's technical reasons which make the above too difficult, they should be described in the MSC itself.

Reviewer 9: Maybe something that would be a decent user experience:

* Define a limit on custom fields (`u.Blah`) since those can only contain string values.
* Leave it up to each individual field to define its limits.
* (Maybe?) Include limits for display name and avatar URL. (But this probably could be a separate MSC.)
* Keep the overall limit for the block (but align to the size of events -- 65 KiB not 64 KiB).

Author: This seems difficult to provide a good user interface for.

This MSC is designed to take an object that has multiple string:string keys in it and provide an interface for clients to quickly add/remove keys on it.

How should a client show a list of numbers, or a large nested JSON object? It feels antithetical to the goal of having easy fields users can quickly add/edit with implementation soon in both clients and servers. 

If I can't tell clients how to deal with the fields I'm telling them to add, it doesn't feel appropriate to include, when deeper functionality could be carefully crafted in a versioned endpoint we can spend time discussing after this is already live, and in the meantime the people can set important personal metadata like bios, timezones, pronouns, social profiles, etc.

Reviewer 11: +1 to only restricting user-defined `u.` fields.

Also if only user-defined fields have restrictions, then I think they should use utf-8 byte counts rather than canonical json lengths for size limiting to make it easier for clients.

> align to the size of events -- 65 KiB not 64 KiB

events are 64 KiB = 65536 bytes ≈ 65.5 KB

Reviewer 9: > [events](https://spec.matrix.org/v1.11/client-server-api/#size-limits) are 64 KiB = 65536 bytes ≈ 65.5 KB

That's what I get for not checking the spec. 👍 Thanks for the clarification.

Author: I've just committed a couple of changes here, so it might be easier to just read the "Key/Namespace Requirements for Custom Fields" and "Size Limits" sections again, but I believe I've answered these concerns: https://github.com/tcpipuk/matrix-spec-proposals/blob/main/proposals/4133-extended-profiles.md#keynamespace-requirements-for-custom-fields

Author: @turt2live are you happy with the changes?

Author: @turt2live As per [this comment](https://github.com/matrix-org/matrix-spec-proposals/pull/4133#pullrequestreview-2338095734), do you think the concern about size limits is safe to resolve now?

I've also updated the proposal to rely on [Common Namespaced Identifier Grammar](https://spec.matrix.org/unstable/appendices/#common-namespaced-identifier-grammar), which I hadn't realised existed before 🙂

  regulations will be enforced, particularly in terms of data deletion and retention policies.
- Custom fields will not trigger state events in rooms, maintaining account-wide metadata without
  creating state events or other moderation issues.

## Potential Issues

Initial challenges are primarily ensuring uniform support for custom fields across different
servers and clients during the rollout phase: users may come to expect that other users will check
their supported languages before communicating with them, but the sender's server does not support
custom profiles.

As such, this MSC is designed to be as simple as possible to get initial functionality and data
structures implemented widely in both clients and servers, so further extensions can be debated,
implemented, and tested over time.

As this data is stored only at the global level, it won't allow users to modify fields per-room,
or track historical changes in profile fields. However, this is for performance and moderation
reasons, as many users will struggle to maintain many fields of personal data between different
rooms, and publishing state events for every field change could become an additional burden on

> servers and moderators.

Author: I agree that further features will likely be desired to make tracking profile changes more convenient in the future, but I don't think anyone is eager to see this MSC dig any deeper into any more fundamental parts of Matrix.

I agree it could be interesting to propose an MSC that prevents all/some profiles from being viewed in a room, but that would definitely make a better MSC on its own, especially as currently a user can easily be banned if their profiles are deemed unacceptable.

Do you foresee it being a common occurrence that someone's profile is found to be unacceptable but room admins choose for them to stay in the room with their profile hidden? That sounds like quite an edge case to me.

> servers and moderators.

Reviewer 5: I meant it in the sense that profiles will be hidden by default and later enabled selectively. Generally this kind of opening up of features as a user participates more in a community is the path that Draupnir is going down.

> servers and moderators.

Author: Ah, like a power level requirement for profiles to be visible?

I'm certain that shouldn't be touched by this MSC, as we only touch the global profile data and not anything inside a room, but it sounds like a sensible (and relatively simple/short) MSC to raise to add that feature.

This MSC is intended to inspire people to explore and add more features, and provide a framework to help that to happen, so I'm excited to hear it's doing that in multiple ways than just specific types of field to add!

> servers and moderators.

Reviewer 17: So, because it seems like there is some disagreement about why this is relevant from a moderation perspective, I want to add my own view on things:

We currently often have people put abusive content into their displayname. This can be removed by redacting the membership event. This is important even when banning a user, because it affects how the historical timeline is rendered.

For a very similar reason status messages (part of presence) aren't that widely supported. They are hard to moderate, because showing them in a visible place means room admins have no control over it.

I think this MSC is in a similar boat. By not having the extensible fields be part of the membership event, they can't be redacted. As such showing these fields is an abuse concern. And for that reason I think this MSC should address that in some way. One option is to put those fields into member events (we already limit them in size, so I assumed that should have been the case). Another option is to explicitly mention that such fields should not be shown for banned members or similar. A third option could be a different mechanism, but I think there should be some mechanism to allow room moderators to remove extensible profile content from being shown for abusive users.

> servers and moderators.

Reviewer 7: >Another option is to explicitly mention that such fields should not be shown for banned members or similar.

I'd go further and say clients probably shouldn't show profiles to users within the context of a room if they are not joined to the room (so banned, or kicked, or w/e). 

However, I think that's more of a concession for privacy than anything. This feature is operating at the global level rather than a room level in Matrix, so it's not really for room admins to police the profiles. Rather, I'd prefer clients treated profiles as untrusted input and the user of the client chooses their level of comfort from showing information from strangers.

So it might be the case that outside of known fields like timezones and so on, clients *don't* display fields by default unless the user is trusted. That might mean they share a DM with you, or some other heuristic. Equally you might decide you don't care and profiles are just globally visible.

Either way, I don't think this MSC needs to put in rules for this to happen, it's an implementation detail how profiles are rendered to users. This MSC deals with the ability to get and set profiles, and clients should ideally choose to protect their users from rogue content. We could include some advisory text in the MSC, as perhaps a hint to the client authors.

> servers and moderators.

Author: That's a great suggestion! I'm a big fan of users getting to choose, as long as clients offer sensible defaults.

For example, if I mention @notinthisroom:server.com and the user wants to be able to see profiles for users not in the room, they should be able to do so. Equally, some users may want to "mute" all custom fields and I think they should have a right to do so. 

Putting some text in to "Implementation" to hint at options for this makes sense, but I agree ultimately it should be up to client implementers to decide on the defaults, and for users to choose what they feel comfortable seeing.

> servers and moderators.

Author: @deepbluev7 @Gnuxie do these solve the problem? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/26c59f7feb097d6dcf19efbe919e90e7fd53bef6

> servers and moderators.

Reviewer 17: @tcpipuk , I am not sure if that addresses the problem for the common case. Like often some rooms are very slow to remove spammers. This means you are very likely to share a room with a spammer, some rooms may even not have any moderators left.

I think you both are misunderstanding the problem. That this MSC is not modifying the existing PDUs is exactly why it is harder to moderate for room moderators. Now I also don't want moderators to decide, what you can display in a room, but they should be able to remove abusive content from their communities.

Let's say for example that someone builds a pronouns field based on this MSC. Innocent clients might assume, that it is fine to display that next to the displayname in the timeline. And that will work fine for most users and be very useful, you will then see "Displayname [pronouns]" for every user. But nothing prevents a user from changing their pronouns to "Displayname [buy sketchy here]" or worse.

Because the displayname and other existing profile fields exist in the member event, those can be redacted by the room admin. This does not apply to external fields like proposed by this MSC, so you will now see "mxid [buy sketchy here]" or similar in the timeline. This is a new complication introduced by this MSC specifically because it moves those fields out of PDUs and allows them to be edited externally at any time. And for that reason I think it is important for this MSC to outline at least some reasonable suggestions for when the profile fields can be considered as trusted and be displayed in the UI without user interaction.

I don't think showing the profile of only people you share a DM with is very useful for what people seem to want to do with them. Otoh I do think the heuristic of sharing any room is insufficient, since not every room has active moderators (and afaik the MSC to check if you share a room isn't merged yet, so it would be hard to check in some clients as well). I think something like "don't show profile fields of banned users by default" probably is sufficient, but I haven't done an extended analysis of that. Another option is of course to put it into the member event, which is something we already do for other fields and as such we can piggy back on the moderation concerns for those. Not showing details for redacted member events could work as well, but I think those are a bit hard to distinguish from people just not having set a displayname, especially with lazy loading.

So in short, I don't think the text in that commit is sufficient, but maybe you can explain why it should be sufficient by explaining the tradeoffs in more detail? Alternatively maybe someone has more T&S experience than me and can tell me that my concerns are unfounded?

And @Half-Shot, I don't consider it an implementation detail. If Matrix wants to take T&S seriously, I don't think T&S concerns should be left to clients to figure out for new MSCs, but features should be safe by default or have their risks documented.

> servers and moderators.

Author: I'm not sure I'm qualified enough in T&S to answer all of these points, but I'll answer what I can about this MSC:

My point in the text was that clients should use *reasonable* defaults, and suggested *at a minimum* it should only include rooms where both users are members, but also suggested only members in the current room as an example.

Regarding pronouns, this MSC would technically allow a user to create a `u.Pronouns` field, but that would be freetext and shouldn't be treated as "special" by a client. 

Like the `m.tz` field proposed by [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175), someone could propose an `m.pronouns` field that would be treated specially by clients, but that would have specific formatting defined by that MSC, with the intention that clients would check this meets the spec before displaying to users. If someone were to manually create a field like `{"m.pronouns": "buy my stuff"}` the client should reject this and not display it.

> servers and moderators.

Reviewer 5: > Regarding pronouns, this MSC would technically allow a user to create a u.Pronouns field, but that would be freetext and shouldn't be treated as "special" by a client.

I've checked the implementation of this MSC in Element web and it doesn't seem like it displays _Custom fields_ (ie `u.*` fields)? Isn't the intent of those fields in this MSC so that the inclusion of those fields can be shown in a profile view within clients? This is the concern for me, we don't know what this will look like or how it will be abused. But it is inevitable that people will put crap in there. And it's obviously a major part of this MSC and why this MSC is highly desired. 

> A specific example is that if you run a closed-federation enterprise server, then showing more fields is more desirable than being default-protective.

Sure but it is far easier to go from protective to liberal than it is in the opposite direction, especially in the context of Matrix. And in a moderation context it's the difference between being proactive (ahead) and reactive (behind).

> servers and moderators.

Author: Assuming we're not about to replicate all custom fields into member events (at least, not in this first MSC, but there's definitely plenty of interest to build on it) what would you like to see change?

I'm absolutely interested in pleasing the community at large (this MSC is building a framework for future MSCs to build upon to add features we lack in other platforms as well as new features other platforms lack) so would like to find a balance that meets these needs within the specific scope this MSC covers. 

My main concern here is that we don't lump the cart in with the horse - the idea here is to introduce very clear specific endpoints for clients and servers to start implementing, then discuss further ways we can make the most of fields.

> servers and moderators.

Reviewer 5: > Assuming we're not about to replicate all custom fields into member events (at least, not in this first MSC, but there's definitely plenty of interest to build on it) what would you like to see change?

At the minimum the custom fields functionality, including UX for entering them needs to be implemented within a client. Then we will need to evaluate what that will look like in a public matrix room, and decide how to hide them. Hiding the profile when the member's membership isn't `join`, `invite`, or `knock` is a good start though.

Alternatively the custom fields need to be taken out of the MSC.

> My main concern here is that we don't lump the cart in with the horse

I don't think that is controversial to ask for the above or stopping this MSC getting over the line. As people will expect clients to implement this functionality if the MSC merged anyhow. The only difference is whether this is an authoritative part of the spec or not.

> servers and moderators.

Author: > Then we will need to evaluate what that will look like in a public matrix room, and decide how to hide them.

Ah, I see the confusion here.

The MSC is only introducing custom freetext fields that will *only* be displayed inside a user's profile - you can already see profiles in a variety of clients, they just don't show extra custom fields underneath.

The choice *in this MSC* is between showing those fields *inside* the user profile or not showing them at all - this MSC does *not* add any special fields or tell clients they should be displayed out in the room anywhere.

Where this will be important is in extensions ([MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175) is an example) that specify fields that should be used by the client to perform extra functions.

This is exactly why I want each of these to occur in its own MSC, so they can be discussed by the community, and won't complete FCP until a sufficient number of people are happy with the UI recommendations and moderation controls.

> Hiding the profile when the member's membership isn't join, invite, or knock is a good start though.

Thank you! I'll clarify the member states in the recommended client deployment defaults.

I totally agree room admins shouldn't have to fear content from abusive users being shown inside a room, but in the same way a room admin doesn't get to choose the theme a client uses or the location of the UI elements within that client, I strongly believe a user should have the *option* to see all profiles if they *choose* - this is an account-wide feature and many clients already include "Ignore" features to suppress all content from users they don't want to see.

I totally agree that options to give admins *additional* controls (e.g. setting a minimum power level) would be worth discussing in a further MSC, but as I mentioned above, this MSC is very intentionally limited to keep visible changes to a minimum moderation and ensure that any bigger client changes would be properly discussed before they are made.

> servers and moderators.

Reviewer 5: > The MSC is only introducing custom freetext fields that will only be displayed inside a user's profile - you can already see profiles in a variety of clients, they just don't show extra custom fields underneath.

There isn't any confusion. The profile pane is typically opened from or displayed alongside the room timeline view. This is also how it has been opened in the screenshot shown on the the Element web PR https://github.com/element-hq/matrix-react-sdk/pull/20. 

> The MSC is only introducing custom freetext fields that will only be displayed inside a user's profile - you can already see profiles in a variety of clients, they just don't show extra custom fields underneath.

Yes, and so far this part of the MSC has not been implemented, and really should be if it is going to remain part of it.

> servers and moderators.

Author: The implementation requirement for this MSC has been met, however I've opened [MSC4202](https://github.com/matrix-org/matrix-spec-proposals/pull/4202) as a possible solution for reporting inappropriate profile content.

There are many possible alternatives you may prefer, such as reporting to room admins by extending [MSC4119](https://github.com/matrix-org/matrix-spec-proposals/pull/4119), but those endpoints are probably best discussed in the relevant MSCs.

> servers and moderators.

Author: Having discussed this with Tulir, it appears [MSC3843](https://github.com/matrix-org/matrix-spec-proposals/pull/3843) already allows reporting users using their `m.room.member` event, so I've updated [MSC4202](https://github.com/matrix-org/matrix-spec-proposals/pull/4202) to just update client behaviour so they show a report button on profiles.

> servers and moderators.

Reviewer 5: > The implementation requirement for this MSC has been met

How is that possible when custom fields unimplemented? The implementation in element web does not show them or allow you to edit them?

> servers and moderators.

Author: To my knowledge, a partial implementation is required. I'd love more implementations on this MSC, but IIRC previous MSCs have cited a demonstration implementation in a client or server that didn't actually exist.

I very much welcome more implementations though. I'm currently working with a couple of server contributors to get more server implementations, but I'd be overjoyed if you happen to know any client contributors that would consider adding more 💜

> servers and moderators.

Reviewer 18: Please do not add custom fields to `m.room.member` events. Those events form part of the core auth rules for the room, and are expensive enough to maintain without bloating the size of them.

Why can't we fix the moderation issues with just mentioning in the MSC:
> When showing profile information in the context of a room, if the `m.room.member` event is redacted (missing displayname/avatar_url) then clients MUST NOT show any other custom profile fields in the context of that room.

This would address @deepbluev7 's concern about:

> Let's say for example that someone builds a pronouns field based on this MSC. Innocent clients might assume, that it is fine to display that next to the displayname in the timeline. And that will work fine for most users and be very useful, you will then see "Displayname [pronouns]" for every user. But nothing prevents a user from changing their pronouns to "Displayname [buy sketchy here]" or worse.

Because a redacted displayname (the current flow in today's Matrix) would also force redaction of _all_ custom fields.

> servers and moderators.

Author: As the area of concern has been exported to [MSC4208](https://github.com/matrix-org/matrix-spec-proposals/pull/4208), I think this thread should be safe to resolve now.

such as when the server policy (e.g.
[MSC4170](https://github.com/matrix-org/matrix-spec-proposals/pull/4170)) restricts such actions:

- **Error Code**: `M_FORBIDDEN`

```json
{
    "errcode": "M_FORBIDDEN",
    "error": "You do not have permission to modify this field."
}
```

**404 Not Found**: When attempting to `GET` or `DELETE` a profile key that does not exist:

- **Error Code**: `M_NOT_FOUND`

```json
{
    "errcode": "M_NOT_FOUND",
    "error": "The requested profile key does not exist."
}
```

### Propagation of Profile Fields to Membership Events

The existing fields, `avatar_url` and `displayname`, will continue to trigger state events in each
room. These fields are replicated per-room via member events. Other fields, however, will **not**
trigger state events in rooms. They will exist solely at the global level and are intended for
storing metadata about the user that does not need to be replicated in each room.

- **avatar_url** and **displayname**: Changes to these fields will generate state events in all
  rooms the user is a member of.
- **Custom fields**: These are stored in the user's global profile and do not generate state events

>   in rooms.

Reviewer 3: What happens if a user specifies a custom field in their membership event? Specifically, thinking about being override them at a room level, similar to displayname and avatar_url currently can be overriden.
This would be nice for eg. a pronouns field, where one might want to use a different set in a private context, such as in a friends group, or eg. a bio field where users may want to add a notice, ... in context of the room.

Example from a moderator on the Discord Developers guild on Discord:
![image](https://github.com/user-attachments/assets/7c19f7cd-3d51-4f39-956a-ff62da8cc91f)

I feel like it's a good idea to make this part of the base MSC rather than making it a separate one, as it may create confusion when some clients implement it and others don't. Besides, this would follow the existing logic of m.room.member's displayname and avatar URL fields.

Author: As far as I can tell, this MSC doesn't touch member events.

As we've already discussed in an earlier thread, it appears #4069 is already raised for this behaviour?

Reviewer 3: Correct, the MSC doesn't touch member events directly, but does touch profiles and how the data is looked up, which as far as I'm aware are pretty tightly coupled with member events as they exist today.

Though, I don't see how #4069 would interact with this MSC, as that is just an opt-out for propagating global displayname/avatar_url changes to rooms (profile owner's side), while this MSC changes where a client looks for profile data (other user's side, omitting m.room.member from that chain as a whole).

The part that I'm missing here, is defining what clients are supposed to do if a given room's member event explicitly contains a custom field (eg. I modified my own member event in a room, with the intent of overriding the global profile's value). #4069 does not touch on this at all, keeping "member event data overrides profile data" as is.

Author: This is my point though - this MSC exclusively extends the global profile, which has always existed and is not touched by this MSC.

It might make sense to have a similar MSC for member events to allow certain custom fields to be created for them, however I think it would need to be far more tightly controlled than this one. 

For example, a future MSC to add a spec'd `m.pronouns` field could additionally say that field can be added to member events to specify pronouns per room, but it wouldn't require any features added to this MSC to do that.

Reviewer 7: I concur that I think another MSC for how profiles relate to member events should be raised, but I wouldn't block this MSC on it.

  attempt to update the content in case it has special requirements.
- The namespace `u.*` is reserved for user-defined fields. The portion of the string after the `u.`
  is defined the display name of this field. These user-defined fields will always be string
  format, and as measured in UTF-8 byte counts, the key and value must not be longer than 128 bytes
  and 512 bytes respectively.
- Client-specific or unstable fields MUST use the Java package naming convention: `tld.name.*`.

Following support for this specification change, a user could enter a "My Timezone" field manually
in their client and the client would be expected to add a `u.My Timezone` key in their profile.
However, this would be expected to be a string value, even if the user types "null" or an integer
into the box.

In contrast, [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175) would use the
unstable key `us.cloke.msc4175.tz` and following approval would then support clients using the
`m.tz` key with the values/validation that MSC requires.

### Size Limits

Whenever "bytes" are referred to as a limit, this is calculated as UTF-8 bytes, so a two-byte

> UTF-16 character consumes two bytes of this limit.

Reviewer 11: I don't think mentioning UTF-16 makes sense in this context. Code points are defined in Unicode. All code points can be encoded in multiple different ways (UTF-8/16/32). Depending on the encoding, some code points take more bytes than others. UTF-8 uses 1-4 bytes per code point.

Might also be good to mention that the string lengths are counted before any JSON-encoding

Author: How does this sound? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/a9546aa358408a69b10da0f08e07461c3e825868

implemented, and tested over time.

As this data is stored only at the global level, it won't allow users to modify fields per-room,
or track historical changes in profile fields. However, this is for performance and moderation
reasons, as many users will struggle to maintain many fields of personal data between different
rooms, and publishing state events for every field change could become an additional burden on
servers and moderators.

In a similar vein, this proposal offers no method to "broadcast" to other users or homeservers that
changes have occurred. This is intentional to keep the scope of this change narrow and maximise
compatibility with existing servers. A future proposal may wish to use an EDU (such as Presence) to
notify users and homeservers that these custom fields have been updated.

## Alternatives

An alternative could be to allow for more specific endpoint modifications or to introduce a
completely new API specifically for extended information. However, these approaches could lead
to more significant changes to the existing API and higher complexity for client developers.

> ## Security Considerations

Reviewer 5: If changes to the extensible fields are not propagated to rooms (via membership), and if there is no [broadcast](https://github.com/matrix-org/matrix-spec-proposals/pull/4133/files#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R318-R321) to stand in place. Then automated scanning of profiles for scam links or other content, becomes a lot lot harder. Since previously each time membership changed a profile could be scanned.

Author: I agree. I've mentioned in a few places that I like the idea of a future MSC "borrowing" an EDU type (e.g. presence) to notify that profiles have been updated, as the presence status field could easily just move into the profile and servers would notify other servers that a user has updated their profile instead.

I also agree it would make sense for a future MSC to allow some custom fields to be replicated in member events, but it could quickly become a lot of PDUs if a state event was created *every* time *any* field was updated.

My initial goal here is for profiles to be requested when a client asks for them, so if you share a room with someone and have no interest in interacting with them, you could quite easily never look at their profile.

Ideally we want to get people storing and reading data from fields soon in an "only when asked for" way, then can expand this further with MSCs depending on what the community wants to see.

Author: Sorry, I'm confused by the phrasing of this - are you asking whether this MSC is going to offer the ability to track the content on the other end of URLs to proactively warn moderators when the content at the destination has changed?

If so, that seems quite a tangent from the functionality provided by this MSC.

Do you have an example of a feature used elsewhere in Matrix (e.g. in URL previews, or any other URL-based content in Matrix) where a process like this is done?

Reviewer 19: I wanted to raise awareness that getting notified about profile changes isn't a silver bullet in the sense that verification of a URL is not possible in practice (at all, ever). It is even harder when the link is entered in a form which is not autolinked (i.e., `t.me/123`). If you were already mitigating this by periodic scanning of URL (for 404 or change), you can implement periodic scanning of profiles the same way without relying on broadcasting their changes.

This in unrelated, but it is easy to target fake URL previews to the user agent & IP of each HS previewer (i.e., showing benign content to the HS of mods while showing edgy content to everyone else).

Author: I agree URLs can be a challenge for moderation, but it's unclear what changes you're asking for.

Do you want me to add text inside the MSC warning that malicious users might change the content at the destination of a URL, or maybe a suggestion for client implementations to warn users before visiting a link in a profile?

# MSC4133: Extending User Profile API with Custom Key:Value Pairs

*This proposal aims to enhance user profiles within the Matrix ecosystem by introducing
customisable key:value pairs to global profiles. Allowing users to add arbitrary public information
to their profiles (such as preferred languages, organisational roles, or other relevant public

> details) should enrich user interactions without impacting existing functionality.*

Author: Corrected, thanks!

## Proposal

Currently, the Matrix protocol supports limited user profile fields: `avatar_url` and
`displayname`. This proposal extends the API to include custom fields, enabling users to add
key:value pairs to their global profiles. This extension provides a flexible framework for users to
share additional public information.

This proposal is designed to be simple and compatible with existing clients and servers,
facilitating quick adoption. It complements, rather than replaces,
[MSC1769](https://github.com/matrix-org/matrix-spec-proposals/pull/1769) (Extensible Profiles as
Rooms), by focusing on global profile data without the complexity of per-room profile management.

## Client-Server API Changes

> - **Get a Profile Field**

Author: I have done this up top: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/9b4741e1f19b909173be1f263bbea6b3fa2ebe8c#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R11

Or would you prefer a specific mention in the actual CS API section itself?

  - **Endpoint**: `GET /_matrix/client/v3/profile/{userId}/{key_name}`
  - **Description**: Retrieve the value of a specified `key_name` from a user's profile.
  - **Response**:

```json
{
  "key_name": "field_value"
}
```

*Example*: Requesting `GET /_matrix/client/v3/profile/@alice:matrix.org/displayname` returns:

```json
{
  "displayname": "Alice"
}
```

- **Set a Profile Field**
  - **Endpoint**: `PUT /_matrix/client/v3/profile/{userId}/{key_name}`
  - **Description**: Set or update the value of a specified `key_name` in the user's profile,
    if permitted by the homeserver.
  - **Request Body**:

```json
{
  "key_name": "new_value"
}
```

*Example*: Setting `PUT /_matrix/client/v3/profile/@alice:matrix.org/displayname` with:

```json
{
  "displayname": "Alice Wonderland"
}
```

> *Note**: Setting a `null` value does not delete the key; the key remains with a `null` value.

Author: Is this roughly what you were looking for? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/9b4741e1f19b909173be1f263bbea6b3fa2ebe8c#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439L61-R66

- **Delete a Profile Field**
  - **Endpoint**: `DELETE /_matrix/client/v3/profile/{userId}/{key_name}`
  - **Description**: Remove a specified `key_name` (and its value) from the user's profile,
    if permitted by the homeserver.

- **Get All Profile Fields**

>   - **Endpoint**: `GET /_matrix/client/v3/profile/{userId}`

Author: Great shout, I've done this here: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/9b4741e1f19b909173be1f263bbea6b3fa2ebe8c#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R77-R78

  - **Request Body**:

```json
{
  "avatar_url": "mxc://matrix.org/MyNewAvatar",
  "displayname": "John Doe",
  "m.example_field": "new_value1"
}
```

- **Replace Profile Fields**
  - **Endpoint**: `PUT /_matrix/client/v3/profile/{userId}`
  - **Description**: Replace the entire user's profile with the provided JSON object,
    adding or updating keys and removing any absent ones, if permitted by the homeserver.
  - **Request Body**:

```json
{
  "avatar_url": "mxc://matrix.org/MyNewAvatar",
  "displayname": "John Doe",
  "m.example_field": "new_value1"
}
```

**Note**: Clients are encouraged to manipulate fields individually to avoid race conditions.
However, this method allows for bulk updates when needed (e.g. bots managing multiple accounts).

## Server-Server API Changes

The federation endpoint `GET /_matrix/federation/v1/query/profile` will mirror the client-server
API changes to ensure profile information is consistent between local and federated users.

There is no method to verify the history of global profile fields over federation, so this endpoint
MUST only accept requests for local users on the current homeserver, and homeservers MUST only

> request a profile from the homeserver specified in that user's MXID.

Author: I have done this here: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/9b4741e1f19b909173be1f263bbea6b3fa2ebe8c#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R291

  "error": "The profile data exceeds the maximum allowed size of 64KiB."
}
```

- **`M_KEY_TOO_LARGE`**: Exceeds individual key length limits.

```json
{
  "errcode": "M_KEY_TOO_LARGE",
  "error": "The key name exceeds the maximum allowed length of 128 bytes."
}
```

- **`M_TOO_MANY_KEYS`**: Exceeds key limits.

```json
{
  "errcode": "M_TOO_MANY_KEYS",
  "error": "The user has exceeded the maximum number of allowed keys in their profile."
}
```

### **403 Forbidden**: User lacks permission

**Note:** See [MSC4170](https://github.com/matrix-org/matrix-spec-proposals/pull/4170) for more
discussion on how server policy may result in 403 errors for profile requests.

### **404 Not Found**: Target cannot be found

- **`M_NOT_FOUND`**: Profile key does not exist.

```json
{
  "errcode": "M_NOT_FOUND",
  "error": "The requested profile key does not exist."
}
```

## Propagation of Profile Fields to Membership Events

The existing fields, `avatar_url` and `displayname`, will continue to trigger state events in each
room. These fields are replicated per-room via member events.

All other fields (unless a future proposal specifies otherwise) will **not** trigger state events

> in rooms and will exist solely at the global level for storing metadata about the user.

Author: Great shout! Is this suitable or were you expecting something more in-depth? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/9b4741e1f19b909173be1f263bbea6b3fa2ebe8c#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R264-R266

Each profile MUST be *at most* 64KiB (65,536 bytes) in size, as measured in Canonical JSON,
including the `avatar_url` and `displayname` fields.

Homeservers are not expected to enforce namespaces, as future expansions may be unknown to the
server, but clients are expected to use the correct namespace for field creation/updates.

Keys MUST follow the [Common Namespaced Identifier Grammar](https://spec.matrix.org/unstable/appendices/#common-namespaced-identifier-grammar),
with the following considerations:

- **Namespace `m.*`**: Reserved for fields explicitly defined in the Matrix specification. These
  fields may have special entry/display requirements, so clients that do not recognise a field in
  this namespace MAY attempt to display it, but SHOULD NOT attempt to update the content unless
  they understand its formatting and validation requirements.

- **Namespace `u.*`**: Reserved for future user-defined custom fields. A future proposal may
  justify a need for users to create custom keys within their client, so this namespace has been
  reserved for that purpose. Clients and servers SHOULD avoid using this namespace until specified

>   in a future proposal.

Reviewer: blocking: this should be moved out to a dedicated proposal. It's not part of the common namespaced identifier grammar, and is somewhat controversial as a feature.

Author: I've removed it from this one, thanks 👍

- **Namespace `tld.name.*`**: For client-specific or unstable fields, using Java package naming
  convention (e.g., `com.example.custom_field`).

> Following this change, for example, clients could use `m.example_field` if that field is defined by
> the Matrix specification, or `org.example.job_title` for organisation or client-specific fields.

Reviewer: ```suggestion
Following this change, for example, clients could use `m.example_field` if that field is defined by
the Matrix specification, or `org.example.job_title` for organisation, client-specific fields, and 
MSC-backed unstable features.
```

The proposal [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175) demonstrates
the process of defining new fields in the `m.*` namespace.

## Implementation Details

> - Custom fields MUST not trigger state events in rooms; their data MUST not be replicated to

Reviewer: ```suggestion
- Custom fields MUST NOT trigger state events in rooms; their data MUST NOT be replicated to
```

  `m.room.member` events unless a future proposal creates exceptions for specific fields.

- Servers SHOULD cache remote profiles for *at least* 5 minutes after retrieval; however, until
  a future proposal establishes a method for servers to notify each other of updates, it is
  *recommended* profiles be cached no longer than 15 minutes to avoid displaying stale data.
  Servers MUST not cache for longer than 24 hours to avoid unwanted data persistence.

- Clients COULD provide a UI for users to view and enter custom fields, respecting the appropriate
  namespaces.

- Clients SHOULD only display profiles of users in the current room whose membership status is

>   `join`, `invite`, or `knock`, and whose `m.room.member` event has *not* been redacted. If a

Author: I have adjusted this here: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/9b4741e1f19b909173be1f263bbea6b3fa2ebe8c#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439L279-R283

```json
{
  "displayname": "Alice Wonderland"
}
```

**Note**: As a `DELETE` endpoint exists to remove a key, setting a `null` value with the `PUT`
method SHOULD NOT delete the key but rather retain it with a `null` value. Servers MAY forbid a
`null` request if their policy does not allow keys to exist with a `null` value.

### Delete a Profile Field

- **Endpoint**: `DELETE /_matrix/client/v3/profile/{userId}/{key_name}`
- **Description**: Remove a specified `key_name` (and its value) from the user's profile,
  if permitted by the homeserver.

### Get All Profile Fields

- **Endpoint**: `GET /_matrix/client/v3/profile/{userId}`
- **Description**: Retrieve all profile fields for a user, identical to the
  [current API](https://spec.matrix.org/latest/client-server-api/#get_matrixclientv3profileuserid).
- **Response**:

```json
{
  "avatar_url": "mxc://matrix.org/MyC00lAvatar",
  "displayname": "John Doe",
  "m.example_field": "value1",
  "org.example.job_title": "Software Engineer"
}
```

### Partially Update Profile Fields

- **Endpoint**: `PATCH /_matrix/client/v3/profile/{userId}`
- **Description**: Merge the provided JSON object into the user's profile, updating any

>   specified keys without altering others, if permitted by the homeserver.

Author: Good point. I'd assumed it would overwrite the top-level key, but I guess that also invites the same race conditions we've been trying to avoid with the single-key operations.

One worth asking the SCT, perhaps? Is merging trees behaviour we want to protect in Matrix, or should we treat each key as an insular piece of data that should always be overwritten entirely?

>   specified keys without altering others, if permitted by the homeserver.

Author: Good point. I'd assumed it would overwrite the top-level key, but I guess that also invites the same race conditions we've been trying to avoid with the single-key operations.

One worth asking the SCT, perhaps? Is merging trees behaviour we want to protect in Matrix, or should we treat each key as an insular piece of data that should always be overwritten entirely?

>   specified keys without altering others, if permitted by the homeserver.

Reviewer 9: I didn't get much traction from the SCT about which way to go.

@tulir Do you know what hungryserv does? The Synapse PR does just a top-level merge.

I think ways forward are to:

1. Just choose one and see what people think.
2. Move `PATCH` into a separate MSC.

>   specified keys without altering others, if permitted by the homeserver.

Reviewer 11: Looks like hungry does a top-level merge too

>   specified keys without altering others, if permitted by the homeserver.

Reviewer 20: I'm not sure I see the benefit of `PATCH` here, whatever merging method we use is going to cause confusion. I think the easiest thing to do is to remove support for this and just have clients either re-upload or set individual keys.

>   specified keys without altering others, if permitted by the homeserver.

Reviewer 9: @tcpipuk To be clear -- it would be best to move `PATCH` support to a separate MSC to avoid blocking the core of this MSC.

>   specified keys without altering others, if permitted by the homeserver.

Reviewer 9: This has now been done.

> If no `field` was specified, the response should include the fields of the user's profile that
> can be made public, such as the display name and avatar.

Given this wording, homeservers currently have the flexibility to decide whether some fields are
published over federation, and this proposal continues to allow this behaviour.

## Capabilities

A new capability `m.profile_fields` controls the ability to *set* profile fields and is advertised
on the `GET /_matrix/client/v3/capabilities` endpoint.

This capability deprecates the use of `m.set_displayname` and `m.set_avatar_url`, which are not
required when this capability is present.

Clients SHOULD check for this capability before attempting to create or modify a profile field.

### Capability Structure

```json
{
  "capabilities": {
    "m.profile_fields": {
      "enabled": true,
      "allowed": ["m.example_field", "org.example.job_title"],
      "disallowed": ["org.example.secret_field"]
    }
  }
}
```

### Behaviour

- **When capability is missing**: Clients SHOULD assume extended profiles are supported and that
  they can be created or modified. If a server intends to deny some (or all) changes, it SHOULD use

>   the capability to advertise this, improving the client experience.

Reviewer 9: I don't think this is backwards compatible -- if it is missing shouldn't it be assumed to be not supported?

Author: That was primarily off the back of this thread: https://github.com/matrix-org/matrix-spec-proposals/pull/4133#discussion_r1760671507

Reviewer 9: Maybe this just needs to state that when the capability is missing, but the homeserver's supported spec versions supports custom profile fields?

Reviewer 4: I came here to make this exact same question; wondering if this got marked "Resolved" accidentally, because AFAICT the current wording has the same problem.

Reviewer 4: To be clear, I think this needs wording along the lines of the following:

```suggestion
- **When capability is missing**: Clients SHOULD assume extended profiles are supported and that
  they can be created or modified, provided the response from [`/versions`](https://spec.matrix.org/v1.13/client-server-api/#get_matrixclientversions)
  indicates support for a spec version that includes this proposal.
  If a server intends to deny some (or all) changes, it SHOULD use
  the capability to advertise this, improving the client experience.
```

Keys MUST follow the [Common Namespaced Identifier Grammar](https://spec.matrix.org/unstable/appendices/#common-namespaced-identifier-grammar),
with the following considerations:

- **Namespace `m.*`**: Reserved for fields explicitly defined in the Matrix specification. Clients
  that do not recognise a field in this namespace MAY attempt to display it but SHOULD NOT attempt
  to update the content unless they understand its formatting and validation requirements.

- **Namespace `tld.name.*`**: For client-specific or unstable fields, using Java package naming
  convention (e.g., `com.example.custom_field`).

Following this change, clients could use `m.example_field` if that field is defined by the Matrix
specification, or `org.example.job_title` for organisation, client-specific fields, or MSC-backed
unstable features. Proposal [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175)
demonstrates the process of defining new fields in the `m.*` namespace.

## Error Handling

### 400 Bad Request: Request Exceeds Limits or Is Malformed

- **`M_BAD_JSON`**: Malformed request.

```json
{
  "errcode": "M_BAD_JSON",
  "error": "The provided JSON is malformed."
}
```

> - **`M_PROFILE_TOO_LARGE`**: Exceeds total profile size limits.

Reviewer 9: Why not just `M_TOO_LARGE` which already exists?

Author: That was because of this thread: https://github.com/matrix-org/matrix-spec-proposals/pull/4133#discussion_r1784323135

Now we've lowered the amount of possible errors, I could switch to using the errors you've proposed, it's just at the time Dave suggested I use clear errors to differentiate which thing was "too large".

```json
{
  "errcode": "M_PROFILE_TOO_LARGE",
  "error": "The profile data exceeds the maximum allowed size of 64 KiB."
}
```

> - **`M_KEY_TOO_LARGE`**: Exceeds individual key length limits.

Reviewer 9: Why not `M_INVALID_PARAM`?

Author: The error should probably already be used if someone tries to do something like use an invalid type for the key name:

> `M_INVALID_PARAM` A parameter that was specified has the wrong value. For example, the server expected an integer and instead received a string.

We could just use this to reject _all_ unacceptable keys though, what do you think? 🤔

Reviewer 9: I was thinking we would reject all unacceptable keys with `M_INVALID_PARAM`.

```json
{
  "errcode": "M_KEY_TOO_LARGE",
  "error": "The key name exceeds the maximum allowed length of 128 bytes."
}

> ```

Reviewer 9: Is the key length 128 bytes or is this meant to server dependent or something else? It is unclear if this is an example or a requirement.

Author: Sorry, I thought I'd updated everywhere to say we're following the common grammar length of 255, I'll fix that now.

- **`M_TOO_MANY_KEYS`**: Exceeds key limits.

```json
{
  "errcode": "M_TOO_MANY_KEYS",
  "error": "The user has exceeded the maximum number of allowed keys in their profile."
}

> ```

Reviewer 9: Same here -- what is the maximum number of keys? If we have the 64 KiB limit why do we have this also?

Author: I _think_ this was an example, so that a homeserver may optionally limit the number of keys a user may create, so clients know to recognise this specific type of situation when it crops up. Should I add explanation for its existence, or just remove it? 🤔

Reviewer 9: I think the size of a key and the total size is probably enough limitations? 🤷

{
  "errcode": "M_NOT_FOUND",
  "error": "The requested profile key does not exist."
}
```

## Propagation of Profile Fields

The existing fields, `avatar_url` and `displayname`, will continue to trigger state events in each
room. These fields are replicated per-room via member events.

All other fields (unless a future proposal specifies otherwise) WILL NOT trigger state events in
rooms and will exist solely at the global level for storing metadata about the user.

Clients SHOULD consider the increased traffic implications when displaying values (e.g. timezones)
outside of the profile. Servers MAY wish to optimise and relax rate limits on these endpoints in
consideration of this.

## Implementation Details

- Custom fields MUST NOT trigger state events in rooms; their data MUST NOT be replicated to
  `m.room.member` events unless a future proposal creates exceptions for specific fields.

- Servers SHOULD cache remote profiles for at least 5 minutes after retrieval. Until a method for
  servers to notify each other of updates is established, it is recommended that profiles be cached
  no longer than an hour to avoid displaying stale data. Servers SHOULD NOT cache profiles for

>   longer than 24 hours to avoid unwanted data persistence.

Reviewer 20: I'm not sure this is really needed if this is triggered by a user-driven API requests? Generally we're OK with doing a federation request in that case, as federation requests aren't that expensive. The issue is when servers end up doing automated tasks that end up doings lots of requests.

The downside with recommending caching is that it makes the UX materially worse over federation

Author: That's good to know. I do think there's some room for optimisation (e.g. if someone joins a room with 10,000 users in it, we don't really want their server asked for timezone information 10,000 times) but we also don't want stale information.

My preference would be to use an EDU to signal when a profile's been updated, so servers/clients can hold onto profile information rather than needing to repeatedly call it. I've avoided including that in this MSC, but this is the rationale behind including caching because it'll likely be encouraged in a future one.

Perhaps I should switch it to a "may" so it's up to servers whether they cache now (and admins can decide how much to compromise UX for performance) but then it's a simple wording change later?

Reviewer 3: I wonder if there could be a separate event type that is only applied server<->server to drop profile caches? (or even better, sending over the new data to save on overhead?)

Author: I do like the idea of an EDU signalling that a profile has been updated, yes. That's not part of this MSC, but it's definitely on my list of things that could be added in a future MSC.

Reviewer: (please add this consideration to the MSC itself)

Author: The consideration does exist in the Potential Issues section, but I've added some to Implementation Details and Potential Issues in case this is what you're looking for? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/43e1e1a13f8bbd3a58373e0fe8f3a93029ea9914

Reviewer: I think there's two conclusions here: the EDU/updates over federation problem, and caching leading to worse experience. The first appears to be well documented (thank you), though I feel like the caching concern may be best resolved with changing the recommendation to "Servers which prefer to cache details should do so for a short period of time to avoid stale data being presented to users". ie: no specific commitments, and an implied recommendation to not do caching in the general case (but still support it).

Reviewer: argh, the caching did make it to the diff, but github was showing me an old copy of the MSC, sorry. 

Considering resolved for now. Threads on the new diff are preferred.

Author: Agreed, new thread would be really helpful for any new changes, but I've tried to reflect this thread's concluded changes here: https://github.com/matrix-org/matrix-spec-proposals/commit/d620259ac07cb38d43fdbb2880ba47a452ed6522

[MSC4202](https://github.com/matrix-org/matrix-spec-proposals/pull/4202) offers a facility for
users to report offensive content to the homeserver that account belongs to. This proposal is not
dependent on MSC4202 but encourages the use of moderation options to allow users to report
offensive content.

## Security Considerations

Since profile fields are public, the server is not directly responsible for the privacy of the data;
however, clients SHOULD make users aware that any information published in their profile will be
visible to others on the federated network.

Likewise, if a server automatically publishes data in user profile fields (e.g. setting a job title
based on an organisation's internal user database), then they SHOULD have consent to do so, and
users SHOULD be made aware that data is published on their behalf.

To minimise the impact of abuse, clients SHOULD offer suitable defaults for displaying user-entered
content. A user MAY choose to display fields from all users globally, but *by default* profiles
SHOULD only be shown when the users share the current room and the other user is in the `join`,

> `invite`, or `knock` membership states.

Reviewer 20: I'm not sure about `knock` here, given for knockable rooms this can be done by any random Matrix user and so there is no extra "trust" gleaned from the fact you "share" a room?

Author: My thinking was that you'd want to see the fields to help decide if you should allow them to enter.

Reviewer: honestly, I'm not sure this should be specified. It can be included as a reminder in the spec that profiles may contain abusive content, but the recommendation portion doesn't really fit the spec's purpose. Clients have their own safety considerations, and being made aware of the potential danger may cause them to adjust rendering to their risk profile.

Author: I'm not 100% sure exactly what the conclusion is asking for, but have had an attempt at rewriting the section: https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/af87bbeb26c6d33725d970676009b52056a87f96

Reviewer: I think the new text captures my concern. @erikjohnston - if your concern isn't addressed, a thread on the new text is probably best.

This proposal does not directly address reporting of user profiles over federation, but
[MSC4202](https://github.com/matrix-org/matrix-spec-proposals/pull/4202) offers a facility for
users to report offensive content to the homeserver that account belongs to. This proposal is not
dependent on MSC4202 but encourages the use of moderation options to allow users to report
offensive content.

## Security Considerations

Since profile fields are public, the server is not directly responsible for the privacy of the data;
however, clients SHOULD make users aware that any information published in their profile will be
visible to others on the federated network.

Likewise, if a server automatically publishes data in user profile fields (e.g. setting a job title
based on an organisation's internal user database), then they SHOULD have consent to do so, and
users SHOULD be made aware that data is published on their behalf.

To minimise the impact of abuse, clients should carefully consider when and how to display
user-entered profile content. While some clients may choose to show profile fields globally, others

> may restrict visibility based on room membership or other trust signals. Clients should be aware
> that profile fields may contain abusive content and implement appropriate safety measures based on
> their risk assessment.

Reviewer 14: It would be helpful to add a client recommendation here to nudge the ecosystem towards consistent behaviour. Consider the scenario where T&S or room moderators are frantically trying to combat a user putting illegal content in their profile, which may then be shown to users.

If there isn't a (mostly) consistent behaviour between clients for when they should/shouldn't show a user's profile field, moderators will be left playing whack-a-mole when advising their users on what to do to hide the content.

```suggestion
may restrict visibility based on room membership or other trust signals. Clients should be aware
that profile fields may contain abusive content and implement appropriate safety measures based on
their risk assessment. For example, a client could hide a user's custom profile fields in the context of
a room if the user in question's latest `m.room.member` state event has been redacted. This gives
room moderators the power to quickly hide abusive content in profile fields from other users.
```

The hope is that by suggesting this behaviour, the majority of clients will implement it, and thus T&S teams can begin to rely on it.

Some additional notes based on [a long discussion](https://matrix.to/#/!NasysSDfxKxZBzJJoE:matrix.org/$_ASJuqvv3iKwtT0KEmONhKFHrM4uT-4GHCd8WZDkLH4?via=matrix.org&via=envs.net&via=element.io) in the `#matrix-spec` room:

* Room moderators [already rely on](https://matrix.to/#/!NasysSDfxKxZBzJJoE:matrix.org/$Kcl5uDrpT3WtNqgQVxDjHudInlOcvLl9_-J1qKYkFsA?via=matrix.org&via=envs.net&via=element.io) redacting the membership event to hide abusive content in a user's profile (displayname/avatar).
* A kick/ban could also solve this problem, though most clients currently don't display the reason why you were kicked/banned from a room. This leads to frustration for when a user doesn't know why they were kicked (it's fairly subtle that one's profile is against the rules of a room).
* The proposal initially had some recommendations for clients on when to display custom user profile fields, but this was removed in [a previous thread](https://github.com/matrix-org/matrix-spec-proposals/pull/4133#discussion_r1906162637
) due to keeping client recommendations out of the spec. However, I argue that including them is beneficial to achieving consistent behaviour across the client ecosystem.
* State events that are redacted should not revert to the previous state event - instead they should continue to affect state, but with the redacted fields removed. See the "info" section [the redactions section of the Client-Server spec](https://spec.matrix.org/v1.13/client-server-api/#redactions).
* Clients must be able to tell whether an event is redacted in order to implement this. An [earlier thread](https://github.com/matrix-org/matrix-spec-proposals/pull/4133#discussion_r1803503479) mentioned that there were edge cases where this wasn't possible, but without an example. @tulir [mentioned on Matrix](https://matrix.to/#/!NasysSDfxKxZBzJJoE:matrix.org/$R2ujiFb1nNwPEdqsAaUdsAIvjFxgdZuSIFWu-3EawA0?via=matrix.org&via=envs.net&via=element.io) that there are cases where you receive an event at one point, then the redaction some time later. I consider this just a fact of life in Matrix, and not a blocker to this suggestion.
* The proposed change above is effectively a blunt tool that would hide ALL custom profile fields, including ones that couldn't possibly include harmful content, such as `m.tz` defined in [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175). MSCs like MSC4175 could alter the client recommendation to opt-out of being hidden in this case.
* An alternative could be defining a new state event to signal to clients that a user's profile should be hidden. But this would require new training for moderators, and add quite a bit of additional complexity to this MSC.

Author: Thank you for this. It's quite a tight balance to walk between under/over stating the requirements for clients, so I appreciate the guidance here.

method SHOULD NOT delete the key but rather retain it with a `null` value. Servers MAY forbid a
`null` request if their policy does not allow keys to exist with a `null` value.

### Delete a Profile Field

- **Endpoint**: `DELETE /_matrix/client/v3/profile/{userId}/{key_name}`
- **Description**: Remove a specified `key_name` (and its value) from the user's profile,
  if permitted by the homeserver.

### Get All Profile Fields

- **Endpoint**: `GET /_matrix/client/v3/profile/{userId}`
- **Description**: Retrieve all profile fields for a user, identical to the
  [current API](https://spec.matrix.org/latest/client-server-api/#get_matrixclientv3profileuserid).
- **Pagination**: Not applicable, the full profile is bounded at 64 KiB total size
- **Response**:

```json
{
  "avatar_url": "mxc://matrix.org/MyC00lAvatar",
  "displayname": "John Doe",
  "m.example_field": "value1",
  "org.example.job_title": "Software Engineer"
}
```

## Server-Server API Changes

The federation endpoint [`GET /_matrix/federation/v1/query/profile](https://spec.matrix.org/v1.13/server-server-api/#get_matrixfederationv1queryprofile)` will mirror the client-server

> API changes to ensure profile information is consistent between local and federated users.

Reviewer 7: This shouldn't block or even make it into this MSC, but I've thought about one of my use-cases and it certainly doesn't want the profile to land on every homeserver but only some of them. 

I think I'd like a future MSC to describe this mirroring as up to the implementation.

Author: Unless I've misread your meaning here, that isn't what "mirror" means in this case - it simply means the _endpoint_ changes are mirrored across both client and federation APIs, but it's up to the homeserver implementation to decide which fields it allows other servers to see when they request a profile.

Reviewer 7: I suppose I didn't read far enough along, although it did confuse me a bit.

>will mirror the client-server API changes to ensure profile information is consistent between local and federated users.

and

>Given this wording, homeservers currently have the flexibility to decide whether some fields are published over federation, and this proposal continues to allow this behaviour.

feel slightly contradictory to me. However, I'm happy to accept that the latter statement takes precedent. I think when the spec comes to be up we should make this very clear.

Author: I'll try to find time today to have a look at that wording to make it a bit more explicit 🙂

Author: Something like this? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/7ca83db96dd93646a3ba56c00ed6b9493023b766 🙂

Reviewer 7: smashing, thanks :+1:

Client-Server and Server-Server endpoints introduced by this MSC. For example:

1. `M_NOT_FOUND` applies to any attempt to retrieve a non-existent profile field.
2. `M_PROFILE_TOO_LARGE` applies to any attempt to create or update profile data exceeding the
   allowed size.

The Server-Server endpoints introduced in this MSC adhere to the existing error structure for
federation, as the federation access remains read-only in this proposal. This means no new error
codes or status code combinations are introduced for Server-Server endpoints beyond what is already
documented in the specification.

## Propagation of Profile Fields

The existing fields, `avatar_url` and `displayname`, will continue to trigger state events in each
room. These fields are replicated per-room via member events.

All other fields (unless a future proposal specifies otherwise) WILL NOT trigger state events in
rooms and will exist solely at the global level for storing metadata about the user.

Clients SHOULD consider the increased traffic implications when displaying values (e.g. timezones)
outside of the profile. Servers MAY wish to optimise and relax rate limits on these endpoints in
consideration of this.

> ## Implementation Details

Reviewer 5: 1. Why is this MSC being considered in isolation of any specified
   custom profile field?
  - All fields contain user generated content.
    - No custom profile field has sufficiently demonstrated itself (via
    the spec process or otherwise).
  - There are no fields defined in this MSC
  - All fields used in the implementation are currently under review
     in other MSCs and have not demonstrated themselves.

2. The MSC does not propose anything in itself except a place to store
   user generated content... without demonstrating consideration for
   user generated content, especially free-text content.
  - Free-text content was originally part of the MSC, but got moved
  out to another proposal
  ([MSC4208](https://github.com/matrix-org/matrix-spec-proposals/pull/4208))
  to "resolve" numerous blocking concerns about it.  But these
  concerns will appear in almost every single follow-up MSC related to
  specifying any field that builds on this proposal.
  - The MSC still has a provision to display and enter custom fields:
    > Clients MAY provide a UI for users to view and enter custom
    > fields, respecting the appropriate namespaces.

  - The demonstrated implementation uses
   [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175)
   which is a parsable field with a very narrow range of values. But
   this proposal is clearly intended to be followed up with fields for
   unparsable user generated content. And we expect there to be many
   more unparsable fields than restricted and narrowly defined ones.

3. Why is there urgency to merge this MSC?
  - Clients can currently use the proposal to implement custom fields
     under the unstable namespace without any changes or hindrance.
     Some of them already have done, so there should be no urgency to
     get this merged.
  - Similar Concerns have been suppressed in Matrix rooms as
    bikeshedding or blocking the MSC. All developers are free to
    implement the proposal as is under the unstable flag. When the
    propsal becomes part of the specification it does not change the
    number of implementations, nor does it allow dependant MSCs to
    make any further progress.

I believe the MSC should be modified to address the concerns with
free-text content. And also demonstrate implementation within a client
that does add free-text fields.

The discussion will have to happen with subsequent MSCs unless they
define fields that are both very limited and restricted, which is
obviously not the intent of this proposal. So it does not make sense
to discuss this proposal in isolation of those concerns.

Reviewer 4: I have split this comment into three threads so we can address them separately.

extending the profile API to include extra fields, enabling users and servers to publish key:value
pairs to their global profiles. This extension provides a flexible framework for users to share
additional public information.

The proposal is designed to be straightforward and compatible with existing clients and servers,
requiring minimal changes to facilitate quick adoption. It complements, rather than replaces,
[MSC1769](https://github.com/matrix-org/matrix-spec-proposals/pull/1769) (Extensible Profiles as
Rooms) by focusing on global profile data without the complexity of per-room profile management.

## Authentication and Rate Limiting

All endpoints in this proposal follow the standard client-server API authentication rules. Specifically:

- All endpoints require authentication except for GET requests which may be accessed without
  authentication
- Servers MUST verify the access token has permission to modify the requested userId's profile
- Rate limiting SHOULD be applied as per the homeserver's normal profile endpoint limits
- Guest access follows the same rules as existing profile endpoints - guests may view profiles but

>   not modify them

Reviewer 9: This isn't fully correct -- guests can set only their display name. I'm assuming this implies that they can still modify their display name (Which for this MSC implies set & now delete), while other fields *cannot* be modified.

Author: Agreed, the intention was "We don't change what Guests are allowed to do/access" 🙂

A new capability `m.profile_fields` controls the ability to *set* profile fields and is advertised
on the `GET /_matrix/client/v3/capabilities` endpoint.

This capability deprecates the use of `m.set_displayname` and `m.set_avatar_url`, which are not
required when this capability is present.

Clients MAY check for this capability before attempting to create or modify a profile field.

### Capability Structure

```json
{
  "capabilities": {
    "m.profile_fields": {
      "enabled": true,
      "allowed": ["m.example_field", "org.example.job_title"],
      "disallowed": ["org.example.secret_field"]
    }
  }
}
```

### Behaviour

- **When capability is missing**: Clients SHOULD assume extended profiles are supported and that
  they can be created or modified, provided the response from [`/versions`](https://spec.matrix.org/v1.13/client-server-api/#get_matrixclientversions)
  indicates support for a spec version that includes this proposal. If a server intends to deny some
  (or all) changes, it SHOULD use the capability to advertise this, improving the client experience.

- **When `enabled` is `false`**: Clients SHOULD expect to display profiles but NOT create or update

>   fields. Any attempt to do so SHOULD result in a `403 Forbidden` error. This does not affect
>   `avatar_url` and `displayname` fields, which are allowed for compatibility purposes.

Reviewer 4: Sounds like we're all agreed that we should remove the second sentence here.

```suggestion
  fields. Any attempt to do so SHOULD result in a `403 Forbidden` error.
```

Perhaps someone could open a PR to make that change to the proposal doc?

> # MSC4133: Extending User Profile API with Custom Key:Value Pairs

Reviewer 18: Overall, I like it.

It was always the intention to allow more keys than just display name / avatar URL, and this proposal considers the impacts of opening up the k:v store more generally: in particular, strong size limits and the ability to PATCH.

Author: Thank you, positive feedback means a lot!

I hoped `PATCH` wasn't assuming too much since [MSC4138](https://github.com/matrix-org/matrix-spec-proposals/pull/4138) was merged - I'm excited to see what bots (especially AppServices) will be able to do with profiles in the future 🙂

*This proposal aims to enhance user profiles within the Matrix ecosystem by introducing
customisable key:value pairs to global profiles. Allowing users to add arbitrary public information
to their profiles (such as preferred languages, organisational roles, or other relevant public
details) should enrich user interactions without impacting existing functionality.*

## Proposal

Currently, the Matrix protocol supports limited user profile fields: `avatar_url` and
`displayname`. This proposal extends the API to include custom fields, enabling users to add
free-text key:value pairs to their global profiles. This extension provides a flexible framework

> for users to share additional public information.

Author: Thanks for the review!

One or two people have mentioned I could've split this into two MSCs, but I was wary of presenting an MSC that didn't technically have a problem to solve (i.e. adding endpoints to set extra keys without having keys to set).

In the short term, following Half-Shot's advice, I've downgraded the UI to a `COULD` so clients aren't compelled to implement it: https://github.com/tcpipuk/matrix-spec-proposals/commit/4afb8b84e2879dbd85affaeb65e468eec9a3710d#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R278

I could go further and remove some further client references so the namespace is reserved but there's less "this is what your client needs to do" but I'd appreciate your perspective on this?

> for users to share additional public information.

Reviewer 21: My worry with a 'COULD' is that essentially it's saying that some clients will display the fields and others won't, which means more fragmentation. Thinking on it a bit more, user defined tags feel kind of okay for room tags where only the user themself will ever see them. It feels a lot more of a challenge for something that other user will see. I'd support splitting user defined tags out into a separate MSC, maybe even as something where we can try the regular MSC process for new profile fields and then come back to if we still feel that it's necessary.

> for users to share additional public information.

Reviewer 9: From discussions with the SCT there seemed to be general consensus that this proposal would benefit from splitting out the user-defined fields, as that seems to be the main point of contention.

Although it is a bit more abstract, I think it is still obvious to see the value in allowing for additional (and custom) profile fields that other MSCs can define. Eg as MSC4175 already does. Alternately a single concrete "custom field" could be added to this proposal, but that's more area to bike shed then.

As a process point, if this is done we'll cancel proposed FCP and restart it later due to major changes made to the MSC.

(I had also previously suggested splitting out some of the endpoint changes to a separate MSC but the SCT generally thought that was unnecessary.)

> for users to share additional public information.

Author: I should have another MSC raised for custom fields this evening, I'm just reviewing what actually wants removing from this one. I think it's actually a (fractionally) small change to remove the custom fields as all of the endpoints (and size/key requirements) stay the same.

Happy to cancel proposed FCP if you'd prefer, though also happy if you want to wait and see what the changes look like later today?

> for users to share additional public information.

Author: I believe this should cover the needed changes - what do you think? https://github.com/matrix-org/matrix-spec-proposals/pull/4133/commits/bb4fc7679c50577662169d6dea5e70848a37ea68

> for users to share additional public information.

Reviewer 7: Looks good to me :+1:

## Client-Server API Changes

- **Get a Profile Field**
  - **Endpoint**: `GET /_matrix/client/v3/profile/{userId}/{key_name}`
  - **Description**: Retrieve the value of a specified `key_name` from a user's profile.
  - **Response**:

```json
{
  "key_name": "field_value"
}
```

*Example*: Requesting `GET /_matrix/client/v3/profile/@alice:matrix.org/displayname` returns:

```json
{
  "displayname": "Alice"
}
```

- **Set a Profile Field**
  - **Endpoint**: `PUT /_matrix/client/v3/profile/{userId}/{key_name}`
  - **Description**: Set or update the value of a specified `key_name` in the user's profile.
  - **Request Body**:

```json
{
  "key_name": "new_value"
}
```

*Example*: Setting `PUT /_matrix/client/v3/profile/@alice:matrix.org/displayname` with:

```json
{
  "displayname": "Alice Wonderland"
}
```

*Note**: Setting a `null` value does not delete the key; the key remains with a `null` value.

- **Delete a Profile Field**
  - **Endpoint**: `DELETE /_matrix/client/v3/profile/{userId}/{key_name}`
  - **Description**: Remove a specified `key_name` (and its value) from the user's profile,

>     if permitted by the homeserver.

Reviewer 21: "if permitted by the homeserver" should probably be on all the write endpoints too?

Author: Thanks! I've done this now: https://github.com/tcpipuk/matrix-spec-proposals/commit/4afb8b84e2879dbd85affaeb65e468eec9a3710d#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439L47-R102

  - **When `enabled` is `true`**: Clients should allow users to create or update custom fields,
    except those listed in the `disallowed` array, if it exists. Servers may optionally specify the
    `allowed` array to allowlist fields that users may set - if the `allowed` key is provided with
    empty content (e.g. `"allowed": []`) this also disallows the setting of any fields. Clients
    will receive `400 Bad Request` or `403 Forbidden` responses from the homeserver if server-side
    policies prevent them.

## Error Handling

Consistent error codes and messages ensure clear communication of issues:

### **400 Bad Request**: Request exceeds limits or is malformed

- **`M_BAD_JSON`**: Malformed request.

```json
{
  "errcode": "M_BAD_JSON",
  "error": "The provided JSON is malformed."
}
```

- **`M_TOO_LARGE`**: Exceeds size limits.

```json
{
  "errcode": "M_TOO_LARGE",
  "error": "The profile data exceeds the maximum allowed size of 64KiB."
}
```

or

```json
{
  "errcode": "M_TOO_LARGE",
  "error": "The key name exceeds the maximum allowed length of 128 bytes."
}
```

or

```json
{

>   "errcode": "M_TOO_LARGE",

Reviewer 21: You've used the same code for three different errors here: the client won't know how to tell the user to fix it. I think they just need to be different, there's no need to feel like you have to stick to the common codes already defined.

Author: Oh cool, I didn't know that was an option! I've proposed these: https://github.com/tcpipuk/matrix-spec-proposals/commit/4afb8b84e2879dbd85affaeb65e468eec9a3710d#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439L184-R209

Or should I keep `M_TOO_LARGE` for the overall profile size to lower the amount of new error codes?

Reviewer 21: I think I like what you've done personally, having M_TOO_LARGE for one where it could apply to either feels unnecessarily confusing. Thanks!

  "error": "The user has exceeded the maximum number of allowed keys in their profile."
}
```

### **403 Forbidden**: User lacks permission

**Note:** See [MSC4170](https://github.com/matrix-org/matrix-spec-proposals/pull/4170) for more
discussion on how server policy may result in 403 errors for profile requests.

### **404 Not Found**: Target cannot be found

- **`M_NOT_FOUND`**: Profile key does not exist.

```json
{
  "errcode": "M_NOT_FOUND",
  "error": "The requested profile key does not exist."
}
```

## Propagation of Profile Fields to Membership Events

The existing fields, `avatar_url` and `displayname`, will continue to trigger state events in each
room. These fields are replicated per-room via member events.

All other fields (unless a future proposal specifies otherwise) will **not** trigger state events
in rooms and will exist solely at the global level for storing metadata about the user.

## Key/Namespace Requirements for Custom Fields

Homeservers are not expected to enforce these namespaces, as future expansions may be unknown to
the server, but clients are expected to use the correct namespace for field creation/updates.

> The *name* of each key must be a valid UTF-8 string of between one and 128 bytes.

Reviewer 21: Please could this be defined as the common identifier grammar: https://spec.matrix.org/unstable/appendices/#common-namespaced-identifier-grammar

Author: I'd never seen the CNIG section before, thanks for bringing that to my attention!

I've stripped out a chunk of my own custom definitions and referenced the common grammar instead: https://github.com/tcpipuk/matrix-spec-proposals/commit/4afb8b84e2879dbd85affaeb65e468eec9a3710d#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439L235-R266

Does that look ok, or could this be better?

Reviewer 21: Yes, they are not super obvious. That looks great, thanks!

  the `u.` defines the display name of this field (e.g. `u.Bio`). The values in this namespace must
  always be UTF-8 strings with a content not exceeding 512 bytes.

- **Namespace `tld.name.*`**: Client-specific or unstable fields use Java package naming convention.

Following this change, for example, a user could enter a "My Timezone" field manually in their
client and the client would be expected to add a `u.My Timezone` key in their profile. Clients
would be expected to treat this as a string inside the profile with no special meaning. However,
following [MSC4175](https://github.com/matrix-org/matrix-spec-proposals/pull/4175), clients would
use the unstable key `us.cloke.msc4175.tz` (or stable key `m.tz`) with the validation and
formatting required by that MSC.

## Size Limits

Whenever "bytes" are referred to as a limit, this is calculated as UTF-8 bytes, so a two-byte code
point consumes two bytes of this limit. This size would be measured before any JSON encoding.

Until a future MSC specifies otherwise:

- Each profile *must* be *at most* 64KiB (65,536 bytes) in size, as measured in Canonical JSON.

> - Each key *must* be a string of *at least* one byte and it's name *must* not exceed 128 bytes.

Reviewer 21: As above for common identifier format.

Author: As above, I've stripped out this section entirely now, thank you: https://github.com/tcpipuk/matrix-spec-proposals/commit/4afb8b84e2879dbd85affaeb65e468eec9a3710d#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439L235-R266

- Each key's *value* in the `u.*` namespace *must* not exceed 512 bytes in length.
- The limit on overall profile size includes `avatar_url` and `displayname`.

## Implementation Details

- The profile data will be public by default, and compliance with GDPR and other privacy

>   regulations will be enforced, particularly in terms of data deletion and retention policies.

Reviewer 21: I'm not really sure what this sentence means, especially by, "will be". Is it requiring that servers must adhere to the GDPR? It feels to vague to be in the spec. A specific recommendation about making sure deleted fields are properly deleted after some time could maybe find a place?

Author: Fair point, I've already made recommendations regarding retention/deletion elsewhere, so will remove this 👍

Author: I agree, I've just removed it entirely because it doesn't seem to add any value: https://github.com/tcpipuk/matrix-spec-proposals/commit/4afb8b84e2879dbd85affaeb65e468eec9a3710d#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439L273-L274

> - Custom fields will not trigger state events in rooms, maintaining account-wide metadata without

Reviewer 7: Nitpick, but to me this clears up the specific request of m.room.member events (although the text is broadly right about state in general).

```suggestion
- Custom fields will not trigger state events in rooms (such as `m.room.member`), maintaining account-wide metadata without
```

Reviewer 7: Albeit, this isn't an implementation detail -> You've already mentioned this in **Propagation of Profile Fields to Membership Events**

Author: Thanks! I've been going through a number of great reviews this afternoon, and implemented this as part of it: https://github.com/tcpipuk/matrix-spec-proposals/commit/4afb8b84e2879dbd85affaeb65e468eec9a3710d#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R270-R271

  creating state events or other moderation issues.

- Servers should cache remote profiles but implement strategies to minimise unintended data

>   persistence (e.g. expire caches within 24 hours).

Reviewer 7: I would be careful here. The timezone MSC calls out a usecase where travelling might provoke several timezone updates, so client implementations should be careful that not all fields will have equal TTLs. 

I think for fields the client doesn't know about, caching sounds reasonable.

Author: The main purpose of this requirement was to set a hard upper-limit, so no one is trying to cache profiles for days.

You're right that I should probably suggest this be a shorter upper-limit for usability, though I doubt someone will move through *many* timezones in one hour unless he's delivering presents on 24th Dec!

I think the actual caching strategy probably wants to be left up to the implementation, but I could (for example) suggest servers and clients cache for 5 minutes by default, maybe suggesting clients could offer users the option to cache longer?

I'm wary that 99% of the time the data won't have changed, but we're pending a good follow-up MSC to decide how servers would notify each other of changes to allow this caching period to be much longer.

Reviewer 7: Yes, I agree that caching is an impl detail. I just don't want people to assume that 24 hours is the lower bound :)

Author: I've written a (hopefully) better guide that suggests a minimum of 5 minutes and recommends servers use 15 minutes until a future MSC introduces a method for servers to proactively notify each other that a profile has been updated: https://github.com/tcpipuk/matrix-spec-proposals/commit/4afb8b84e2879dbd85affaeb65e468eec9a3710d#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R273-R276

- Clients are encouraged to provide settings for users to choose the scope of users they present
  the freetext (i.e. `u.*` namespaced) fields from (e.g. none, only users on the local server,
  users in the current room, any users sharing a room with this user, all users globally) with
  defaults sensitive to their implementation's UI and intended audience. This proposal *recommends*
  clients only display profiles of users in the current room whose membership status is `join`,
  `invite`, or `knock`, and whose `m.room.member` event has *not* been redacted. If a client offers
  an option for custom fields to always be available in the UI, an option should be provided to
  hide/minimise them automatically. These recommendations are to moderation concerns associated
  with displaying information from banned or departed users.

- Clients should provide a UI for users to enter their own free-text custom fields in the `u.*`

>   namespace of their own profile.

Reviewer 7: Nitpick, but I'd actually make this a `COULD`. I wouldn't go so far as saying clients would need to support this behavior to acceptably implement the MSC.

Author: Good point, and probably would've saved some trouble this week 🙂

Author: I've done exactly that, thank you for the advice! https://github.com/tcpipuk/matrix-spec-proposals/commit/4afb8b84e2879dbd85affaeb65e468eec9a3710d#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R278

- Servers may add/remove/modify fields in their own user's global profile data, whether for
  moderation purposes or for other policy reasons, e.g. to automatically populate a job title based
  on the user's organisation. Server administrators must make users aware that these fields exist

>   and that they will be made available to every server that this server federates with, if any.

Author: I've removed those lines here: https://github.com/tcpipuk/matrix-spec-proposals/commit/4afb8b84e2879dbd85affaeb65e468eec9a3710d#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R290-L298

- Servers are *allowed* to suppress some or all local/remote profile fields they do not wish to
  share with their local users (e.g. if there are moderation concerns during a go-live phase);
  however, server admins *must* disclose to users if they are publishing profile fields on behalf
  of a user over federation that they cannot see, as this could be considered a breach of trust.
  This feature could cause confusion if some users can see fields that other users cannot, so this

>   should be used sparingly.

Reviewer 7: In what situation would this occur? This seems extremely unusual. I'd almost say this seems like a bad practice thing to do, and contorts how profiles work.

Author: I believe Jim raised this as an option for large servers to roll out profile fields, i.e. they'd update to a server that supports them but want to block fields being seen/updated until they were ready.

I agree it feels like a bit of a hack, I'm just not sure whether there's a better way to soft-enable/disable profile fields, or if we should even do that at all?

Reviewer 7: Hmm, I'd like to understand what @jimmackenzie had in mind here. This feels like it raises the complexity of profiles AND reduces trust in profiles being accurate to me.

Author: In the meantime, I've rewritten this to be a lot less "servers can suppress whatever they like, whenever they like" and more "this entire feature can be temporarily disabled, but only sparingly": https://github.com/tcpipuk/matrix-spec-proposals/commit/4afb8b84e2879dbd85affaeb65e468eec9a3710d#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439R294-R296

>   should be used sparingly.

Reviewer 21: This is another stipulation on server admins rather than impls - see other comment.

Writing nitpick: I'd have put these (hiding fields from federated, adding additional fields to federated) as two separate things as this took me 3 reads to parse.

Lastly, I think sticking to the MUST/SHOULD/MAY specifiers would help this be clearer rather than going off-piste with things like, "are allowed".

Author: Agreed! I'll cut this down, and replace the use of "allowed" with "MAY" 🙂

Author: I've chopped down the fluff and used `MAY` here, thank you! https://github.com/tcpipuk/matrix-spec-proposals/commit/4afb8b84e2879dbd85affaeb65e468eec9a3710d#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439L300-R296

- This proposal focuses on introducing custom free-text fields in a controlled manner. Future
  extensions, such as fields with special behaviours or administrative controls, will be addressed
  in separate MSCs to allow for thorough community discussion and consideration of UI and

>   moderation impacts.

Reviewer 7: This sort of feels like a rehash of what **Key/Namespace Requirements for Custom Fields** is saying? I don't quite see how this is an implementation detail.

Author: This was in here very early in the MSC, and the rest has grown up around it 🙂

You're right this is a bit redundant now - I'll remove it!

Author: Done, thank you! https://github.com/tcpipuk/matrix-spec-proposals/commit/4afb8b84e2879dbd85affaeb65e468eec9a3710d#diff-e408d6c936a5ae3f7dd4386ed643830fb2191b459b35cbec73b2987d73cae439L307-L310

notify users and homeservers that these custom fields have been updated.

This proposal does not directly address reporting of user profiles over federation, but
[MSC4202](https://github.com/matrix-org/matrix-spec-proposals/pull/4202) offers a facility for
users to report offensive content to the homeserver that account belongs to. This proposal is not
dependent on [MSC4202](https://github.com/matrix-org/matrix-spec-proposals/pull/4202) but encourages
the use of moderation options to allow users to report offensive content.

## Alternatives

An alternative approach could involve introducing a completely new API for extended profile
information. However, this may lead to increased complexity for client and server implementations.

At the time of writing, Extensible Profiles as Rooms
([MSC1769](https://github.com/matrix-org/matrix-spec-proposals/pull/1769) and variant
[MSC4201](https://github.com/matrix-org/matrix-spec-proposals/pull/4201)) is under development for
richer and more granular content and privacy controls, which this proposal does not intend to
replace - this proposal focuses on basic global profile data without the complexity of per-room

> profile management.

Reviewer 2: > this proposal focuses on basic global profile data without the complexity of per-room
profile management.

Imho this proposal shouldnt be merged without the T&S considerations. If this is out of scope of this MSC this should be not merged before this is solved. T&S imho should be a MAJOR concern especially considering the recent spam waves we have seen...

Reviewer 11: I don't see how privacy controls and being able to use different profiles in different rooms would be relevant to spam waves? Those are privacy features, not anti-abuse features.

*This proposal aims to enhance the usability and richness of user profiles within the Matrix ecosystem by
introducing additional, customizable key:value pairs. This feature will facilitate the sharing of more
diverse user-defined public information across the federated network. The primary goal is to enable users to
publish a wide variety of possible information, such as preferred communication languages, pronouns,
public-facing organisation roles, or other relevant public details, thereby enriching the user interaction
experience without impacting existing functionalities.*

## Proposal

The Matrix protocol's current user profile structure supports very limited fields (`displayname` and
`avatar_url`). This proposal suggests expanding this structure to include an `extended` object capable of
storing arbitrary key:value pairs, thus allowing for a more versatile user profile.

### Client-Server API Changes

1. **GET `/_matrix/client/v3/profile/{userId}`**: This endpoint will be extended to include an `extended`
   field in the JSON response:

    ```json
    {
      "avatar_url": "mxc://matrix.org/MyC00lAvatar",
      "displayname": "John Doe",

>       "extended": {

Reviewer 8: Why put the new stuff under `extended`?  Why not make the entire profile a key:value store?  

This way just makes things more complicated.  Once the new fields become established, no one will remember which fields were the historical ones and which ones are the "new" extended ones, and it will be a source of confusion.

Author: My main concern was that this is user-entered data, so it'd be safer to keep the keys in a special object separate to the core profile information.

We also have a special API for "extended" fields because there's already separate ones for avatar and display name - making the fields look like established ones won't change those API endpoints immediately.

My hope was that this would get data structures added so as fields enter clients and become more common/standardised, future MSCs can then pick profile info that should be "promoted" to having state events, per-room data, etc.

Author: (I can certainly clarify these bits, but the reason the `/extended` API route returns an object containing the "extended" key is that it adds the ability for this to be improved in the future as the community desires)

Reviewer 8: errr...  Actually according to @clokep's comment above, it looks like the profile is *already* supposed to be a generic k:v store?

Author: There's no endpoint to update it though - it seems to be suggesting fields that the homeserver admin's authentication system would publish into the user profile.

This MSC is creating the ability for users to add their own custom metadata to their profiles.

Reviewer 11: I think a new key makes sense with the assumption that it doesn't modify the `m.room.member` event. There could be a future MSC to allow custom top-level fields that do go into member events. If this MSC used top-level fields for non-member-event stuff, it would cause conflicts.

Reviewer 8: > There's no endpoint to update it though 

Maybe this is an opportunity to make that change?

Personally I would love to see a `/_matrix/client/v3/profile/{userId}/{key}` replace the existing `/displayname` and `/avatar_url` endpoints.

Reviewer 9: > Personally I would love to see a `/_matrix/client/v3/profile/{userId}/{key}` replace the existing `/displayname` and `/avatar_url` endpoints.

This was also my initial thought. 👍

Author: Yeah, my main concern was about abuse, or complicating clients/timelines. People updating avatars and names is a big deal, but updating your work cellphone or your next vacation day is probably not so much.

I do think it would make sense for an MSC to follow that expands the array of per-room fields that can be copied to member events, but the intention of this account-wide set of freetext fields is to be as un-intrusive as possible for users until they actively visit a profile to see what is there.

Reviewer 7: I can see the merit in the `extended` field / API paths, but I think it's probably creating more overhead than necessary. I kinda see the profile logic here and the membership updating logic as separate issues and I don't think the object structure here necessarily has to line up with the membership event.

As @clokep says, the spec already specifies that only two fields should be copied to the event, so servers *shouldn't* be blindly copying all the fields. 

My feeling is rather it's enough to have the profile endpoints on an unstable path during the MSC process and see if the theoretical problems with abuse or complication actually manifests? I've written some thoughts on abuse in https://github.com/matrix-org/matrix-spec-proposals/pull/4133/files#r1576137660 which I think might also be more relevant here.

Author: I've rewritten this to place the fields at the root of the profile, and specified specifically only "avatar_url" and "displayname" should be copied into member events - does this sound good now?

        "field_name": "field_value",
        "field_2": ["one value", "another value"]
      }
    }
    ```

2. **GET `/_matrix/client/v3/profile/{userId}/extended`** will retrieve only the `extended` object. To ensure
   backward compatibility, it should *always* return an object when this user can update an extended profile,
   and return a 403 or 404 error when the server either will not allow or does not support them.

3. **PUT `/_matrix/client/v3/profile/{userId}/extended`** will accept a JSON object to replace the entire
   `extended` object, simplifying data management by allowing bulk updates instead of individual key

>    modifications:

Reviewer 11: A separate endpoint for updating individual keys may also be wanted to avoid race conditions, although I'm not sure how big of a problem that would be in practice

Author: My hope is that for a single user this wouldn't be a serious problem, and that in practice clients would likely GET while making a profile update to ensure no changes have been made.

I'm happy to include an extra endpoint, but likewise this might be one to raise a future MSC for if it actually proves to be an issue in practice?

Author: @tulir I've moved the "overwriting" method to `POST` so `PUT` will allow updating just a subset of keys to avoid race conditions - hopefully that looks better?

      }
    }
    ```

### Server-Server API Changes

1. **GET `/_matrix/federation/v1/query/profile`** will mirror the client-server API changes to ensure
   extended profile information is consistently available across the federated network.

### Implementation Details

- This feature will be implemented as optional but recommended, enabling a smooth transition and minimal
  disruption to existing deployments.
- The extended profile data will be public by default, and compliance with GDPR and other privacy regulations
  will be enforced, particularly in terms of data deletion and retention policies.

## Potential Issues

Initial challenges are primarily ensuring uniform support for the `extended` field across different servers
and clients during the rollout phase: users may come to expect that other users will check their supported
languages before communicating with them, but the sender's server does not support extended profiles.

As such, this MSC is designed to be as simple as possible to get initial functionality and data structures
implemented widely in both clients and servers, so further extensions can be debated, implemented, and tested
over time in the future.

> ## Alternatives

Reviewer 11: This should document that the `extended` object is *not* copied into `m.room.member` like other profile events, and perhaps mention some of the caveats caused by that fact (no per-room profiles, no historical profiles, clients need to be careful not to show profiles of banned users and such)

Author: Good point! Done 👍

An alternative could be to allow for more specific endpoint modifications or to introduce a completely new
API specifically for extended information. However, these approaches could lead to more significant changes
to the existing API and higher complexity for client developers.

## Security Considerations

Since the `extended` data is public, there are minimal security risks associated with the transmission of
sensitive information; however, it is critical to ensure that all data handled through these endpoints
complies with GDPR and other privacy regulations, particularly in the event of user data deletion.

It is crucial to ensure that clients inform users this data will be public, does not include sensitive
personal information, and complies with legal frameworks such as GDPR. Homeservers will be encouraged to
implement data caching strategies that do not exceed 24 hours to minimise the risk of unintended data
persistence.

## Unstable prefix

Until this proposal is accepted and finalized within the Matrix specification, implementers are encouraged to
use the `/unstable` prefix (see [MSC2324](https://github.com/matrix-org/matrix-doc/pull/2324)) for endpoints

> and a feature flag to manage client support.

Reviewer 11: This section is supposed to document the actual unstable prefixes to use, so in this case a prefix for the new endpoints and probably one for the `extended` field itself. (e.g. `/_matrix/client/unstable/uk.tcpip.msc4133/profile/{userId}/extended` and `uk.tcpip.msc4133.extended` for the field)

Author: Based on what @clokep mentioned above, is this field already allowed to exist in the profile according to the current spec?

I'd need to add unstable fields for `/profile/{userId}/extended` but would the main profile stay the same as the field is technically allowed to appear in the payload?

Reviewer 11: Custom fields always need to be prefixed, only standard fields (or fields within already-prefixed objects) can use stable prefixes. After the MSC is accepted, the unstable field would be replaced with the stable one.

Author: Got it, just writing this up now... so the object would be `uk.tcpip.msc4133.extended` in the unstable profile, then back to `extended` again in the stable version?

Author: Or did I misunderstand and the entire API route should be prefixed?

Author: I _think_ I've got this right now, thanks for the clarity!
